{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eafd7e01",
   "metadata": {},
   "source": [
    "## Reconstruct missing residues using Modeller\n",
    "Links:\n",
    "- [Comparative protein structure modeling using MODELLER](https://doi.org/10.1002/cpbi.3)\n",
    "\n",
    "- [Comparative modeling of protein 3D structures using MODELLER](https://salilab.org/modeller/wiki/Missing%20residues)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cafae907",
   "metadata": {},
   "source": [
    "### Get the sequence from the template (pdb file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4076cc2a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      ">P1;2xa0_A\n",
      "structureX:2xa0_A:10:A:+137:A:::-1.00:-1.00\n",
      "DNREIVMKYIHYKLSQRGYEWDVVHLTLRQAGDDFSRRYRRDFAEMSSQLHLTPFTARGRFATVVEELFRDGVNW\n",
      "GRIVAFFEFGGVMCVESVNREMSPLVDNIALWMTEYLNRHLHTWIQDNGGWDAFVELYGPSM*\n"
     ]
    }
   ],
   "source": [
    "# Get the sequence of the target file (2xa0_A.PDB), and write to an alignment file\n",
    "from modeller import *\n",
    "code = '2xa0_A'     # 2xa0_A.pdb\n",
    "\n",
    "e = Environ()\n",
    "m = Model(e, file=code)\n",
    "aln = Alignment(e)\n",
    "aln.append_model(m, align_codes=code)\n",
    "outSeq = \"2xa0_A.seq\"\n",
    "aln.write(file=outSeq)\n",
    "[print (l,end=\"\") for l in open (outSeq).readlines()];"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3c0d56a",
   "metadata": {},
   "source": [
    "## Build structure using automatic modelling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3061a86",
   "metadata": {},
   "source": [
    "Because either AutoModel or LoopModel will build a comparative model using your input PDB as a template, potentially all of the atoms in your final model could move. If you really don't want the non-missing residues to move, you can override the select_atoms method to select only the missing residues with a script similar to that below (note that the residue numbers are off by 1, since Modeller numbers the model starting at 1 in chain A, while the original PDB started numbering at 2):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6ebdef53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "openf___224_> Open           $(LIB)/restyp.lib\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/resgrp.lib\n",
      "rdresgr_266_> Number of residue groups:        2\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/sstruc.lib\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:       953673     931.321     0.909\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:       954201     931.837     0.910\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/resdih.lib\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:      1002801     979.298     0.956\n",
      "rdrdih__263_> Number of dihedral angle types         :        9\n",
      "              Maximal number of dihedral angle optima:        3\n",
      "              Dihedral angle names                   :  Alph Phi Psi Omeg chi1 chi2 chi3 chi4 chi5\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/radii.lib\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:      1016101     992.286     0.969\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/af_mnchdef.lib\n",
      "rdwilmo_274_> Mainchain residue conformation classes:  APBLE\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/mnch.lib\n",
      "rdclass_257_> Number of classes:        5\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/mnch1.lib\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/mnch2.lib\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/mnch3.lib\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/xs4.mat\n",
      "rdrrwgh_268_> Number of residue types:       21\n",
      "openf___224_> Open           modeller.ali\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      1044532    1020.051     0.996\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      1059624    1034.789     1.011\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      1075024    1049.828     1.025\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      1105824    1079.906     1.055\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      1167424    1140.062     1.113\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1167968    1140.594     1.114\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1168792    1141.398     1.115\n",
      "\n",
      "Read the alignment from file       : modeller.ali\n",
      "\n",
      "Total number of alignment positions:   207\n",
      "\n",
      "  #  Code        #_Res #_Segm PDB_code    Name\n",
      "-------------------------------------------------------------------------------\n",
      "  1   Template     137      1      2xa0_A\n",
      "  2     Target     207      1\n",
      "check_a_343_> >> BEGINNING OF COMMAND\n",
      "openf___224_> Open           ./2xa0_A.pdb\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1211617    1183.220     1.155\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1211617    1183.220     1.155\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1211788    1183.387     1.156\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1212638    1184.217     1.156\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1213896    1185.445     1.158\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1215800    1187.305     1.159\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1218656    1190.094     1.162\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1218656    1190.094     1.162\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1218746    1190.182     1.162\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1223030    1194.365     1.166\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1223030    1194.365     1.166\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1223165    1194.497     1.167\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1229591    1200.772     1.173\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1229591    1200.772     1.173\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1229789    1200.966     1.173\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1239411    1210.362     1.182\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1239411    1210.362     1.182\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1239708    1210.652     1.182\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1239708    1210.652     1.182\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1240158    1211.092     1.183\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1254608    1225.203     1.196\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      1254491    1225.089     1.196\n",
      "\n",
      "check_ali___> Checking the sequence-structure alignment. \n",
      "\n",
      "Implied intrachain target CA(i)-CA(i+1) distances longer than  8.0 angstroms:\n",
      "\n",
      "ALN_POS  TMPL  RID1  RID2  NAM1  NAM2     DIST\n",
      "----------------------------------------------\n",
      "END OF TABLE\n",
      "check_a_344_> << END OF COMMAND\n",
      "openf___224_> Open           ${LIB}/top_heav.lib\n",
      "read_to_681_> topology.submodel read from topology file:        3\n",
      "openf___224_> Open           ${LIB}/par.lib\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1393743    1361.077     1.329\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1395343    1362.640     1.331\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1399323    1366.526     1.334\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1401323    1368.479     1.336\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1404323    1371.409     1.339\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1408823    1375.804     1.344\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1414395    1381.245     1.349\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1417195    1383.979     1.352\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1421395    1388.081     1.356\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1427695    1394.233     1.362\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1430467    1396.940     1.364\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1431867    1398.308     1.366\n",
      "read_pa_232_> parameters    BONDS   ANGLS  DIHEDS IMPROPS    MODE\n",
      "                              227     561     661     112       0\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      1464209    1429.892     1.396\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      1624047    1585.983     1.549\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      1648831    1610.187     1.572\n",
      "\n",
      "mkilst______> segment topology constructed from sequence and RTF:\n",
      "              segments residues  atoms  bonds angles dihedrals impropers:\n",
      "                     1      207   1612   1664      0         0       717\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      1719519    1679.218     1.640\n",
      "\n",
      "patch_______> segment topology patched using RTF:  1 ;  MET  ;  NTER\n",
      "              segments residues  atoms  bonds angles dihedrals impropers:\n",
      "                     1      207   1612   1664   2269      2718       717\n",
      "\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      1796295    1754.194     1.713\n",
      "patch_______> segment topology patched using RTF:  207 ;  ARG  ;  CTER\n",
      "              segments residues  atoms  bonds angles dihedrals impropers:\n",
      "                     1      207   1613   1665   2271      2720       718\n",
      "\n",
      "genseg______> segment topology constructed from sequence and RTF:\n",
      "              segments residues  atoms  bonds angles dihedrals impropers:\n",
      "                     1      207   1613   1665   2271      2720       718\n",
      "transfe_506_> MODEL is an average of all templates.\n",
      "transfe_511_> Number of templates for coordinate transfer:        1\n",
      "              After transferring coordinates of the equivalent template atoms,\n",
      "              there are defined, undefined atoms in MODEL:      854      759\n",
      "openf___224_> Open           Target.ini\n",
      "wrpdb___568_> Residues, atoms, selected atoms:      207     1613     1613\n",
      "make_re_417_> Restraint type to be calculated:  stereo\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      1903744    1859.125     1.816\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      1911936    1867.125     1.823\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      1928320    1883.125     1.839\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      1961088    1915.125     1.870\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      1993856    1947.125     1.901\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2059392    2011.125     1.964\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2092160    2043.125     1.995\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2124928    2075.125     2.026\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2190464    2139.125     2.089\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2321536    2267.125     2.214\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2387072    2331.125     2.276\n",
      "r_stere_606_> Stereochemical restraints were constructed from RTF & PRMF.\n",
      "              Added bond,angle,dihedral,improper restraints  :     1665    2271    2657     700\n",
      "              Total number of restraints before, now         :        0     7293\n",
      "make_re_422_> Number of previous, current restraints         :        0     7293\n",
      "make_re_423_> Number of previous, current selected restraints:        0     7293\n",
      "make_re_417_> Restraint type to be calculated:  phi-psi_binormal\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      2504912    2446.203     2.389\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/mnch1.bin\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/mnch1.mdt\n",
      "getdata_643_> Protein accepted:  Template\n",
      "getdata_289_> Proteins (all/accepted):        1        1\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2570464    2510.219     2.451\n",
      "make_re_422_> Number of previous, current restraints         :     7293     7498\n",
      "make_re_423_> Number of previous, current selected restraints:     7293     7498\n",
      "make_re_417_> Restraint type to be calculated:  omega_dihedral\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      2570464    2510.219     2.451\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/omega.bin\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/omega.mdt\n",
      "getdata_643_> Protein accepted:  Template\n",
      "getdata_289_> Proteins (all/accepted):        1        1\n",
      "omgdel__425_> Unselected all O C +N +CA dihedrals:      224\n",
      "              (This is to avoid clashes between STEREO\n",
      "              and OMEGA_DIHEDRAL restraints)\n",
      "make_re_422_> Number of previous, current restraints         :     7498     7704\n",
      "make_re_423_> Number of previous, current selected restraints:     7498     7480\n",
      "make_re_417_> Restraint type to be calculated:  chi1_dihedral\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      2570464    2510.219     2.451\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/chi1234.bin\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/chi1.mdt\n",
      "getdata_643_> Protein accepted:  Template\n",
      "getdata_289_> Proteins (all/accepted):        1        1\n",
      "make_re_422_> Number of previous, current restraints         :     7704     7868\n",
      "make_re_423_> Number of previous, current selected restraints:     7480     7644\n",
      "make_re_417_> Restraint type to be calculated:  chi2_dihedral\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      2570464    2510.219     2.451\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/chi1234.bin\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/chi2.mdt\n",
      "getdata_643_> Protein accepted:  Template\n",
      "getdata_289_> Proteins (all/accepted):        1        1\n",
      "make_re_422_> Number of previous, current restraints         :     7868     7994\n",
      "make_re_423_> Number of previous, current selected restraints:     7644     7770\n",
      "make_re_417_> Restraint type to be calculated:  chi3_dihedral\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      2570464    2510.219     2.451\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/chi1234.bin\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/chi3.mdt\n",
      "getdata_643_> Protein accepted:  Template\n",
      "getdata_289_> Proteins (all/accepted):        1        1\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2701536    2638.219     2.576\n",
      "make_re_422_> Number of previous, current restraints         :     7994     8054\n",
      "make_re_423_> Number of previous, current selected restraints:     7770     7830\n",
      "make_re_417_> Restraint type to be calculated:  chi4_dihedral\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      2701536    2638.219     2.576\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/chi1234.bin\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/chi4.mdt\n",
      "mdtrsr__446W> A potential that relies on one protein is used, yet you have at\n",
      "              least one known structure available. MDT, not library, potential is used.\n",
      "getdata_643_> Protein accepted:  Template\n",
      "getdata_289_> Proteins (all/accepted):        1        1\n",
      "make_re_422_> Number of previous, current restraints         :     8054     8073\n",
      "make_re_423_> Number of previous, current selected restraints:     7830     7849\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "\n",
      "Dynamically allocated memory at amaxhash_contac [B,KiB,MiB]:      3166044    3091.840     3.019\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      3297116    3219.840     3.144\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      3559260    3475.840     3.394\n",
      "make_re_422_> Number of previous, current restraints         :     8073    10470\n",
      "make_re_423_> Number of previous, current selected restraints:     7849    10246\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      3690332    3603.840     3.519\n",
      "make_re_422_> Number of previous, current restraints         :    10470    12792\n",
      "make_re_423_> Number of previous, current selected restraints:    10246    12568\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "make_re_422_> Number of previous, current restraints         :    12792    13961\n",
      "make_re_423_> Number of previous, current selected restraints:    12568    13737\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "make_re_422_> Number of previous, current restraints         :    13961    14196\n",
      "make_re_423_> Number of previous, current selected restraints:    13737    13972\n",
      "0 atoms in HETATM/BLK residues constrained\n",
      "to protein atoms within 2.30 angstroms\n",
      "and protein CA atoms within 10.00 angstroms\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "make_re_422_> Number of previous, current restraints         :    14196    14196\n",
      "make_re_423_> Number of previous, current selected restraints:    13972    13972\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "make_re_422_> Number of previous, current restraints         :    14196    14196\n",
      "make_re_423_> Number of previous, current selected restraints:    13972    13972\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "make_re_422_> Number of previous, current restraints         :    14196    14196\n",
      "make_re_423_> Number of previous, current selected restraints:    13972    13972\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "make_re_422_> Number of previous, current restraints         :    14196    14196\n",
      "make_re_423_> Number of previous, current selected restraints:    13972    13972\n",
      "0 atoms in residues without defined topology\n",
      "constrained to be rigid bodies\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "make_re_422_> Number of previous, current restraints         :    14196    14196\n",
      "make_re_423_> Number of previous, current selected restraints:    13972    13972\n",
      "rmdupl__427_>     1279 redundant cosine dihedral restraints were unselected.\n",
      "condens_443_> Restraints marked for deletion were removed.\n",
      "              Total number of restraints before, now:    14196    12693\n",
      "openf___224_> Open           Target.rsr\n",
      "openf___224_> Open           Target.rsr\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      3604812    3520.324     3.438\n",
      "rdcsr2__307_> Number of restraints read    :    12693\n",
      "              Number of excluded pairs read:        0\n",
      "              Number of pseudo atoms read  :        0\n",
      "rdcsrs__304_> Restraints in memory, selected restraints:    12693    12693\n",
      "              Explicitly excluded atom pairs in memory :        0\n",
      "              Pseudo atoms in memory                   :        0\n",
      "457 (of 1613 total) atoms selected for optimization\n",
      "openf___224_> Open           Target.ini\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      3604844    3520.355     3.438\n",
      "read_mo_297_> Segments, residues, atoms:        1      207     1613\n",
      "read_mo_298_> Segment:        1  1:A  207:A     1613\n",
      "randomi_498_> Atoms,selected atoms,random_seed,amplitude:     1613      457        1        4.0000\n",
      "randomi_496_> Amplitude is > 0; randomization is done.\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1927\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1933\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1938\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "conmin__610W> Precision problem; starting from scratch.\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1956\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      3936128    3843.875     3.754\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "iupac_m_486_> OE1/2 will be swapped:     -105.1969       13  13\n",
      "iupac_m_487_> NH1/2 swapped:      179.7925       26  26\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:      169.7476       28  28\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_485_> OD1/2 will be swapped:      107.2833       34  34\n",
      "iupac_m_487_> NH1/2 swapped:      179.9441       98  98\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      103  103\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0163      104  104\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      107  107\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      109  109\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      110  110\n",
      "iupac_m_485_> OD1/2 will be swapped:     -179.9802      111  111\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      114  114\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0292      124  124\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9657      127  127\n",
      "iupac_m_487_> NH1/2 swapped:      179.9558      129  129\n",
      "iupac_m_486_> OE1/2 will be swapped:      154.4214      135  135\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      136  136\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0239      138  138\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9558      139  139\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9720      140  140\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9407      146  146\n",
      "iupac_m_486_> OE1/2 will be swapped:      179.9604      152  152\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      160  160\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      164  164\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      165  165\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      179  179\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:     -176.2274      180  180\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -180.0000      183  183\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      196  196\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      200  200\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0309      202  202\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      207  207\n",
      "\n",
      "\n",
      ">> ENERGY; Differences between the model's features and restraints:\n",
      "Number of all residues in MODEL                   :      207\n",
      "Number of all, selected real atoms                :     1613     457\n",
      "Number of all, selected pseudo atoms              :        0       0\n",
      "Number of all static, selected restraints         :    12693    1983\n",
      "COVALENT_CYS                                      :        F\n",
      "NONBONDED_SEL_ATOMS                               :        1\n",
      "Number of non-bonded pairs (excluding 1-2,1-3,1-4):      598\n",
      "Dynamic pairs routine                             : 5, NATM_SEL x NATM double loop\n",
      "Atomic shift for contacts update (UPDATE_DYNAMIC) :    0.390\n",
      "LENNARD_JONES_SWITCH                              :    6.500   7.500\n",
      "COULOMB_JONES_SWITCH                              :    6.500   7.500\n",
      "RESIDUE_SPAN_RANGE                                :        0   99999\n",
      "NLOGN_USE                                         :       15\n",
      "CONTACT_SHELL                                     :    4.000\n",
      "DYNAMIC_PAIRS,_SPHERE,_COULOMB,_LENNARD,_MODELLER :        T       T       F       F       F\n",
      "SPHERE_STDV                                       :    0.050\n",
      "RADII_FACTOR                                      :    0.820\n",
      "Current energy                                    :         303.1090\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Summary of the restraint violations: \n",
      "\n",
      "   NUM     ... number of restraints.\n",
      "   NUMVI   ... number of restraints with RVIOL > VIOL_REPORT_CUT[i].\n",
      "   RVIOL   ... relative difference from the best value.\n",
      "   NUMVP   ... number of restraints with -Ln(pdf) > VIOL_REPORT_CUT2[i].\n",
      "   RMS_1   ... RMS(feature, minimally_violated_basis_restraint, NUMB).\n",
      "   RMS_2   ... RMS(feature, best_value, NUMB).\n",
      "   MOL.PDF ... scaled contribution to -Ln(Molecular pdf).\n",
      "\n",
      " #                     RESTRAINT_GROUP      NUM   NUMVI  NUMVP   RMS_1   RMS_2         MOL.PDF     S_i\n",
      "------------------------------------------------------------------------------------------------------\n",
      " 1 Bond length potential              :     478       0      0   0.003   0.003      1.0271       1.000\n",
      " 2 Bond angle potential               :     667       0      1   2.090   2.090      56.133       1.000\n",
      " 3 Stereochemical cosine torsion poten:     343       0      4  51.997  51.997      123.83       1.000\n",
      " 4 Stereochemical improper torsion pot:     181       0      0   0.759   0.759      2.3100       1.000\n",
      " 5 Soft-sphere overlap restraints     :     598       0      0   0.005   0.005      1.7397       1.000\n",
      " 6 Lennard-Jones 6-12 potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 7 Coulomb point-point electrostatic p:       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 8 H-bonding potential                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 9 Distance restraints 1 (CA-CA)      :      29       0      0   0.343   0.343     0.45127       1.000\n",
      "10 Distance restraints 2 (N-O)        :      30       0      0   1.070   1.070      3.3529       1.000\n",
      "11 Mainchain Phi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "12 Mainchain Psi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "13 Mainchain Omega dihedral restraints:      70       0      1   2.650   2.650      5.7983       1.000\n",
      "14 Sidechain Chi_1 dihedral restraints:      43       0      0  89.703  89.703      12.513       1.000\n",
      "15 Sidechain Chi_2 dihedral restraints:      31       0      0  75.359  75.359      8.4806       1.000\n",
      "16 Sidechain Chi_3 dihedral restraints:      21       0      0  53.696  53.696      5.3615       1.000\n",
      "17 Sidechain Chi_4 dihedral restraints:       3       0      0 145.906 145.906      1.6359       1.000\n",
      "18 Disulfide distance restraints      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "19 Disulfide angle restraints         :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "20 Disulfide dihedral angle restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "21 Lower bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "22 Upper bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "23 Distance restraints 3 (SDCH-MNCH)  :      16       0      0   0.742   0.742     0.15227       1.000\n",
      "24 Sidechain Chi_5 dihedral restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "25 Phi/Psi pair of dihedral restraints:      71      52      9  27.026 148.668      80.327       1.000\n",
      "26 Distance restraints 4 (SDCH-SDCH)  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "27 Distance restraints 5 (X-Y)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "28 NMR distance restraints 6 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "29 NMR distance restraints 7 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "30 Minimal distance restraints        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "31 Non-bonded restraints              :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "32 Atomic accessibility restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "33 Atomic density restraints          :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "34 Absolute position restraints       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "35 Dihedral angle difference restraint:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "36 GBSA implicit solvent potential    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "37 EM density fitting potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "38 SAXS restraints                    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "39 Symmetry restraints                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "\n",
      "\n",
      "\n",
      "openf___224_> Open           Target.V99990001\n",
      "# Heavy relative violation of each residue is written to: Target.V99990001\n",
      "# The profile is NOT normalized by the number of restraints.\n",
      "# The profiles are smoothed over a window of residues:    1\n",
      "# The sum of all numbers in the file:   13089.0938\n",
      "\n",
      "\n",
      "\n",
      "List of the violated restraints:\n",
      "   A restraint is violated when the relative difference\n",
      "   from the best value (RVIOL) is larger than CUTOFF.\n",
      "\n",
      "   ICSR   ... index of a restraint in the current set.\n",
      "   RESNO  ... residue numbers of the first two atoms.\n",
      "   ATM    ... IUPAC atom names of the first two atoms.\n",
      "   FEAT   ... the value of the feature in the model.\n",
      "   restr  ... the mean of the basis restraint with the smallest\n",
      "              difference from the model (local minimum).\n",
      "   viol   ... difference from the local minimum.\n",
      "   rviol  ... relative difference from the local minimum.\n",
      "   RESTR  ... the best value (global minimum).\n",
      "   VIOL   ... difference from the best value.\n",
      "   RVIOL  ... relative difference from the best value.\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 25                           : Phi/Psi pair of dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       6.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   5791   1M   2A C   N       7    9 -123.70 -134.00   10.31    0.32  -62.50 -176.84   33.29\n",
      "    1          2A   2A N   CA      9   10  146.47  147.00                  -40.90\n",
      "    2   5792   2A   3H C   N      12   14  -94.47  -67.60   27.37    1.90  -63.20  175.32   23.95\n",
      "    2          3H   3H N   CA     14   15  145.19  140.00                  -42.30\n",
      "    3   5793   3H   4A C   N      22   24 -121.18 -134.00   12.87    0.47  -62.50 -179.26   32.75\n",
      "    3          4A   4A N   CA     24   25  148.15  147.00                  -40.90\n",
      "    4   5794   4A   5G C   N      27   29  177.78 -167.20   15.02    0.28   82.20 -168.23   13.57\n",
      "    4          5G   5G N   CA     29   30  174.76  174.60                    8.50\n",
      "    5   5795   5G   6R C   N      31   33  -74.20  -72.10    7.25    0.51  -63.00  170.43   23.78\n",
      "    5          6R   6R N   CA     33   34  148.84  141.90                  -41.10\n",
      "    6   5796   6R   7T C   N      42   44 -117.17 -124.80    9.65    0.58  -63.20  176.93   26.28\n",
      "    6          7T   7T N   CA     44   45  149.40  143.50                  -42.10\n",
      "    7   5797   7T   8G C   N      49   51  125.45   78.70   49.63    1.99   82.20  174.19   10.55\n",
      "    7          8G   8G N   CA     51   52  177.24 -166.10                    8.50\n",
      "    8   5798   8G   9Y C   N      53   55 -135.95 -124.30   47.82    2.08  -63.50  153.05   27.79\n",
      "    8          9Y   9Y N   CA     55   56 -178.22  135.40                  -43.40\n",
      "    9   5799   9Y  10D C   N      65   67 -105.57  -63.30   48.41    8.27  -63.30   48.41    8.27\n",
      "    9         10D  10D N   CA     67   68  -63.60  -40.00                  -40.00\n",
      "   10   5821  31D  32A C   N     265  267  -80.42  -68.20   13.04    0.94  -62.50  170.20   28.81\n",
      "   10         32A  32A N   CA    267  268  149.85  145.30                  -40.90\n",
      "   11   5822  32A  33G C   N     270  272  178.49 -167.20   47.60    1.96   82.20  154.40   11.54\n",
      "   11         33G  33G N   CA    272  273  129.19  174.60                    8.50\n",
      "   12   5823  33G  34D C   N     274  276 -148.24  -96.50   58.88    2.44  -63.30 -163.04   29.79\n",
      "   12         34D  34D N   CA    276  277  142.30  114.20                  -40.00\n",
      "   13   5825  35V  36G C   N     289  291 -151.37 -167.20   23.23    0.54   82.20 -164.52   14.72\n",
      "   13         36G  36G N   CA    291  292  157.59  174.60                    8.50\n",
      "   14   5826  36G  37A C   N     293  295  -73.38  -68.20    7.17    0.45  -62.50  169.19   28.27\n",
      "   14         37A  37A N   CA    295  296  150.26  145.30                  -40.90\n",
      "   15   5827  37A  38A C   N     298  300 -124.60 -134.00    9.43    0.34  -62.50 -177.71   33.19\n",
      "   15         38A  38A N   CA    300  301  147.72  147.00                  -40.90\n",
      "   16   5829  39P  40P C   N     310  312  -54.67  -58.70    6.51    0.41  -64.50  177.46   13.77\n",
      "   16         40P  40P N   CA    312  313  -35.61  -30.50                  147.20\n",
      "   17   5830  40P  41G C   N     317  319   85.08   78.70   27.37    0.53   82.20  158.80    7.85\n",
      "   17         41G  41G N   CA    319  320  167.28 -166.10                    8.50\n",
      "   18   5831  41G  42A C   N     321  323 -139.49 -134.00    8.44    0.28  -62.50 -177.29   34.07\n",
      "   18         42A  42A N   CA    323  324  153.41  147.00                  -40.90\n",
      "   19   5832  42A  43A C   N     326  328  -71.42  -68.20    6.06    0.39  -62.50  168.90   28.13\n",
      "   19         43A  43A N   CA    328  329  150.44  145.30                  -40.90\n",
      "   20   5833  43A  44P C   N     331  333  -65.74  -58.70    8.07    0.74  -64.50  173.75   13.03\n",
      "   20         44P  44P N   CA    333  334  -26.55  -30.50                  147.20\n",
      "   21   5834  44P  45A C   N     338  340 -142.05 -134.00    9.70    0.24  -62.50 -175.28   34.53\n",
      "   21         45A  45A N   CA    340  341  152.39  147.00                  -40.90\n",
      "   22   5835  45A  46P C   N     343  345  -69.14  -58.70   14.62    1.04  -64.50  167.53   12.75\n",
      "   22         46P  46P N   CA    345  346  -20.26  -30.50                  147.20\n",
      "   23   5836  46P  47G C   N     350  352   94.33   78.70   46.93    0.77   82.20  141.66    7.48\n",
      "   23         47G  47G N   CA    352  353  149.65 -166.10                    8.50\n",
      "   24   5838  48I  49F C   N     362  364 -110.81 -124.20   15.90    0.42  -63.20 -174.74   22.76\n",
      "   24         49F  49F N   CA    364  365  134.73  143.30                  -44.30\n",
      "   25   5839  49F  50S C   N     373  375 -133.88 -136.60    2.80    0.13  -64.10 -173.37   18.47\n",
      "   25         50S  50S N   CA    375  376  151.90  151.20                  -35.00\n",
      "   26   5840  50S  51S C   N     379  381 -124.72 -136.60   12.73    0.40  -64.10 -171.59   17.90\n",
      "   26         51S  51S N   CA    381  382  146.61  151.20                  -35.00\n",
      "   27   5841  51S  52Q C   N     385  387 -113.53 -121.10   11.14    0.37  -63.80  178.88   23.99\n",
      "   27         52Q  52Q N   CA    387  388  131.53  139.70                  -40.30\n",
      "   28   5843  53P  54G C   N     401  403 -157.76 -167.20   24.69    0.80   82.20 -173.08   14.06\n",
      "   28         54G  54G N   CA    403  404  151.78  174.60                    8.50\n",
      "   29   5844  54G  55H C   N     405  407 -130.83 -125.60   17.68    0.56  -63.20  175.56   26.37\n",
      "   29         55H  55H N   CA    407  408  155.69  138.80                  -42.30\n",
      "   30   5845  55H  56T C   N     415  417 -111.80 -124.80   15.44    0.57  -63.20 -176.19   20.74\n",
      "   30         56T  56T N   CA    417  418  135.17  143.50                  -42.10\n",
      "   31   5847  57P  58H C   N     429  431 -124.95 -125.60   20.57    0.77  -63.20  172.01   18.58\n",
      "   31         58H  58H N   CA    431  432  118.24  138.80                  -42.30\n",
      "   32   5849  59P  60A C   N     446  448  -92.88  -68.20   30.28    2.98  -62.50  171.38   26.76\n",
      "   32         60A  60A N   CA    448  449  127.77  145.30                  -40.90\n",
      "   33   5850  60A  61A C   N     451  453 -161.51 -134.00   29.51    0.68  -62.50 -170.61   36.31\n",
      "   33         61A  61A N   CA    453  454  157.66  147.00                  -40.90\n",
      "   34   5851  61A  62S C   N     456  458  -73.44  -72.40   15.44    0.94  -64.10  172.25   11.96\n",
      "   34         62S  62S N   CA    458  459  137.00  152.40                  -35.00\n",
      "   35   5852  62S  63R C   N     462  464 -131.69 -125.20   14.04    0.77  -63.00 -177.35   21.06\n",
      "   35         63R  63R N   CA    464  465  128.15  140.60                  -41.10\n",
      "   36   5853  63R  64D C   N     473  475 -115.17  -63.30   51.92    8.34  -63.30   51.92    8.34\n",
      "   36         64D  64D N   CA    475  476  -42.32  -40.00                  -40.00\n",
      "   37   5856  66V  67A C   N     495  497 -126.60 -134.00   16.13    0.68  -62.50 -174.98   27.98\n",
      "   37         67A  67A N   CA    497  498  132.66  147.00                  -40.90\n",
      "   38   5858  68R  69T C   N     511  513 -107.61 -124.80   21.56    0.75  -63.20  178.21   20.24\n",
      "   38         69T  69T N   CA    513  514  130.49  143.50                  -42.10\n",
      "   39   5859  69T  70S C   N     518  520  -82.85  -72.40   27.97    1.34  -64.10  147.85   11.93\n",
      "   39         70S  70S N   CA    520  521  178.35  152.40                  -35.00\n",
      "   40   5861  71P  72L C   N     531  533 -111.15 -108.50   13.64    0.71  -63.50  179.36   27.83\n",
      "   40         72L  72L N   CA    533  534  145.89  132.50                  -41.20\n",
      "   41   5862  72L  73Q C   N     539  541 -115.21 -121.10   29.65    1.49  -63.80  159.46   26.18\n",
      "   41         73Q  73Q N   CA    541  542  168.75  139.70                  -40.30\n",
      "   42   5863  73Q  74T C   N     548  550  -62.63  -78.10   15.47    0.94  -63.20  167.97   21.50\n",
      "   42         74T  74T N   CA    550  551  149.93  149.80                  -42.10\n",
      "   43   5865  75P  76A C   N     562  564  -66.96  -68.20    8.24    0.62  -62.50  178.11   28.97\n",
      "   43         76A  76A N   CA    564  565  137.16  145.30                  -40.90\n",
      "   44   5866  76A  77A C   N     567  569 -131.17 -134.00   12.17    0.61  -62.50 -171.02   28.52\n",
      "   44         77A  77A N   CA    569  570  135.17  147.00                  -40.90\n",
      "   45   5868  78P  79G C   N     579  581 -126.77 -167.20  102.01    3.27   82.20  167.50   12.99\n",
      "   45         79G  79G N   CA    581  582   80.94  174.60                    8.50\n",
      "   46   5869  79G  80A C   N     583  585  176.45 -134.00   53.17    1.22  -62.50 -165.04   38.27\n",
      "   46         80A  80A N   CA    585  586  166.27  147.00                  -40.90\n",
      "   47   5870  80A  81A C   N     588  590 -124.25 -134.00    9.85    0.38  -62.50 -178.51   33.04\n",
      "   47         81A  81A N   CA    590  591  148.44  147.00                  -40.90\n",
      "   48   5871  81A  82A C   N     593  595 -117.21 -134.00   16.80    0.58  -62.50 -179.95   32.42\n",
      "   48         82A  82A N   CA    595  596  147.57  147.00                  -40.90\n",
      "   49   5872  82A  83G C   N     598  600  174.31 -167.20   21.71    0.37   82.20 -160.04   13.88\n",
      "   49         83G  83G N   CA    600  601 -174.02  174.60                    8.50\n",
      "   50   5874  84P  85A C   N     609  611  -49.01  -68.20   25.13    1.60  -62.50  170.51   28.63\n",
      "   50         85A  85A N   CA    611  612  129.08  145.30                  -40.90\n",
      "   51   5876  86L  87S C   N     622  624  -83.53  -72.40   24.38    1.13  -64.10  152.15   12.28\n",
      "   51         87S  87S N   CA    624  625  174.09  152.40                  -35.00\n",
      "   52   5880  90P  91P C   N     649  651  -80.58  -64.50   28.64    2.97  -58.70  155.55   11.17\n",
      "   52         91P  91P N   CA    651  652  123.50  147.20                  -30.50\n",
      "\n",
      "\n",
      "report______> Distribution of short non-bonded contacts:\n",
      "\n",
      "  serious non-bonded atom clash:    90 1354   1.714\n",
      "  serious non-bonded atom clash:   122 1320   2.215\n",
      "  serious non-bonded atom clash:   144  727   2.264\n",
      "  serious non-bonded atom clash:   151 1196   1.945\n",
      "  serious non-bonded atom clash:   190 1197   1.758\n",
      "  serious non-bonded atom clash:   664  694   2.295\n",
      "  serious non-bonded atom clash:   671  702   1.565\n",
      "  serious non-bonded atom clash:   719 1120   2.189\n",
      "  serious non-bonded atom clash:   720 1119   1.741\n",
      "  serious non-bonded atom clash:   721 1119   1.648\n",
      "  serious non-bonded atom clash:   726 1152   1.768\n",
      "  serious non-bonded atom clash:   726 1154   2.269\n",
      "  serious non-bonded atom clash:   727 1152   2.294\n",
      "  serious non-bonded atom clash:   728 1152   2.162\n",
      "  serious non-bonded atom clash:   728 1154   1.683\n",
      "  serious non-bonded atom clash:   729 1152   2.041\n",
      "  serious non-bonded atom clash:   729 1153   2.155\n",
      "  serious non-bonded atom clash:   729 1154   1.124\n",
      "  serious non-bonded atom clash:   757  797   2.239\n",
      "  serious non-bonded atom clash:   757  798   2.053\n",
      "  serious non-bonded atom clash:   809  817   2.094\n",
      "  serious non-bonded atom clash:   823  827   2.273\n",
      "  serious non-bonded atom clash:   882 1210   2.153\n",
      "  serious non-bonded atom clash:  1000 1371   1.984\n",
      "  serious non-bonded atom clash:  1019 1041   1.587\n",
      "  serious non-bonded atom clash:  1019 1043   1.762\n",
      "  serious non-bonded atom clash:  1063 1067   2.231\n",
      "  serious non-bonded atom clash:  1111 1455   1.857\n",
      "  serious non-bonded atom clash:  1111 1457   1.212\n",
      "  serious non-bonded atom clash:  1111 1459   1.921\n",
      "  serious non-bonded atom clash:  1139 1383   2.285\n",
      "  serious non-bonded atom clash:  1141 1382   2.235\n",
      "  serious non-bonded atom clash:  1141 1383   2.023\n",
      "  serious non-bonded atom clash:  1168 1342   1.700\n",
      "  serious non-bonded atom clash:  1168 1343   2.160\n",
      "  serious non-bonded atom clash:  1169 1340   1.956\n",
      "  serious non-bonded atom clash:  1169 1342   2.045\n",
      "  serious non-bonded atom clash:  1169 1343   1.522\n",
      "  serious non-bonded atom clash:  1169 1345   2.168\n",
      "  serious non-bonded atom clash:  1170 1312   1.597\n",
      "  serious non-bonded atom clash:  1170 1339   2.166\n",
      "  serious non-bonded atom clash:  1170 1340   0.933\n",
      "  serious non-bonded atom clash:  1170 1342   2.001\n",
      "  serious non-bonded atom clash:  1170 1343   0.763\n",
      "  serious non-bonded atom clash:  1170 1344   1.750\n",
      "  serious non-bonded atom clash:  1170 1345   1.567\n",
      "  serious non-bonded atom clash:  1170 1346   2.186\n",
      "  serious non-bonded atom clash:  1170 1347   2.117\n",
      "  serious non-bonded atom clash:  1172 1310   2.198\n",
      "  serious non-bonded atom clash:  1172 1312   1.497\n",
      "  serious non-bonded atom clash:  1172 1313   2.013\n",
      "  serious non-bonded atom clash:  1172 1314   1.631\n",
      "  serious non-bonded atom clash:  1172 1340   1.411\n",
      "  serious non-bonded atom clash:  1172 1343   2.110\n",
      "  serious non-bonded atom clash:  1172 1344   0.608\n",
      "  serious non-bonded atom clash:  1172 1346   1.043\n",
      "  serious non-bonded atom clash:  1172 1347   1.878\n",
      "  serious non-bonded atom clash:  1174 1313   2.173\n",
      "  serious non-bonded atom clash:  1174 1315   2.177\n",
      "  serious non-bonded atom clash:  1174 1344   1.525\n",
      "  serious non-bonded atom clash:  1174 1346   1.595\n",
      "  serious non-bonded atom clash:  1195 1289   1.622\n",
      "  serious non-bonded atom clash:  1196 1289   1.533\n",
      "  serious non-bonded atom clash:  1197 1287   2.140\n",
      "  serious non-bonded atom clash:  1197 1288   1.553\n",
      "  serious non-bonded atom clash:  1197 1289   2.144\n",
      "  serious non-bonded atom clash:  1216 1262   2.299\n",
      "  serious non-bonded atom clash:  1218 1284   1.935\n",
      "  serious non-bonded atom clash:  1219 1262   2.021\n",
      "  serious non-bonded atom clash:  1219 1277   1.863\n",
      "  serious non-bonded atom clash:  1219 1278   2.125\n",
      "  serious non-bonded atom clash:  1219 1279   1.883\n",
      "  serious non-bonded atom clash:  1219 1280   2.163\n",
      "  serious non-bonded atom clash:  1262 1279   1.800\n",
      "  serious non-bonded atom clash:  1308 1346   1.922\n",
      "  serious non-bonded atom clash:  1309 1346   1.814\n",
      "  serious non-bonded atom clash:  1309 1347   1.856\n",
      "  serious non-bonded atom clash:  1310 1345   1.752\n",
      "  serious non-bonded atom clash:  1310 1346   1.549\n",
      "  serious non-bonded atom clash:  1310 1347   0.442\n",
      "  serious non-bonded atom clash:  1311 1345   1.843\n",
      "  serious non-bonded atom clash:  1311 1347   1.407\n",
      "  serious non-bonded atom clash:  1312 1340   2.114\n",
      "  serious non-bonded atom clash:  1312 1343   1.842\n",
      "  serious non-bonded atom clash:  1312 1344   1.964\n",
      "  serious non-bonded atom clash:  1312 1345   1.299\n",
      "  serious non-bonded atom clash:  1312 1346   1.407\n",
      "  serious non-bonded atom clash:  1312 1347   0.954\n",
      "  serious non-bonded atom clash:  1313 1344   1.825\n",
      "  serious non-bonded atom clash:  1313 1346   1.568\n",
      "  serious non-bonded atom clash:  1314 1344   1.208\n",
      "  serious non-bonded atom clash:  1314 1346   1.796\n",
      "  serious non-bonded atom clash:  1436 1474   1.768\n",
      "  serious non-bonded atom clash:  1436 1475   0.966\n",
      "  serious non-bonded atom clash:  1437 1475   2.218\n",
      "  serious non-bonded atom clash:  1438 1475   1.747\n",
      "  serious non-bonded atom clash:  1486 1489   2.181\n",
      "  serious non-bonded atom clash:  1486 1522   2.155\n",
      "  serious non-bonded atom clash:  1489 1522   2.077\n",
      "  serious non-bonded atom clash:  1490 1519   1.863\n",
      "  serious non-bonded atom clash:  1490 1520   2.124\n",
      "  serious non-bonded atom clash:  1491 1519   1.983\n",
      "  serious non-bonded atom clash:  1491 1520   1.662\n",
      "  serious non-bonded atom clash:  1491 1521   1.787\n",
      "  serious non-bonded atom clash:  1492 1494   2.068\n",
      "  serious non-bonded atom clash:  1564 1583   1.439\n",
      "  serious non-bonded atom clash:  1564 1585   2.034\n",
      "  serious non-bonded atom clash:  1575 1583   2.278\n",
      "  serious non-bonded atom clash:  1576 1583   2.243\n",
      "\n",
      "DISTANCE1:  0.00 2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40\n",
      "DISTANCE2:  2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40 3.50\n",
      "FREQUENCY:    74   21   14    7   24   32   40   72  119  150  160  205  219  209  275\n",
      "\n",
      "\n",
      "<< end of ENERGY.\n",
      "openf___224_> Open           Target.B99990001.pdb\n",
      "wrpdb___568_> Residues, atoms, selected atoms:      207     1613     1613\n",
      "openf___224_> Open           Target.ini\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      3936160    3843.906     3.754\n",
      "read_mo_297_> Segments, residues, atoms:        1      207     1613\n",
      "read_mo_298_> Segment:        1  1:A  207:A     1613\n",
      "randomi_498_> Atoms,selected atoms,random_seed,amplitude:     1613      457        1        4.0000\n",
      "randomi_496_> Amplitude is > 0; randomization is done.\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1927\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1933\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1938\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1956\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "iupac_m_486_> OE1/2 will be swapped:     -105.1969       13  13\n",
      "iupac_m_487_> NH1/2 swapped:      179.7925       26  26\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:      169.7476       28  28\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_485_> OD1/2 will be swapped:     -175.5649       34  34\n",
      "iupac_m_487_> NH1/2 swapped:      179.7537       63  63\n",
      "iupac_m_485_> OD1/2 will be swapped:     -178.8687       64  64\n",
      "iupac_m_487_> NH1/2 swapped:      178.0681       68  68\n",
      "iupac_m_487_> NH1/2 swapped:      179.9441       98  98\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      103  103\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0163      104  104\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      107  107\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      109  109\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      110  110\n",
      "iupac_m_485_> OD1/2 will be swapped:     -179.9802      111  111\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      114  114\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0292      124  124\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9657      127  127\n",
      "iupac_m_487_> NH1/2 swapped:      179.9558      129  129\n",
      "iupac_m_486_> OE1/2 will be swapped:      154.4214      135  135\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      136  136\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0239      138  138\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9558      139  139\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9720      140  140\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9407      146  146\n",
      "iupac_m_486_> OE1/2 will be swapped:      179.9604      152  152\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      160  160\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      164  164\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      165  165\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      179  179\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:     -176.2274      180  180\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -180.0000      183  183\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      196  196\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      200  200\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0309      202  202\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      207  207\n",
      "\n",
      "\n",
      ">> ENERGY; Differences between the model's features and restraints:\n",
      "Number of all residues in MODEL                   :      207\n",
      "Number of all, selected real atoms                :     1613     457\n",
      "Number of all, selected pseudo atoms              :        0       0\n",
      "Number of all static, selected restraints         :    12693    1983\n",
      "COVALENT_CYS                                      :        F\n",
      "NONBONDED_SEL_ATOMS                               :        1\n",
      "Number of non-bonded pairs (excluding 1-2,1-3,1-4):      685\n",
      "Dynamic pairs routine                             : 5, NATM_SEL x NATM double loop\n",
      "Atomic shift for contacts update (UPDATE_DYNAMIC) :    0.390\n",
      "LENNARD_JONES_SWITCH                              :    6.500   7.500\n",
      "COULOMB_JONES_SWITCH                              :    6.500   7.500\n",
      "RESIDUE_SPAN_RANGE                                :        0   99999\n",
      "NLOGN_USE                                         :       15\n",
      "CONTACT_SHELL                                     :    4.000\n",
      "DYNAMIC_PAIRS,_SPHERE,_COULOMB,_LENNARD,_MODELLER :        T       T       F       F       F\n",
      "SPHERE_STDV                                       :    0.050\n",
      "RADII_FACTOR                                      :    0.820\n",
      "Current energy                                    :         259.4150\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Summary of the restraint violations: \n",
      "\n",
      "   NUM     ... number of restraints.\n",
      "   NUMVI   ... number of restraints with RVIOL > VIOL_REPORT_CUT[i].\n",
      "   RVIOL   ... relative difference from the best value.\n",
      "   NUMVP   ... number of restraints with -Ln(pdf) > VIOL_REPORT_CUT2[i].\n",
      "   RMS_1   ... RMS(feature, minimally_violated_basis_restraint, NUMB).\n",
      "   RMS_2   ... RMS(feature, best_value, NUMB).\n",
      "   MOL.PDF ... scaled contribution to -Ln(Molecular pdf).\n",
      "\n",
      " #                     RESTRAINT_GROUP      NUM   NUMVI  NUMVP   RMS_1   RMS_2         MOL.PDF     S_i\n",
      "------------------------------------------------------------------------------------------------------\n",
      " 1 Bond length potential              :     478       0      0   0.002   0.002     0.41314       1.000\n",
      " 2 Bond angle potential               :     667       0      0   1.962   1.962      50.459       1.000\n",
      " 3 Stereochemical cosine torsion poten:     343       0      3  52.252  52.252      123.08       1.000\n",
      " 4 Stereochemical improper torsion pot:     181       0      0   0.739   0.739      2.2359       1.000\n",
      " 5 Soft-sphere overlap restraints     :     685       0      0   0.001   0.001     0.17762       1.000\n",
      " 6 Lennard-Jones 6-12 potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 7 Coulomb point-point electrostatic p:       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 8 H-bonding potential                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 9 Distance restraints 1 (CA-CA)      :      29       0      0   0.084   0.084     0.20427E-01   1.000\n",
      "10 Distance restraints 2 (N-O)        :      30       0      0   0.514   0.514     0.97203       1.000\n",
      "11 Mainchain Phi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "12 Mainchain Psi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "13 Mainchain Omega dihedral restraints:      70       0      0   2.090   2.090      3.6071       1.000\n",
      "14 Sidechain Chi_1 dihedral restraints:      43       0      0  84.201  84.201      12.604       1.000\n",
      "15 Sidechain Chi_2 dihedral restraints:      31       0      0  74.097  74.097      3.4641       1.000\n",
      "16 Sidechain Chi_3 dihedral restraints:      21       0      0  47.737  47.737      5.2560       1.000\n",
      "17 Sidechain Chi_4 dihedral restraints:       3       0      0  76.352  76.352      2.0580       1.000\n",
      "18 Disulfide distance restraints      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "19 Disulfide angle restraints         :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "20 Disulfide dihedral angle restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "21 Lower bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "22 Upper bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "23 Distance restraints 3 (SDCH-MNCH)  :      16       0      0   0.375   0.375     0.34580E-01   1.000\n",
      "24 Sidechain Chi_5 dihedral restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "25 Phi/Psi pair of dihedral restraints:      71      53      3  21.366 153.486      55.036       1.000\n",
      "26 Distance restraints 4 (SDCH-SDCH)  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "27 Distance restraints 5 (X-Y)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "28 NMR distance restraints 6 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "29 NMR distance restraints 7 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "30 Minimal distance restraints        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "31 Non-bonded restraints              :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "32 Atomic accessibility restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "33 Atomic density restraints          :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "34 Absolute position restraints       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "35 Dihedral angle difference restraint:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "36 GBSA implicit solvent potential    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "37 EM density fitting potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "38 SAXS restraints                    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "39 Symmetry restraints                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "\n",
      "\n",
      "\n",
      "openf___224_> Open           Target.V99990002\n",
      "# Heavy relative violation of each residue is written to: Target.V99990002\n",
      "# The profile is NOT normalized by the number of restraints.\n",
      "# The profiles are smoothed over a window of residues:    1\n",
      "# The sum of all numbers in the file:   13188.0283\n",
      "\n",
      "\n",
      "\n",
      "List of the violated restraints:\n",
      "   A restraint is violated when the relative difference\n",
      "   from the best value (RVIOL) is larger than CUTOFF.\n",
      "\n",
      "   ICSR   ... index of a restraint in the current set.\n",
      "   RESNO  ... residue numbers of the first two atoms.\n",
      "   ATM    ... IUPAC atom names of the first two atoms.\n",
      "   FEAT   ... the value of the feature in the model.\n",
      "   restr  ... the mean of the basis restraint with the smallest\n",
      "              difference from the model (local minimum).\n",
      "   viol   ... difference from the local minimum.\n",
      "   rviol  ... relative difference from the local minimum.\n",
      "   RESTR  ... the best value (global minimum).\n",
      "   VIOL   ... difference from the best value.\n",
      "   RVIOL  ... relative difference from the best value.\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 25                           : Phi/Psi pair of dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       6.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   5791   1M   2A C   N       7    9  -70.98  -68.20    4.20    0.26  -62.50  170.87   28.43\n",
      "    1          2A   2A N   CA      9   10  148.45  145.30                  -40.90\n",
      "    2   5792   2A   3H C   N      12   14  -67.07  -67.60    0.97    0.09  -63.20  176.92   22.36\n",
      "    2          3H   3H N   CA     14   15  140.82  140.00                  -42.30\n",
      "    3   5793   3H   4A C   N      22   24 -121.33 -134.00   12.70    0.46  -62.50 -179.00   32.80\n",
      "    3          4A   4A N   CA     24   25  147.93  147.00                  -40.90\n",
      "    4   5794   4A   5G C   N      27   29  174.53 -167.20   19.04    0.28   82.20 -165.24   13.61\n",
      "    4          5G   5G N   CA     29   30  179.98  174.60                    8.50\n",
      "    5   5795   5G   6R C   N      31   33 -115.74 -125.20   20.04    1.10  -63.00  169.08   26.48\n",
      "    5          6R   6R N   CA     33   34  158.26  140.60                  -41.10\n",
      "    6   5796   6R   7T C   N      42   44 -116.13 -124.80    8.71    0.48  -63.20 -178.46   26.80\n",
      "    6          7T   7T N   CA     44   45  144.25  143.50                  -42.10\n",
      "    7   5797   7T   8G C   N      49   51   89.56   78.70   29.11    0.45   82.20  158.56    8.07\n",
      "    7          8G   8G N   CA     51   52  166.89 -166.10                    8.50\n",
      "    8   5798   8G   9Y C   N      53   55 -137.41 -124.30   50.53    2.18  -63.50  151.63   27.62\n",
      "    8          9Y   9Y N   CA     55   56 -175.80  135.40                  -43.40\n",
      "    9   5799   9Y  10D C   N      65   67 -108.67  -63.30   51.14    8.73  -63.30   51.14    8.73\n",
      "    9         10D  10D N   CA     67   68  -63.60  -40.00                  -40.00\n",
      "   10   5821  31D  32A C   N     265  267  -73.48  -68.20    5.29    0.44  -62.50  173.84   29.04\n",
      "   10         32A  32A N   CA    267  268  145.61  145.30                  -40.90\n",
      "   11   5822  32A  33G C   N     270  272 -114.07  -80.20   70.69    4.40   82.20 -166.27   15.14\n",
      "   11         33G  33G N   CA    272  273  112.05  174.10                    8.50\n",
      "   12   5823  33G  34D C   N     274  276  -94.69  -96.50    4.33    0.18  -63.30  161.22   18.22\n",
      "   12         34D  34D N   CA    276  277  118.14  114.20                  -40.00\n",
      "   13   5825  35V  36G C   N     289  291  156.30 -167.20   39.02    1.11   82.20  169.38   11.61\n",
      "   13         36G  36G N   CA    291  292  160.81  174.60                    8.50\n",
      "   14   5826  36G  37A C   N     293  295 -131.99 -134.00    2.79    0.16  -62.50 -176.19   33.84\n",
      "   14         37A  37A N   CA    295  296  148.93  147.00                  -40.90\n",
      "   15   5827  37A  38A C   N     298  300 -116.64 -134.00   18.02    0.44  -62.50 -174.96   33.20\n",
      "   15         38A  38A N   CA    300  301  142.16  147.00                  -40.90\n",
      "   16   5830  40P  41G C   N     317  319 -157.84 -167.20   21.08    0.65   82.20 -170.10   14.23\n",
      "   16         41G  41G N   CA    319  320  155.72  174.60                    8.50\n",
      "   17   5831  41G  42A C   N     321  323 -141.39 -134.00   13.34    0.50  -62.50  179.29   33.61\n",
      "   17         42A  42A N   CA    323  324  158.10  147.00                  -40.90\n",
      "   18   5832  42A  43A C   N     326  328  -67.01  -68.20    2.34    0.15  -62.50  175.87   29.04\n",
      "   18         43A  43A N   CA    328  329  143.29  145.30                  -40.90\n",
      "   19   5833  43A  44P C   N     331  333  -69.22  -58.70   13.91    1.05  -64.50  168.66   12.84\n",
      "   19         44P  44P N   CA    333  334  -21.40  -30.50                  147.20\n",
      "   20   5834  44P  45A C   N     338  340  -69.35  -68.20    2.26    0.22  -62.50  175.88   29.16\n",
      "   20         45A  45A N   CA    340  341  143.36  145.30                  -40.90\n",
      "   21   5836  46P  47G C   N     350  352  -72.77  -80.20   19.76    0.47   82.20 -146.20    6.88\n",
      "   21         47G  47G N   CA    352  353  155.79  174.10                    8.50\n",
      "   22   5838  48I  49F C   N     362  364 -113.61 -124.20   11.05    0.34  -63.20 -177.34   28.73\n",
      "   22         49F  49F N   CA    364  365  140.13  143.30                  -44.30\n",
      "   23   5839  49F  50S C   N     373  375 -124.94 -136.60   11.71    0.44  -64.10 -174.77   17.69\n",
      "   23         50S  50S N   CA    375  376  150.04  151.20                  -35.00\n",
      "   24   5840  50S  51S C   N     379  381 -128.67 -136.60    7.97    0.30  -64.10 -173.91   18.03\n",
      "   24         51S  51S N   CA    381  382  150.47  151.20                  -35.00\n",
      "   25   5841  51S  52Q C   N     385  387 -111.91 -121.10   11.65    0.36  -63.80  179.41   24.13\n",
      "   25         52Q  52Q N   CA    387  388  132.54  139.70                  -40.30\n",
      "   26   5843  53P  54G C   N     401  403 -166.54 -167.20   24.61    0.96   82.20  179.99   13.42\n",
      "   26         54G  54G N   CA    403  404  149.99  174.60                    8.50\n",
      "   27   5844  54G  55H C   N     405  407 -135.81 -125.60   20.27    0.56  -63.20  176.98   26.86\n",
      "   27         55H  55H N   CA    407  408  156.30  138.80                  -42.30\n",
      "   28   5845  55H  56T C   N     415  417 -114.63 -124.80   13.45    0.45  -63.20 -175.88   20.65\n",
      "   28         56T  56T N   CA    417  418  134.69  143.50                  -42.10\n",
      "   29   5847  57P  58H C   N     429  431 -103.49 -125.60   31.99    0.80  -63.20  163.03   18.23\n",
      "   29         58H  58H N   CA    431  432  115.68  138.80                  -42.30\n",
      "   30   5849  59P  60A C   N     446  448  -77.24  -68.20   10.13    0.70  -62.50  169.87   28.58\n",
      "   30         60A  60A N   CA    448  449  149.87  145.30                  -40.90\n",
      "   31   5850  60A  61A C   N     451  453 -151.65 -134.00   20.27    0.47  -62.50 -174.98   35.08\n",
      "   31         61A  61A N   CA    453  454  156.98  147.00                  -40.90\n",
      "   32   5851  61A  62S C   N     456  458  -79.31  -72.40   18.79    0.90  -64.10  155.86   12.26\n",
      "   32         62S  62S N   CA    458  459  169.88  152.40                  -35.00\n",
      "   33   5852  62S  63R C   N     462  464  -73.87  -72.10   22.21    1.75  -63.00  161.22   21.07\n",
      "   33         63R  63R N   CA    464  465  119.76  141.90                  -41.10\n",
      "   34   5853  63R  64D C   N     473  475 -122.49  -96.50   35.36    1.48  -63.30 -172.25   20.37\n",
      "   34         64D  64D N   CA    475  476  138.17  114.20                  -40.00\n",
      "   35   5854  64D  65P C   N     481  483  -64.52  -58.70    8.29    0.58  -64.50  171.79   12.82\n",
      "   35         65P  65P N   CA    483  484  -24.59  -30.50                  147.20\n",
      "   36   5856  66V  67A C   N     495  497 -139.78 -134.00    7.01    0.17  -62.50 -174.94   34.47\n",
      "   36         67A  67A N   CA    497  498  150.95  147.00                  -40.90\n",
      "   37   5857  67A  68R C   N     500  502 -142.94 -125.20   25.51    0.78  -63.00  178.83   29.59\n",
      "   37         68R  68R N   CA    502  503  158.93  140.60                  -41.10\n",
      "   38   5858  68R  69T C   N     511  513 -127.37 -124.80   16.26    0.62  -63.20  170.86   26.14\n",
      "   38         69T  69T N   CA    513  514  159.55  143.50                  -42.10\n",
      "   39   5859  69T  70S C   N     518  520  -78.90  -72.40   11.21    0.86  -64.10  178.88   12.16\n",
      "   39         70S  70S N   CA    520  521  143.26  152.40                  -35.00\n",
      "   40   5861  71P  72L C   N     531  533 -121.87 -108.50   13.62    0.68  -63.50 -179.21   22.52\n",
      "   40         72L  72L N   CA    533  534  129.90  132.50                  -41.20\n",
      "   41   5862  72L  73Q C   N     539  541  -94.22  -73.00   21.49    1.59  -63.80 -179.80   24.99\n",
      "   41         73Q  73Q N   CA    541  542  137.32  140.70                  -40.30\n",
      "   42   5863  73Q  74T C   N     548  550  -84.76  -78.10    9.04    0.59  -63.20  175.54   23.95\n",
      "   42         74T  74T N   CA    550  551  143.68  149.80                  -42.10\n",
      "   43   5865  75P  76A C   N     562  564 -128.55 -134.00   15.40    0.71  -62.50 -174.36   28.04\n",
      "   43         76A  76A N   CA    564  565  132.59  147.00                  -40.90\n",
      "   44   5866  76A  77A C   N     567  569 -134.57 -134.00    8.09    0.47  -62.50 -166.27   29.20\n",
      "   44         77A  77A N   CA    569  570  138.93  147.00                  -40.90\n",
      "   45   5868  78P  79G C   N     579  581 -118.74  -80.20   52.65    3.75   82.20 -154.75   15.96\n",
      "   45         79G  79G N   CA    581  582  138.23  174.10                    8.50\n",
      "   46   5869  79G  80A C   N     583  585 -135.48 -134.00    5.46    0.27  -62.50 -177.89   33.76\n",
      "   46         80A  80A N   CA    585  586  152.25  147.00                  -40.90\n",
      "   47   5870  80A  81A C   N     588  590 -122.51 -134.00   14.18    0.36  -62.50 -170.65   28.75\n",
      "   47         81A  81A N   CA    590  591  138.69  147.00                  -40.90\n",
      "   48   5871  81A  82A C   N     593  595  -87.34  -68.20   23.46    1.53  -62.50  162.14   27.86\n",
      "   48         82A  82A N   CA    595  596  158.87  145.30                  -40.90\n",
      "   49   5872  82A  83G C   N     598  600  158.48 -167.20   34.93    0.54   82.20 -171.28   12.69\n",
      "   49         83G  83G N   CA    600  601 -178.88  174.60                    8.50\n",
      "   50   5874  84P  85A C   N     609  611  -94.82  -68.20   29.19    2.80  -62.50  177.20   27.63\n",
      "   50         85A  85A N   CA    611  612  133.32  145.30                  -40.90\n",
      "   51   5875  85A  86L C   N     614  616  -80.38  -70.70   17.66    1.69  -63.50  168.89   22.54\n",
      "   51         86L  86L N   CA    616  617  126.84  141.60                  -41.20\n",
      "   52   5876  86L  87S C   N     622  624 -124.47 -136.60   12.58    0.42  -64.10 -172.85   17.79\n",
      "   52         87S  87S N   CA    624  625  147.86  151.20                  -35.00\n",
      "   53   5877  87S  88P C   N     628  630  -62.11  -58.70    4.43    0.34  -64.50  174.88   12.92\n",
      "   53         88P  88P N   CA    630  631  -27.66  -30.50                  147.20\n",
      "\n",
      "\n",
      "report______> Distribution of short non-bonded contacts:\n",
      "\n",
      "  serious non-bonded atom clash:    90 1354   1.714\n",
      "  serious non-bonded atom clash:   122 1320   2.215\n",
      "  serious non-bonded atom clash:   144  727   2.264\n",
      "  serious non-bonded atom clash:   151 1196   1.945\n",
      "  serious non-bonded atom clash:   190 1197   1.758\n",
      "  serious non-bonded atom clash:   664  694   2.295\n",
      "  serious non-bonded atom clash:   671  702   1.565\n",
      "  serious non-bonded atom clash:   719 1120   2.189\n",
      "  serious non-bonded atom clash:   720 1119   1.741\n",
      "  serious non-bonded atom clash:   721 1119   1.648\n",
      "  serious non-bonded atom clash:   726 1152   1.768\n",
      "  serious non-bonded atom clash:   726 1154   2.269\n",
      "  serious non-bonded atom clash:   727 1152   2.294\n",
      "  serious non-bonded atom clash:   728 1152   2.162\n",
      "  serious non-bonded atom clash:   728 1154   1.683\n",
      "  serious non-bonded atom clash:   729 1152   2.041\n",
      "  serious non-bonded atom clash:   729 1153   2.155\n",
      "  serious non-bonded atom clash:   729 1154   1.124\n",
      "  serious non-bonded atom clash:   757  797   2.239\n",
      "  serious non-bonded atom clash:   757  798   2.053\n",
      "  serious non-bonded atom clash:   809  817   2.094\n",
      "  serious non-bonded atom clash:   823  827   2.273\n",
      "  serious non-bonded atom clash:   882 1210   2.153\n",
      "  serious non-bonded atom clash:  1000 1371   1.984\n",
      "  serious non-bonded atom clash:  1019 1041   1.587\n",
      "  serious non-bonded atom clash:  1019 1043   1.762\n",
      "  serious non-bonded atom clash:  1063 1067   2.231\n",
      "  serious non-bonded atom clash:  1111 1455   1.857\n",
      "  serious non-bonded atom clash:  1111 1457   1.212\n",
      "  serious non-bonded atom clash:  1111 1459   1.921\n",
      "  serious non-bonded atom clash:  1139 1383   2.285\n",
      "  serious non-bonded atom clash:  1141 1382   2.235\n",
      "  serious non-bonded atom clash:  1141 1383   2.023\n",
      "  serious non-bonded atom clash:  1168 1342   1.700\n",
      "  serious non-bonded atom clash:  1168 1343   2.160\n",
      "  serious non-bonded atom clash:  1169 1340   1.956\n",
      "  serious non-bonded atom clash:  1169 1342   2.045\n",
      "  serious non-bonded atom clash:  1169 1343   1.522\n",
      "  serious non-bonded atom clash:  1169 1345   2.168\n",
      "  serious non-bonded atom clash:  1170 1312   1.597\n",
      "  serious non-bonded atom clash:  1170 1339   2.166\n",
      "  serious non-bonded atom clash:  1170 1340   0.933\n",
      "  serious non-bonded atom clash:  1170 1342   2.001\n",
      "  serious non-bonded atom clash:  1170 1343   0.763\n",
      "  serious non-bonded atom clash:  1170 1344   1.750\n",
      "  serious non-bonded atom clash:  1170 1345   1.567\n",
      "  serious non-bonded atom clash:  1170 1346   2.186\n",
      "  serious non-bonded atom clash:  1170 1347   2.117\n",
      "  serious non-bonded atom clash:  1172 1310   2.198\n",
      "  serious non-bonded atom clash:  1172 1312   1.497\n",
      "  serious non-bonded atom clash:  1172 1313   2.013\n",
      "  serious non-bonded atom clash:  1172 1314   1.631\n",
      "  serious non-bonded atom clash:  1172 1340   1.411\n",
      "  serious non-bonded atom clash:  1172 1343   2.110\n",
      "  serious non-bonded atom clash:  1172 1344   0.608\n",
      "  serious non-bonded atom clash:  1172 1346   1.043\n",
      "  serious non-bonded atom clash:  1172 1347   1.878\n",
      "  serious non-bonded atom clash:  1174 1313   2.173\n",
      "  serious non-bonded atom clash:  1174 1315   2.177\n",
      "  serious non-bonded atom clash:  1174 1344   1.525\n",
      "  serious non-bonded atom clash:  1174 1346   1.595\n",
      "  serious non-bonded atom clash:  1195 1289   1.622\n",
      "  serious non-bonded atom clash:  1196 1289   1.533\n",
      "  serious non-bonded atom clash:  1197 1287   2.140\n",
      "  serious non-bonded atom clash:  1197 1288   1.553\n",
      "  serious non-bonded atom clash:  1197 1289   2.144\n",
      "  serious non-bonded atom clash:  1216 1262   2.299\n",
      "  serious non-bonded atom clash:  1218 1284   1.935\n",
      "  serious non-bonded atom clash:  1219 1262   2.021\n",
      "  serious non-bonded atom clash:  1219 1277   1.863\n",
      "  serious non-bonded atom clash:  1219 1278   2.125\n",
      "  serious non-bonded atom clash:  1219 1279   1.883\n",
      "  serious non-bonded atom clash:  1219 1280   2.163\n",
      "  serious non-bonded atom clash:  1262 1279   1.800\n",
      "  serious non-bonded atom clash:  1308 1346   1.922\n",
      "  serious non-bonded atom clash:  1309 1346   1.814\n",
      "  serious non-bonded atom clash:  1309 1347   1.856\n",
      "  serious non-bonded atom clash:  1310 1345   1.752\n",
      "  serious non-bonded atom clash:  1310 1346   1.549\n",
      "  serious non-bonded atom clash:  1310 1347   0.442\n",
      "  serious non-bonded atom clash:  1311 1345   1.843\n",
      "  serious non-bonded atom clash:  1311 1347   1.407\n",
      "  serious non-bonded atom clash:  1312 1340   2.114\n",
      "  serious non-bonded atom clash:  1312 1343   1.842\n",
      "  serious non-bonded atom clash:  1312 1344   1.964\n",
      "  serious non-bonded atom clash:  1312 1345   1.299\n",
      "  serious non-bonded atom clash:  1312 1346   1.407\n",
      "  serious non-bonded atom clash:  1312 1347   0.954\n",
      "  serious non-bonded atom clash:  1313 1344   1.825\n",
      "  serious non-bonded atom clash:  1313 1346   1.568\n",
      "  serious non-bonded atom clash:  1314 1344   1.208\n",
      "  serious non-bonded atom clash:  1314 1346   1.796\n",
      "  serious non-bonded atom clash:  1436 1474   1.768\n",
      "  serious non-bonded atom clash:  1436 1475   0.966\n",
      "  serious non-bonded atom clash:  1437 1475   2.218\n",
      "  serious non-bonded atom clash:  1438 1475   1.747\n",
      "  serious non-bonded atom clash:  1486 1489   2.181\n",
      "  serious non-bonded atom clash:  1486 1522   2.155\n",
      "  serious non-bonded atom clash:  1489 1522   2.077\n",
      "  serious non-bonded atom clash:  1490 1519   1.863\n",
      "  serious non-bonded atom clash:  1490 1520   2.124\n",
      "  serious non-bonded atom clash:  1491 1519   1.983\n",
      "  serious non-bonded atom clash:  1491 1520   1.662\n",
      "  serious non-bonded atom clash:  1491 1521   1.787\n",
      "  serious non-bonded atom clash:  1492 1494   2.068\n",
      "  serious non-bonded atom clash:  1564 1583   1.439\n",
      "  serious non-bonded atom clash:  1564 1585   2.034\n",
      "  serious non-bonded atom clash:  1575 1583   2.278\n",
      "  serious non-bonded atom clash:  1576 1583   2.243\n",
      "\n",
      "DISTANCE1:  0.00 2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40\n",
      "DISTANCE2:  2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40 3.50\n",
      "FREQUENCY:    74   21   14    7   24   33   39   78  116  147  164  207  210  226  282\n",
      "\n",
      "\n",
      "<< end of ENERGY.\n",
      "openf___224_> Open           Target.B99990002.pdb\n",
      "wrpdb___568_> Residues, atoms, selected atoms:      207     1613     1613\n",
      "openf___224_> Open           Target.ini\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      3936160    3843.906     3.754\n",
      "read_mo_297_> Segments, residues, atoms:        1      207     1613\n",
      "read_mo_298_> Segment:        1  1:A  207:A     1613\n",
      "randomi_498_> Atoms,selected atoms,random_seed,amplitude:     1613      457        1        4.0000\n",
      "randomi_496_> Amplitude is > 0; randomization is done.\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1927\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1933\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1938\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1956\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "iupac_m_487_> NH1/2 swapped:      179.0202        6  6\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:      140.7287        9  9\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_486_> OE1/2 will be swapped:     -105.1969       13  13\n",
      "iupac_m_487_> NH1/2 swapped:      179.7925       26  26\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:      169.7476       28  28\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_485_> OD1/2 will be swapped:      150.9763       34  34\n",
      "iupac_m_487_> NH1/2 swapped:      179.9441       98  98\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      103  103\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0163      104  104\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      107  107\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      109  109\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      110  110\n",
      "iupac_m_485_> OD1/2 will be swapped:     -179.9802      111  111\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      114  114\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0292      124  124\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9657      127  127\n",
      "iupac_m_487_> NH1/2 swapped:      179.9558      129  129\n",
      "iupac_m_486_> OE1/2 will be swapped:      154.4214      135  135\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      136  136\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0239      138  138\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9558      139  139\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9720      140  140\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9407      146  146\n",
      "iupac_m_486_> OE1/2 will be swapped:      179.9604      152  152\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      160  160\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      164  164\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      165  165\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      179  179\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:     -176.2274      180  180\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -180.0000      183  183\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      196  196\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      200  200\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0309      202  202\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      207  207\n",
      "\n",
      "\n",
      ">> ENERGY; Differences between the model's features and restraints:\n",
      "Number of all residues in MODEL                   :      207\n",
      "Number of all, selected real atoms                :     1613     457\n",
      "Number of all, selected pseudo atoms              :        0       0\n",
      "Number of all static, selected restraints         :    12693    1983\n",
      "COVALENT_CYS                                      :        F\n",
      "NONBONDED_SEL_ATOMS                               :        1\n",
      "Number of non-bonded pairs (excluding 1-2,1-3,1-4):      530\n",
      "Dynamic pairs routine                             : 5, NATM_SEL x NATM double loop\n",
      "Atomic shift for contacts update (UPDATE_DYNAMIC) :    0.390\n",
      "LENNARD_JONES_SWITCH                              :    6.500   7.500\n",
      "COULOMB_JONES_SWITCH                              :    6.500   7.500\n",
      "RESIDUE_SPAN_RANGE                                :        0   99999\n",
      "NLOGN_USE                                         :       15\n",
      "CONTACT_SHELL                                     :    4.000\n",
      "DYNAMIC_PAIRS,_SPHERE,_COULOMB,_LENNARD,_MODELLER :        T       T       F       F       F\n",
      "SPHERE_STDV                                       :    0.050\n",
      "RADII_FACTOR                                      :    0.820\n",
      "Current energy                                    :         317.3790\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Summary of the restraint violations: \n",
      "\n",
      "   NUM     ... number of restraints.\n",
      "   NUMVI   ... number of restraints with RVIOL > VIOL_REPORT_CUT[i].\n",
      "   RVIOL   ... relative difference from the best value.\n",
      "   NUMVP   ... number of restraints with -Ln(pdf) > VIOL_REPORT_CUT2[i].\n",
      "   RMS_1   ... RMS(feature, minimally_violated_basis_restraint, NUMB).\n",
      "   RMS_2   ... RMS(feature, best_value, NUMB).\n",
      "   MOL.PDF ... scaled contribution to -Ln(Molecular pdf).\n",
      "\n",
      " #                     RESTRAINT_GROUP      NUM   NUMVI  NUMVP   RMS_1   RMS_2         MOL.PDF     S_i\n",
      "------------------------------------------------------------------------------------------------------\n",
      " 1 Bond length potential              :     478       0      0   0.003   0.003      1.1327       1.000\n",
      " 2 Bond angle potential               :     667       0      1   2.258   2.258      66.829       1.000\n",
      " 3 Stereochemical cosine torsion poten:     343       0      5  52.698  52.698      124.69       1.000\n",
      " 4 Stereochemical improper torsion pot:     181       0      1   1.296   1.296      6.3277       1.000\n",
      " 5 Soft-sphere overlap restraints     :     530       0      0   0.003   0.003     0.47964       1.000\n",
      " 6 Lennard-Jones 6-12 potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 7 Coulomb point-point electrostatic p:       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 8 H-bonding potential                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 9 Distance restraints 1 (CA-CA)      :      29       0      0   0.038   0.038     0.43559E-02   1.000\n",
      "10 Distance restraints 2 (N-O)        :      30       0      0   0.483   0.483     0.88253       1.000\n",
      "11 Mainchain Phi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "12 Mainchain Psi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "13 Mainchain Omega dihedral restraints:      70       1      1   4.391   4.391      15.919       1.000\n",
      "14 Sidechain Chi_1 dihedral restraints:      43       0      1  81.440  81.440      11.587       1.000\n",
      "15 Sidechain Chi_2 dihedral restraints:      31       0      1  75.625  75.625      7.4270       1.000\n",
      "16 Sidechain Chi_3 dihedral restraints:      21       0      0  61.053  61.053      4.3956       1.000\n",
      "17 Sidechain Chi_4 dihedral restraints:       3       0      0  73.364  73.364      1.9631       1.000\n",
      "18 Disulfide distance restraints      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "19 Disulfide angle restraints         :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "20 Disulfide dihedral angle restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "21 Lower bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "22 Upper bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "23 Distance restraints 3 (SDCH-MNCH)  :      16       0      0   0.343   0.343     0.29661E-01   1.000\n",
      "24 Sidechain Chi_5 dihedral restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "25 Phi/Psi pair of dihedral restraints:      71      50     10  31.186 147.490      75.709       1.000\n",
      "26 Distance restraints 4 (SDCH-SDCH)  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "27 Distance restraints 5 (X-Y)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "28 NMR distance restraints 6 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "29 NMR distance restraints 7 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "30 Minimal distance restraints        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "31 Non-bonded restraints              :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "32 Atomic accessibility restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "33 Atomic density restraints          :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "34 Absolute position restraints       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "35 Dihedral angle difference restraint:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "36 GBSA implicit solvent potential    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "37 EM density fitting potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "38 SAXS restraints                    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "39 Symmetry restraints                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "\n",
      "\n",
      "\n",
      "openf___224_> Open           Target.V99990003\n",
      "# Heavy relative violation of each residue is written to: Target.V99990003\n",
      "# The profile is NOT normalized by the number of restraints.\n",
      "# The profiles are smoothed over a window of residues:    1\n",
      "# The sum of all numbers in the file:   13086.8291\n",
      "\n",
      "\n",
      "\n",
      "List of the violated restraints:\n",
      "   A restraint is violated when the relative difference\n",
      "   from the best value (RVIOL) is larger than CUTOFF.\n",
      "\n",
      "   ICSR   ... index of a restraint in the current set.\n",
      "   RESNO  ... residue numbers of the first two atoms.\n",
      "   ATM    ... IUPAC atom names of the first two atoms.\n",
      "   FEAT   ... the value of the feature in the model.\n",
      "   restr  ... the mean of the basis restraint with the smallest\n",
      "              difference from the model (local minimum).\n",
      "   viol   ... difference from the local minimum.\n",
      "   rviol  ... relative difference from the local minimum.\n",
      "   RESTR  ... the best value (global minimum).\n",
      "   VIOL   ... difference from the best value.\n",
      "   RVIOL  ... relative difference from the best value.\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 13                           : Mainchain Omega dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       4.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   6004   9Y   9Y CA  C      56   65 -156.48 -180.00   23.52    4.70 -180.00   23.52    4.70\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 25                           : Phi/Psi pair of dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       6.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   5791   1M   2A C   N       7    9  -64.74  -68.20    5.18    0.33  -62.50  177.68   29.23\n",
      "    1          2A   2A N   CA      9   10  141.44  145.30                  -40.90\n",
      "    2   5792   2A   3H C   N      12   14  -78.29  -67.60   11.49    0.74  -63.20  174.15   22.73\n",
      "    2          3H   3H N   CA     14   15  144.20  140.00                  -42.30\n",
      "    3   5793   3H   4A C   N      22   24 -121.84 -134.00   12.72    0.57  -62.50  178.52   32.43\n",
      "    3          4A   4A N   CA     24   25  150.73  147.00                  -40.90\n",
      "    4   5794   4A   5G C   N      27   29  107.88   78.70   72.14    1.09   82.20  122.15    7.18\n",
      "    4          5G   5G N   CA     29   30  127.92 -166.10                    8.50\n",
      "    5   5795   5G   6R C   N      31   33 -147.43 -125.20   62.70    2.47  -63.00  146.46   25.34\n",
      "    5          6R   6R N   CA     33   34 -160.77  140.60                  -41.10\n",
      "    6   5798   8G   9Y C   N      53   55 -120.86 -124.30   48.17    2.70  -63.50  144.97   25.85\n",
      "    6          9Y   9Y N   CA     55   56 -176.55  135.40                  -43.40\n",
      "    7   5799   9Y  10D C   N      65   67 -116.08  -63.30   57.82    9.86  -63.30   57.82    9.86\n",
      "    7         10D  10D N   CA     67   68  -63.60  -40.00                  -40.00\n",
      "    8   5821  31D  32A C   N     265  267 -164.18 -134.00   30.59    0.84  -62.50 -164.37   37.47\n",
      "    8         32A  32A N   CA    267  268  151.98  147.00                  -40.90\n",
      "    9   5822  32A  33G C   N     270  272   99.18   78.70   67.85    1.16   82.20  121.90    6.76\n",
      "    9         33G  33G N   CA    272  273  129.21 -166.10                    8.50\n",
      "   10   5823  33G  34D C   N     274  276 -135.43  -96.50   57.61    2.41  -63.30  178.56   26.70\n",
      "   10         34D  34D N   CA    276  277  156.67  114.20                  -40.00\n",
      "   11   5825  35V  36G C   N     289  291 -132.76 -167.20   41.69    0.74   82.20 -156.61   15.62\n",
      "   11         36G  36G N   CA    291  292  151.10  174.60                    8.50\n",
      "   12   5826  36G  37A C   N     293  295  -77.97  -68.20   14.81    0.93  -62.50  163.40   27.56\n",
      "   12         37A  37A N   CA    295  296  156.43  145.30                  -40.90\n",
      "   13   5827  37A  38A C   N     298  300  -66.87  -68.20    1.75    0.17  -62.50  172.72   28.52\n",
      "   13         38A  38A N   CA    300  301  146.43  145.30                  -40.90\n",
      "   14   5830  40P  41G C   N     317  319 -172.86 -167.20   18.92    0.78   82.20 -178.54   13.32\n",
      "   14         41G  41G N   CA    319  320  156.54  174.60                    8.50\n",
      "   15   5831  41G  42A C   N     321  323 -128.00 -134.00    6.49    0.15  -62.50 -173.54   34.06\n",
      "   15         42A  42A N   CA    323  324  144.52  147.00                  -40.90\n",
      "   16   5832  42A  43A C   N     326  328 -142.67 -134.00   11.83    0.34  -62.50 -177.40   34.22\n",
      "   16         43A  43A N   CA    328  329  155.05  147.00                  -40.90\n",
      "   17   5833  43A  44P C   N     331  333  -65.27  -58.70    7.10    0.72  -64.50  175.01   13.10\n",
      "   17         44P  44P N   CA    333  334  -27.81  -30.50                  147.20\n",
      "   18   5834  44P  45A C   N     338  340   64.15  -68.20  132.62   11.52  -62.50 -151.71   31.49\n",
      "   18         45A  45A N   CA    340  341  153.74  145.30                  -40.90\n",
      "   19   5836  46P  47G C   N     350  352 -160.49 -167.20   30.87    1.11   82.20  179.58   13.56\n",
      "   19         47G  47G N   CA    352  353  144.47  174.60                    8.50\n",
      "   20   5838  48I  49F C   N     362  364 -107.34 -124.20   20.21    0.53  -63.20 -178.11   22.49\n",
      "   20         49F  49F N   CA    364  365  132.16  143.30                  -44.30\n",
      "   21   5839  49F  50S C   N     373  375 -136.24 -136.60    5.31    0.28  -64.10 -176.71   18.41\n",
      "   21         50S  50S N   CA    375  376  156.49  151.20                  -35.00\n",
      "   22   5840  50S  51S C   N     379  381 -126.03 -136.60   10.57    0.43  -64.10 -175.79   17.70\n",
      "   22         51S  51S N   CA    381  382  151.51  151.20                  -35.00\n",
      "   23   5841  51S  52Q C   N     385  387 -108.97 -121.10   13.62    0.42  -63.80  179.59   24.27\n",
      "   23         52Q  52Q N   CA    387  388  133.52  139.70                  -40.30\n",
      "   24   5843  53P  54G C   N     401  403 -165.45 -167.20   18.13    0.69   82.20 -174.15   13.79\n",
      "   24         54G  54G N   CA    403  404  156.55  174.60                    8.50\n",
      "   25   5844  54G  55H C   N     405  407 -146.85 -125.60   28.06    0.69  -63.20 -178.93   28.06\n",
      "   25         55H  55H N   CA    407  408  157.12  138.80                  -42.30\n",
      "   26   5845  55H  56T C   N     415  417 -106.26 -124.80   23.24    0.81  -63.20  176.92   20.14\n",
      "   26         56T  56T N   CA    417  418  129.50  143.50                  -42.10\n",
      "   27   5846  56T  57P C   N     422  424  -52.53  -58.70    9.55    0.62  -64.50  175.42   13.73\n",
      "   27         57P  57P N   CA    424  425  -37.79  -30.50                  147.20\n",
      "   28   5847  57P  58H C   N     429  431   51.65   56.30   11.00    0.80  -63.20  147.82   25.16\n",
      "   28         58H  58H N   CA    431  432   50.77   40.80                  -42.30\n",
      "   29   5849  59P  60A C   N     446  448 -109.81 -134.00   25.28    0.61  -62.50 -174.41   32.92\n",
      "   29         60A  60A N   CA    448  449  139.65  147.00                  -40.90\n",
      "   30   5850  60A  61A C   N     451  453 -142.43 -134.00   14.79    0.55  -62.50  178.81   33.58\n",
      "   30         61A  61A N   CA    453  454  159.15  147.00                  -40.90\n",
      "   31   5851  61A  62S C   N     456  458  -68.28  -72.40    5.70    0.27  -64.10  176.58   13.05\n",
      "   31         62S  62S N   CA    458  459  148.46  152.40                  -35.00\n",
      "   32   5852  62S  63R C   N     462  464 -117.97 -125.20    7.25    0.28  -63.00 -172.88   29.08\n",
      "   32         63R  63R N   CA    464  465  140.04  140.60                  -41.10\n",
      "   33   5853  63R  64D C   N     473  475  -82.72  -70.90   27.55    0.93  -63.30  146.12   19.26\n",
      "   33         64D  64D N   CA    475  476  175.18  150.30                  -40.00\n",
      "   34   5854  64D  65P C   N     481  483  -66.24  -58.70   10.24    0.75  -64.50  170.78   12.84\n",
      "   34         65P  65P N   CA    483  484  -23.58  -30.50                  147.20\n",
      "   35   5856  66V  67A C   N     495  497 -107.89 -134.00   28.55    0.65  -62.50 -177.90   27.99\n",
      "   35         67A  67A N   CA    497  498  135.45  147.00                  -40.90\n",
      "   36   5857  67A  68R C   N     500  502 -135.45 -125.20   18.61    0.63  -63.00  178.16   29.02\n",
      "   36         68R  68R N   CA    502  503  156.13  140.60                  -41.10\n",
      "   37   5858  68R  69T C   N     511  513  -76.76  -78.10    4.25    0.23  -63.20  164.62   22.01\n",
      "   37         69T  69T N   CA    513  514  153.84  149.80                  -42.10\n",
      "   38   5859  69T  70S C   N     518  520  -81.04  -72.40   13.42    0.62  -64.10  163.21   12.91\n",
      "   38         70S  70S N   CA    520  521  162.67  152.40                  -35.00\n",
      "   39   5861  71P  72L C   N     531  533 -107.98 -108.50    3.76    0.21  -63.50 -177.08   23.25\n",
      "   39         72L  72L N   CA    533  534  136.23  132.50                  -41.20\n",
      "   40   5862  72L  73Q C   N     539  541  -84.60  -73.00   17.57    1.01  -63.80  167.10   25.68\n",
      "   40         73Q  73Q N   CA    541  542  153.90  140.70                  -40.30\n",
      "   41   5863  73Q  74T C   N     548  550 -106.24 -124.80   32.72    1.01  -63.20  164.39   18.57\n",
      "   41         74T  74T N   CA    550  551  116.55  143.50                  -42.10\n",
      "   42   5865  75P  76A C   N     562  564 -127.32 -134.00    6.88    0.29  -62.50 -177.64   33.35\n",
      "   42         76A  76A N   CA    564  565  148.65  147.00                  -40.90\n",
      "   43   5866  76A  77A C   N     567  569 -116.01 -134.00   19.36    0.44  -62.50 -172.93   33.50\n",
      "   43         77A  77A N   CA    569  570  139.84  147.00                  -40.90\n",
      "   44   5868  78P  79G C   N     579  581 -147.60 -167.20   39.57    1.16   82.20 -174.79   14.19\n",
      "   44         79G  79G N   CA    581  582  140.22  174.60                    8.50\n",
      "   45   5869  79G  80A C   N     583  585 -151.02 -134.00   20.43    0.50  -62.50 -176.43   34.81\n",
      "   45         80A  80A N   CA    585  586  158.29  147.00                  -40.90\n",
      "   46   5870  80A  81A C   N     588  590 -110.24 -134.00   26.93    0.62  -62.50 -178.38   27.84\n",
      "   46         81A  81A N   CA    590  591  134.33  147.00                  -40.90\n",
      "   47   5871  81A  82A C   N     593  595 -114.56 -134.00   24.18    0.62  -62.50 -178.84   27.64\n",
      "   47         82A  82A N   CA    595  596  132.62  147.00                  -40.90\n",
      "   48   5874  84P  85A C   N     609  611 -119.24 -134.00   24.09    0.84  -62.50  178.13   27.04\n",
      "   48         85A  85A N   CA    611  612  127.95  147.00                  -40.90\n",
      "   49   5875  85A  86L C   N     614  616 -149.11 -108.50   42.71    1.91  -63.50 -166.91   32.10\n",
      "   49         86L  86L N   CA    616  617  145.72  132.50                  -41.20\n",
      "   50   5876  86L  87S C   N     622  624  -87.97  -72.40   21.67    1.73  -64.10  173.98   11.40\n",
      "   50         87S  87S N   CA    624  625  137.34  152.40                  -35.00\n",
      "\n",
      "\n",
      "report______> Distribution of short non-bonded contacts:\n",
      "\n",
      "  serious non-bonded atom clash:    90 1354   1.714\n",
      "  serious non-bonded atom clash:   122 1320   2.215\n",
      "  serious non-bonded atom clash:   144  727   2.264\n",
      "  serious non-bonded atom clash:   151 1196   1.945\n",
      "  serious non-bonded atom clash:   190 1197   1.758\n",
      "  serious non-bonded atom clash:   664  694   2.295\n",
      "  serious non-bonded atom clash:   671  702   1.565\n",
      "  serious non-bonded atom clash:   719 1120   2.189\n",
      "  serious non-bonded atom clash:   720 1119   1.741\n",
      "  serious non-bonded atom clash:   721 1119   1.648\n",
      "  serious non-bonded atom clash:   726 1152   1.768\n",
      "  serious non-bonded atom clash:   726 1154   2.269\n",
      "  serious non-bonded atom clash:   727 1152   2.294\n",
      "  serious non-bonded atom clash:   728 1152   2.162\n",
      "  serious non-bonded atom clash:   728 1154   1.683\n",
      "  serious non-bonded atom clash:   729 1152   2.041\n",
      "  serious non-bonded atom clash:   729 1153   2.155\n",
      "  serious non-bonded atom clash:   729 1154   1.124\n",
      "  serious non-bonded atom clash:   757  797   2.239\n",
      "  serious non-bonded atom clash:   757  798   2.053\n",
      "  serious non-bonded atom clash:   809  817   2.094\n",
      "  serious non-bonded atom clash:   823  827   2.273\n",
      "  serious non-bonded atom clash:   882 1210   2.153\n",
      "  serious non-bonded atom clash:  1000 1371   1.984\n",
      "  serious non-bonded atom clash:  1019 1041   1.587\n",
      "  serious non-bonded atom clash:  1019 1043   1.762\n",
      "  serious non-bonded atom clash:  1063 1067   2.231\n",
      "  serious non-bonded atom clash:  1111 1455   1.857\n",
      "  serious non-bonded atom clash:  1111 1457   1.212\n",
      "  serious non-bonded atom clash:  1111 1459   1.921\n",
      "  serious non-bonded atom clash:  1139 1383   2.285\n",
      "  serious non-bonded atom clash:  1141 1382   2.235\n",
      "  serious non-bonded atom clash:  1141 1383   2.023\n",
      "  serious non-bonded atom clash:  1168 1342   1.700\n",
      "  serious non-bonded atom clash:  1168 1343   2.160\n",
      "  serious non-bonded atom clash:  1169 1340   1.956\n",
      "  serious non-bonded atom clash:  1169 1342   2.045\n",
      "  serious non-bonded atom clash:  1169 1343   1.522\n",
      "  serious non-bonded atom clash:  1169 1345   2.168\n",
      "  serious non-bonded atom clash:  1170 1312   1.597\n",
      "  serious non-bonded atom clash:  1170 1339   2.166\n",
      "  serious non-bonded atom clash:  1170 1340   0.933\n",
      "  serious non-bonded atom clash:  1170 1342   2.001\n",
      "  serious non-bonded atom clash:  1170 1343   0.763\n",
      "  serious non-bonded atom clash:  1170 1344   1.750\n",
      "  serious non-bonded atom clash:  1170 1345   1.567\n",
      "  serious non-bonded atom clash:  1170 1346   2.186\n",
      "  serious non-bonded atom clash:  1170 1347   2.117\n",
      "  serious non-bonded atom clash:  1172 1310   2.198\n",
      "  serious non-bonded atom clash:  1172 1312   1.497\n",
      "  serious non-bonded atom clash:  1172 1313   2.013\n",
      "  serious non-bonded atom clash:  1172 1314   1.631\n",
      "  serious non-bonded atom clash:  1172 1340   1.411\n",
      "  serious non-bonded atom clash:  1172 1343   2.110\n",
      "  serious non-bonded atom clash:  1172 1344   0.608\n",
      "  serious non-bonded atom clash:  1172 1346   1.043\n",
      "  serious non-bonded atom clash:  1172 1347   1.878\n",
      "  serious non-bonded atom clash:  1174 1313   2.173\n",
      "  serious non-bonded atom clash:  1174 1315   2.177\n",
      "  serious non-bonded atom clash:  1174 1344   1.525\n",
      "  serious non-bonded atom clash:  1174 1346   1.595\n",
      "  serious non-bonded atom clash:  1195 1289   1.622\n",
      "  serious non-bonded atom clash:  1196 1289   1.533\n",
      "  serious non-bonded atom clash:  1197 1287   2.140\n",
      "  serious non-bonded atom clash:  1197 1288   1.553\n",
      "  serious non-bonded atom clash:  1197 1289   2.144\n",
      "  serious non-bonded atom clash:  1216 1262   2.299\n",
      "  serious non-bonded atom clash:  1218 1284   1.935\n",
      "  serious non-bonded atom clash:  1219 1262   2.021\n",
      "  serious non-bonded atom clash:  1219 1277   1.863\n",
      "  serious non-bonded atom clash:  1219 1278   2.125\n",
      "  serious non-bonded atom clash:  1219 1279   1.883\n",
      "  serious non-bonded atom clash:  1219 1280   2.163\n",
      "  serious non-bonded atom clash:  1262 1279   1.800\n",
      "  serious non-bonded atom clash:  1308 1346   1.922\n",
      "  serious non-bonded atom clash:  1309 1346   1.814\n",
      "  serious non-bonded atom clash:  1309 1347   1.856\n",
      "  serious non-bonded atom clash:  1310 1345   1.752\n",
      "  serious non-bonded atom clash:  1310 1346   1.549\n",
      "  serious non-bonded atom clash:  1310 1347   0.442\n",
      "  serious non-bonded atom clash:  1311 1345   1.843\n",
      "  serious non-bonded atom clash:  1311 1347   1.407\n",
      "  serious non-bonded atom clash:  1312 1340   2.114\n",
      "  serious non-bonded atom clash:  1312 1343   1.842\n",
      "  serious non-bonded atom clash:  1312 1344   1.964\n",
      "  serious non-bonded atom clash:  1312 1345   1.299\n",
      "  serious non-bonded atom clash:  1312 1346   1.407\n",
      "  serious non-bonded atom clash:  1312 1347   0.954\n",
      "  serious non-bonded atom clash:  1313 1344   1.825\n",
      "  serious non-bonded atom clash:  1313 1346   1.568\n",
      "  serious non-bonded atom clash:  1314 1344   1.208\n",
      "  serious non-bonded atom clash:  1314 1346   1.796\n",
      "  serious non-bonded atom clash:  1436 1474   1.768\n",
      "  serious non-bonded atom clash:  1436 1475   0.966\n",
      "  serious non-bonded atom clash:  1437 1475   2.218\n",
      "  serious non-bonded atom clash:  1438 1475   1.747\n",
      "  serious non-bonded atom clash:  1486 1489   2.181\n",
      "  serious non-bonded atom clash:  1486 1522   2.155\n",
      "  serious non-bonded atom clash:  1489 1522   2.077\n",
      "  serious non-bonded atom clash:  1490 1519   1.863\n",
      "  serious non-bonded atom clash:  1490 1520   2.124\n",
      "  serious non-bonded atom clash:  1491 1519   1.983\n",
      "  serious non-bonded atom clash:  1491 1520   1.662\n",
      "  serious non-bonded atom clash:  1491 1521   1.787\n",
      "  serious non-bonded atom clash:  1492 1494   2.068\n",
      "  serious non-bonded atom clash:  1564 1583   1.439\n",
      "  serious non-bonded atom clash:  1564 1585   2.034\n",
      "  serious non-bonded atom clash:  1575 1583   2.278\n",
      "  serious non-bonded atom clash:  1576 1583   2.243\n",
      "\n",
      "DISTANCE1:  0.00 2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40\n",
      "DISTANCE2:  2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40 3.50\n",
      "FREQUENCY:    74   21   14    7   24   33   38   73  107  142  150  205  216  209  269\n",
      "\n",
      "\n",
      "<< end of ENERGY.\n",
      "openf___224_> Open           Target.B99990003.pdb\n",
      "wrpdb___568_> Residues, atoms, selected atoms:      207     1613     1613\n",
      "openf___224_> Open           Target.ini\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      3936160    3843.906     3.754\n",
      "read_mo_297_> Segments, residues, atoms:        1      207     1613\n",
      "read_mo_298_> Segment:        1  1:A  207:A     1613\n",
      "randomi_498_> Atoms,selected atoms,random_seed,amplitude:     1613      457        1        4.0000\n",
      "randomi_496_> Amplitude is > 0; randomization is done.\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1927\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1933\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1938\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1956\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "iupac_m_487_> NH1/2 swapped:      179.3987        6  6\n",
      "iupac_m_486_> OE1/2 will be swapped:     -105.1969       13  13\n",
      "iupac_m_487_> NH1/2 swapped:      179.7925       26  26\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:      169.7476       28  28\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_485_> OD1/2 will be swapped:      172.2769       34  34\n",
      "iupac_m_487_> NH1/2 swapped:      179.9441       98  98\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      103  103\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0163      104  104\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      107  107\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      109  109\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      110  110\n",
      "iupac_m_485_> OD1/2 will be swapped:     -179.9802      111  111\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      114  114\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0292      124  124\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9657      127  127\n",
      "iupac_m_487_> NH1/2 swapped:      179.9558      129  129\n",
      "iupac_m_486_> OE1/2 will be swapped:      154.4214      135  135\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      136  136\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0239      138  138\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9558      139  139\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9720      140  140\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9407      146  146\n",
      "iupac_m_486_> OE1/2 will be swapped:      179.9604      152  152\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      160  160\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      164  164\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      165  165\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      179  179\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:     -176.2274      180  180\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -180.0000      183  183\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      196  196\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      200  200\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0309      202  202\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      207  207\n",
      "\n",
      "\n",
      ">> ENERGY; Differences between the model's features and restraints:\n",
      "Number of all residues in MODEL                   :      207\n",
      "Number of all, selected real atoms                :     1613     457\n",
      "Number of all, selected pseudo atoms              :        0       0\n",
      "Number of all static, selected restraints         :    12693    1983\n",
      "COVALENT_CYS                                      :        F\n",
      "NONBONDED_SEL_ATOMS                               :        1\n",
      "Number of non-bonded pairs (excluding 1-2,1-3,1-4):      541\n",
      "Dynamic pairs routine                             : 5, NATM_SEL x NATM double loop\n",
      "Atomic shift for contacts update (UPDATE_DYNAMIC) :    0.390\n",
      "LENNARD_JONES_SWITCH                              :    6.500   7.500\n",
      "COULOMB_JONES_SWITCH                              :    6.500   7.500\n",
      "RESIDUE_SPAN_RANGE                                :        0   99999\n",
      "NLOGN_USE                                         :       15\n",
      "CONTACT_SHELL                                     :    4.000\n",
      "DYNAMIC_PAIRS,_SPHERE,_COULOMB,_LENNARD,_MODELLER :        T       T       F       F       F\n",
      "SPHERE_STDV                                       :    0.050\n",
      "RADII_FACTOR                                      :    0.820\n",
      "Current energy                                    :         309.3078\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Summary of the restraint violations: \n",
      "\n",
      "   NUM     ... number of restraints.\n",
      "   NUMVI   ... number of restraints with RVIOL > VIOL_REPORT_CUT[i].\n",
      "   RVIOL   ... relative difference from the best value.\n",
      "   NUMVP   ... number of restraints with -Ln(pdf) > VIOL_REPORT_CUT2[i].\n",
      "   RMS_1   ... RMS(feature, minimally_violated_basis_restraint, NUMB).\n",
      "   RMS_2   ... RMS(feature, best_value, NUMB).\n",
      "   MOL.PDF ... scaled contribution to -Ln(Molecular pdf).\n",
      "\n",
      " #                     RESTRAINT_GROUP      NUM   NUMVI  NUMVP   RMS_1   RMS_2         MOL.PDF     S_i\n",
      "------------------------------------------------------------------------------------------------------\n",
      " 1 Bond length potential              :     478       0      0   0.003   0.003      1.0825       1.000\n",
      " 2 Bond angle potential               :     667       0      1   2.237   2.237      63.166       1.000\n",
      " 3 Stereochemical cosine torsion poten:     343       0      8  53.039  53.039      134.99       1.000\n",
      " 4 Stereochemical improper torsion pot:     181       0      0   0.929   0.929      3.2773       1.000\n",
      " 5 Soft-sphere overlap restraints     :     541       0      0   0.002   0.002     0.30485       1.000\n",
      " 6 Lennard-Jones 6-12 potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 7 Coulomb point-point electrostatic p:       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 8 H-bonding potential                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 9 Distance restraints 1 (CA-CA)      :      29       0      0   0.277   0.277     0.19268       1.000\n",
      "10 Distance restraints 2 (N-O)        :      30       0      0   0.818   0.818      1.6585       1.000\n",
      "11 Mainchain Phi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "12 Mainchain Psi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "13 Mainchain Omega dihedral restraints:      70       0      0   2.433   2.433      4.8872       1.000\n",
      "14 Sidechain Chi_1 dihedral restraints:      43       0      0  79.110  79.110      9.8793       1.000\n",
      "15 Sidechain Chi_2 dihedral restraints:      31       0      0  72.831  72.831      4.0145       1.000\n",
      "16 Sidechain Chi_3 dihedral restraints:      21       0      0  54.231  54.231      3.7785       1.000\n",
      "17 Sidechain Chi_4 dihedral restraints:       3       0      0 123.710 123.710      2.1987       1.000\n",
      "18 Disulfide distance restraints      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "19 Disulfide angle restraints         :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "20 Disulfide dihedral angle restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "21 Lower bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "22 Upper bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "23 Distance restraints 3 (SDCH-MNCH)  :      16       0      0   0.737   0.737     0.13760       1.000\n",
      "24 Sidechain Chi_5 dihedral restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "25 Phi/Psi pair of dihedral restraints:      71      48      7  30.602 146.327      79.738       1.000\n",
      "26 Distance restraints 4 (SDCH-SDCH)  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "27 Distance restraints 5 (X-Y)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "28 NMR distance restraints 6 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "29 NMR distance restraints 7 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "30 Minimal distance restraints        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "31 Non-bonded restraints              :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "32 Atomic accessibility restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "33 Atomic density restraints          :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "34 Absolute position restraints       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "35 Dihedral angle difference restraint:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "36 GBSA implicit solvent potential    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "37 EM density fitting potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "38 SAXS restraints                    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "39 Symmetry restraints                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "\n",
      "\n",
      "\n",
      "openf___224_> Open           Target.V99990004\n",
      "# Heavy relative violation of each residue is written to: Target.V99990004\n",
      "# The profile is NOT normalized by the number of restraints.\n",
      "# The profiles are smoothed over a window of residues:    1\n",
      "# The sum of all numbers in the file:   13039.7070\n",
      "\n",
      "\n",
      "\n",
      "List of the violated restraints:\n",
      "   A restraint is violated when the relative difference\n",
      "   from the best value (RVIOL) is larger than CUTOFF.\n",
      "\n",
      "   ICSR   ... index of a restraint in the current set.\n",
      "   RESNO  ... residue numbers of the first two atoms.\n",
      "   ATM    ... IUPAC atom names of the first two atoms.\n",
      "   FEAT   ... the value of the feature in the model.\n",
      "   restr  ... the mean of the basis restraint with the smallest\n",
      "              difference from the model (local minimum).\n",
      "   viol   ... difference from the local minimum.\n",
      "   rviol  ... relative difference from the local minimum.\n",
      "   RESTR  ... the best value (global minimum).\n",
      "   VIOL   ... difference from the best value.\n",
      "   RVIOL  ... relative difference from the best value.\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 25                           : Phi/Psi pair of dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       6.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   5791   1M   2A C   N       7    9 -114.28 -134.00   23.75    0.58  -62.50 -177.83   27.81\n",
      "    1          2A   2A N   CA      9   10  133.76  147.00                  -40.90\n",
      "    2   5792   2A   3H C   N      12   14 -104.60 -125.60   25.61    0.64  -63.20  171.50   19.22\n",
      "    2          3H   3H N   CA     14   15  124.13  138.80                  -42.30\n",
      "    3   5793   3H   4A C   N      22   24 -126.18 -134.00    7.85    0.24  -62.50 -175.87   33.58\n",
      "    3          4A   4A N   CA     24   25  146.33  147.00                  -40.90\n",
      "    4   5794   4A   5G C   N      27   29 -158.57 -167.20   15.80    0.44   82.20 -166.14   14.44\n",
      "    4          5G   5G N   CA     29   30  161.36  174.60                    8.50\n",
      "    5   5795   5G   6R C   N      31   33 -128.93 -125.20    9.24    0.35  -63.00 -177.81   29.15\n",
      "    5          6R   6R N   CA     33   34  149.05  140.60                  -41.10\n",
      "    6   5796   6R   7T C   N      42   44  -91.60  -78.10   34.60    1.93  -63.20  162.54   19.10\n",
      "    6          7T   7T N   CA     44   45  117.94  149.80                  -42.10\n",
      "    7   5799   9Y  10D C   N      65   67 -171.00  -63.30  110.26   18.40  -63.30  110.26   18.40\n",
      "    7         10D  10D N   CA     67   68  -63.60  -40.00                  -40.00\n",
      "    8   5821  31D  32A C   N     265  267  -96.54  -68.20   29.24    2.70  -62.50 -177.76   28.39\n",
      "    8         32A  32A N   CA    267  268  138.13  145.30                  -40.90\n",
      "    9   5822  32A  33G C   N     270  272  147.84 -167.20   63.02    2.35   82.20  138.48    9.68\n",
      "    9         33G  33G N   CA    272  273  130.45  174.60                    8.50\n",
      "   10   5823  33G  34D C   N     274  276 -141.47  -96.50   70.06    2.94  -63.30  171.00   26.14\n",
      "   10         34D  34D N   CA    276  277  167.92  114.20                  -40.00\n",
      "   11   5825  35V  36G C   N     289  291 -155.68 -167.20   11.64    0.27   82.20 -152.47   15.30\n",
      "   11         36G  36G N   CA    291  292  176.31  174.60                    8.50\n",
      "   12   5826  36G  37A C   N     293  295 -118.62 -134.00   20.95    1.21  -62.50  167.55   30.46\n",
      "   12         37A  37A N   CA    295  296  161.23  147.00                  -40.90\n",
      "   13   5827  37A  38A C   N     298  300 -116.48 -134.00   26.69    0.87  -62.50  176.23   26.81\n",
      "   13         38A  38A N   CA    300  301  126.86  147.00                  -40.90\n",
      "   14   5830  40P  41G C   N     317  319 -164.43 -167.20   28.41    1.08   82.20  178.45   13.39\n",
      "   14         41G  41G N   CA    319  320  146.32  174.60                    8.50\n",
      "   15   5831  41G  42A C   N     321  323 -160.55 -134.00   30.44    0.70  -62.50 -174.72   35.58\n",
      "   15         42A  42A N   CA    323  324  161.89  147.00                  -40.90\n",
      "   16   5832  42A  43A C   N     326  328 -144.14 -134.00   11.67    0.27  -62.50 -174.71   34.74\n",
      "   16         43A  43A N   CA    328  329  152.76  147.00                  -40.90\n",
      "   17   5834  44P  45A C   N     338  340 -132.14 -134.00    2.75    0.09  -62.50 -172.45   34.46\n",
      "   17         45A  45A N   CA    340  341  144.96  147.00                  -40.90\n",
      "   18   5836  46P  47G C   N     350  352 -152.08 -167.20   24.49    0.62   82.20 -166.69   14.58\n",
      "   18         47G  47G N   CA    352  353  155.34  174.60                    8.50\n",
      "   19   5838  48I  49F C   N     362  364 -111.54 -124.20   13.08    0.42  -63.20 -177.77   28.54\n",
      "   19         49F  49F N   CA    364  365  140.00  143.30                  -44.30\n",
      "   20   5839  49F  50S C   N     373  375 -128.71 -136.60    7.94    0.34  -64.10 -175.40   17.93\n",
      "   20         50S  50S N   CA    375  376  152.07  151.20                  -35.00\n",
      "   21   5840  50S  51S C   N     379  381 -132.35 -136.60    6.29    0.35  -64.10 -177.59   18.05\n",
      "   21         51S  51S N   CA    381  382  155.83  151.20                  -35.00\n",
      "   22   5841  51S  52Q C   N     385  387 -117.72 -121.10   11.73    0.60  -63.80  177.17   28.92\n",
      "   22         52Q  52Q N   CA    387  388  150.94  139.70                  -40.30\n",
      "   23   5842  52Q  53P C   N     394  396  -53.60  -58.70    9.28    0.56  -64.50  174.88   13.63\n",
      "   23         53P  53P N   CA    396  397  -38.26  -30.50                  147.20\n",
      "   24   5844  54G  55H C   N     405  407  -71.51  -67.60    4.02    0.27  -63.20  176.96   22.65\n",
      "   24         55H  55H N   CA    407  408  140.93  140.00                  -42.30\n",
      "   25   5845  55H  56T C   N     415  417 -126.96 -124.80    2.95    0.10  -63.20 -176.19   27.79\n",
      "   25         56T  56T N   CA    417  418  145.51  143.50                  -42.10\n",
      "   26   5847  57P  58H C   N     429  431 -131.78 -125.60    6.72    0.18  -63.20 -170.86   28.13\n",
      "   26         58H  58H N   CA    431  432  141.43  138.80                  -42.30\n",
      "   27   5849  59P  60A C   N     446  448 -128.42 -134.00    6.02    0.14  -62.50 -173.57   34.08\n",
      "   27         60A  60A N   CA    448  449  144.72  147.00                  -40.90\n",
      "   28   5850  60A  61A C   N     451  453 -161.03 -134.00   32.12    0.77  -62.50 -176.55   35.30\n",
      "   28         61A  61A N   CA    453  454  164.36  147.00                  -40.90\n",
      "   29   5851  61A  62S C   N     456  458  -76.97  -72.40   17.37    0.88  -64.10  156.37   12.14\n",
      "   29         62S  62S N   CA    458  459  169.16  152.40                  -35.00\n",
      "   30   5852  62S  63R C   N     462  464  -85.30  -72.10   43.34    3.60  -63.00  143.46   17.98\n",
      "   30         63R  63R N   CA    464  465  100.62  141.90                  -41.10\n",
      "   31   5853  63R  64D C   N     473  475 -121.97  -96.50   30.78    1.28  -63.30 -178.76   19.62\n",
      "   31         64D  64D N   CA    475  476  131.48  114.20                  -40.00\n",
      "   32   5856  66V  67A C   N     495  497  -62.64  -68.20    7.11    0.46  -62.50  178.23   29.22\n",
      "   32         67A  67A N   CA    497  498  140.87  145.30                  -40.90\n",
      "   33   5858  68R  69T C   N     511  513  -76.23  -78.10   10.24    0.38  -63.20  178.64   23.77\n",
      "   33         69T  69T N   CA    513  514  139.74  149.80                  -42.10\n",
      "   34   5859  69T  70S C   N     518  520 -140.81 -136.60   16.43    0.73  -64.10  175.57   18.21\n",
      "   34         70S  70S N   CA    520  521  167.08  151.20                  -35.00\n",
      "   35   5861  71P  72L C   N     531  533 -149.43 -108.50   43.85    1.95  -63.50 -169.02   31.83\n",
      "   35         72L  72L N   CA    533  534  148.24  132.50                  -41.20\n",
      "   36   5862  72L  73Q C   N     539  541 -112.89 -121.10   19.07    0.74  -63.80  170.03   22.74\n",
      "   36         73Q  73Q N   CA    541  542  122.48  139.70                  -40.30\n",
      "   37   5863  73Q  74T C   N     548  550 -138.26  -63.20   79.59   13.50  -63.20   79.59   13.50\n",
      "   37         74T  74T N   CA    550  551  -68.55  -42.10                  -42.10\n",
      "   38   5865  75P  76A C   N     562  564 -167.39 -134.00   34.81    0.84  -62.50 -166.77   37.23\n",
      "   38         76A  76A N   CA    564  565  156.83  147.00                  -40.90\n",
      "   39   5866  76A  77A C   N     567  569 -150.89 -134.00   18.46    0.90  -62.50 -159.86   37.53\n",
      "   39         77A  77A N   CA    569  570  139.54  147.00                  -40.90\n",
      "   40   5868  78P  79G C   N     579  581 -163.31 -167.20   36.29    1.37   82.20  173.24   13.11\n",
      "   40         79G  79G N   CA    581  582  138.52  174.60                    8.50\n",
      "   41   5869  79G  80A C   N     583  585 -173.27 -134.00   42.55    0.97  -62.50 -168.90   37.16\n",
      "   41         80A  80A N   CA    585  586  163.38  147.00                  -40.90\n",
      "   42   5870  80A  81A C   N     588  590 -116.36 -134.00   17.98    0.47  -62.50 -176.31   32.97\n",
      "   42         81A  81A N   CA    590  591  143.49  147.00                  -40.90\n",
      "   43   5871  81A  82A C   N     593  595 -167.43 -134.00   39.64    0.95  -62.50 -176.27   35.65\n",
      "   43         82A  82A N   CA    595  596  168.29  147.00                  -40.90\n",
      "   44   5872  82A  83G C   N     598  600  166.72 -167.20   36.12    1.34   82.20  164.49   11.73\n",
      "   44         83G  83G N   CA    600  601  149.62  174.60                    8.50\n",
      "   45   5874  84P  85A C   N     609  611  -74.10  -68.20    6.00    0.47  -62.50  173.09   28.95\n",
      "   45         85A  85A N   CA    611  612  146.40  145.30                  -40.90\n",
      "   46   5875  85A  86L C   N     614  616 -110.92 -108.50    9.54    0.54  -63.50  171.18   21.56\n",
      "   46         86L  86L N   CA    616  617  123.27  132.50                  -41.20\n",
      "   47   5876  86L  87S C   N     622  624 -140.02 -136.60    4.67    0.15  -64.10 -173.25   18.94\n",
      "   47         87S  87S N   CA    624  625  154.37  151.20                  -35.00\n",
      "   48   5880  90P  91P C   N     649  651  -76.01  -64.50   14.41    1.01  -58.70  174.50   15.84\n",
      "   48         91P  91P N   CA    651  652  155.86  147.20                  -30.50\n",
      "\n",
      "\n",
      "report______> Distribution of short non-bonded contacts:\n",
      "\n",
      "  serious non-bonded atom clash:    90 1354   1.714\n",
      "  serious non-bonded atom clash:   122 1320   2.215\n",
      "  serious non-bonded atom clash:   144  727   2.264\n",
      "  serious non-bonded atom clash:   151 1196   1.945\n",
      "  serious non-bonded atom clash:   190 1197   1.758\n",
      "  serious non-bonded atom clash:   664  694   2.295\n",
      "  serious non-bonded atom clash:   671  702   1.565\n",
      "  serious non-bonded atom clash:   719 1120   2.189\n",
      "  serious non-bonded atom clash:   720 1119   1.741\n",
      "  serious non-bonded atom clash:   721 1119   1.648\n",
      "  serious non-bonded atom clash:   726 1152   1.768\n",
      "  serious non-bonded atom clash:   726 1154   2.269\n",
      "  serious non-bonded atom clash:   727 1152   2.294\n",
      "  serious non-bonded atom clash:   728 1152   2.162\n",
      "  serious non-bonded atom clash:   728 1154   1.683\n",
      "  serious non-bonded atom clash:   729 1152   2.041\n",
      "  serious non-bonded atom clash:   729 1153   2.155\n",
      "  serious non-bonded atom clash:   729 1154   1.124\n",
      "  serious non-bonded atom clash:   757  797   2.239\n",
      "  serious non-bonded atom clash:   757  798   2.053\n",
      "  serious non-bonded atom clash:   809  817   2.094\n",
      "  serious non-bonded atom clash:   823  827   2.273\n",
      "  serious non-bonded atom clash:   882 1210   2.153\n",
      "  serious non-bonded atom clash:  1000 1371   1.984\n",
      "  serious non-bonded atom clash:  1019 1041   1.587\n",
      "  serious non-bonded atom clash:  1019 1043   1.762\n",
      "  serious non-bonded atom clash:  1063 1067   2.231\n",
      "  serious non-bonded atom clash:  1111 1455   1.857\n",
      "  serious non-bonded atom clash:  1111 1457   1.212\n",
      "  serious non-bonded atom clash:  1111 1459   1.921\n",
      "  serious non-bonded atom clash:  1139 1383   2.285\n",
      "  serious non-bonded atom clash:  1141 1382   2.235\n",
      "  serious non-bonded atom clash:  1141 1383   2.023\n",
      "  serious non-bonded atom clash:  1168 1342   1.700\n",
      "  serious non-bonded atom clash:  1168 1343   2.160\n",
      "  serious non-bonded atom clash:  1169 1340   1.956\n",
      "  serious non-bonded atom clash:  1169 1342   2.045\n",
      "  serious non-bonded atom clash:  1169 1343   1.522\n",
      "  serious non-bonded atom clash:  1169 1345   2.168\n",
      "  serious non-bonded atom clash:  1170 1312   1.597\n",
      "  serious non-bonded atom clash:  1170 1339   2.166\n",
      "  serious non-bonded atom clash:  1170 1340   0.933\n",
      "  serious non-bonded atom clash:  1170 1342   2.001\n",
      "  serious non-bonded atom clash:  1170 1343   0.763\n",
      "  serious non-bonded atom clash:  1170 1344   1.750\n",
      "  serious non-bonded atom clash:  1170 1345   1.567\n",
      "  serious non-bonded atom clash:  1170 1346   2.186\n",
      "  serious non-bonded atom clash:  1170 1347   2.117\n",
      "  serious non-bonded atom clash:  1172 1310   2.198\n",
      "  serious non-bonded atom clash:  1172 1312   1.497\n",
      "  serious non-bonded atom clash:  1172 1313   2.013\n",
      "  serious non-bonded atom clash:  1172 1314   1.631\n",
      "  serious non-bonded atom clash:  1172 1340   1.411\n",
      "  serious non-bonded atom clash:  1172 1343   2.110\n",
      "  serious non-bonded atom clash:  1172 1344   0.608\n",
      "  serious non-bonded atom clash:  1172 1346   1.043\n",
      "  serious non-bonded atom clash:  1172 1347   1.878\n",
      "  serious non-bonded atom clash:  1174 1313   2.173\n",
      "  serious non-bonded atom clash:  1174 1315   2.177\n",
      "  serious non-bonded atom clash:  1174 1344   1.525\n",
      "  serious non-bonded atom clash:  1174 1346   1.595\n",
      "  serious non-bonded atom clash:  1195 1289   1.622\n",
      "  serious non-bonded atom clash:  1196 1289   1.533\n",
      "  serious non-bonded atom clash:  1197 1287   2.140\n",
      "  serious non-bonded atom clash:  1197 1288   1.553\n",
      "  serious non-bonded atom clash:  1197 1289   2.144\n",
      "  serious non-bonded atom clash:  1216 1262   2.299\n",
      "  serious non-bonded atom clash:  1218 1284   1.935\n",
      "  serious non-bonded atom clash:  1219 1262   2.021\n",
      "  serious non-bonded atom clash:  1219 1277   1.863\n",
      "  serious non-bonded atom clash:  1219 1278   2.125\n",
      "  serious non-bonded atom clash:  1219 1279   1.883\n",
      "  serious non-bonded atom clash:  1219 1280   2.163\n",
      "  serious non-bonded atom clash:  1262 1279   1.800\n",
      "  serious non-bonded atom clash:  1308 1346   1.922\n",
      "  serious non-bonded atom clash:  1309 1346   1.814\n",
      "  serious non-bonded atom clash:  1309 1347   1.856\n",
      "  serious non-bonded atom clash:  1310 1345   1.752\n",
      "  serious non-bonded atom clash:  1310 1346   1.549\n",
      "  serious non-bonded atom clash:  1310 1347   0.442\n",
      "  serious non-bonded atom clash:  1311 1345   1.843\n",
      "  serious non-bonded atom clash:  1311 1347   1.407\n",
      "  serious non-bonded atom clash:  1312 1340   2.114\n",
      "  serious non-bonded atom clash:  1312 1343   1.842\n",
      "  serious non-bonded atom clash:  1312 1344   1.964\n",
      "  serious non-bonded atom clash:  1312 1345   1.299\n",
      "  serious non-bonded atom clash:  1312 1346   1.407\n",
      "  serious non-bonded atom clash:  1312 1347   0.954\n",
      "  serious non-bonded atom clash:  1313 1344   1.825\n",
      "  serious non-bonded atom clash:  1313 1346   1.568\n",
      "  serious non-bonded atom clash:  1314 1344   1.208\n",
      "  serious non-bonded atom clash:  1314 1346   1.796\n",
      "  serious non-bonded atom clash:  1436 1474   1.768\n",
      "  serious non-bonded atom clash:  1436 1475   0.966\n",
      "  serious non-bonded atom clash:  1437 1475   2.218\n",
      "  serious non-bonded atom clash:  1438 1475   1.747\n",
      "  serious non-bonded atom clash:  1486 1489   2.181\n",
      "  serious non-bonded atom clash:  1486 1522   2.155\n",
      "  serious non-bonded atom clash:  1489 1522   2.077\n",
      "  serious non-bonded atom clash:  1490 1519   1.863\n",
      "  serious non-bonded atom clash:  1490 1520   2.124\n",
      "  serious non-bonded atom clash:  1491 1519   1.983\n",
      "  serious non-bonded atom clash:  1491 1520   1.662\n",
      "  serious non-bonded atom clash:  1491 1521   1.787\n",
      "  serious non-bonded atom clash:  1492 1494   2.068\n",
      "  serious non-bonded atom clash:  1564 1583   1.439\n",
      "  serious non-bonded atom clash:  1564 1585   2.034\n",
      "  serious non-bonded atom clash:  1575 1583   2.278\n",
      "  serious non-bonded atom clash:  1576 1583   2.243\n",
      "\n",
      "DISTANCE1:  0.00 2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40\n",
      "DISTANCE2:  2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40 3.50\n",
      "FREQUENCY:    74   21   14    7   24   31   38   75  106  144  153  196  217  212  273\n",
      "\n",
      "\n",
      "<< end of ENERGY.\n",
      "openf___224_> Open           Target.B99990004.pdb\n",
      "wrpdb___568_> Residues, atoms, selected atoms:      207     1613     1613\n",
      "openf___224_> Open           Target.ini\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      3936160    3843.906     3.754\n",
      "read_mo_297_> Segments, residues, atoms:        1      207     1613\n",
      "read_mo_298_> Segment:        1  1:A  207:A     1613\n",
      "randomi_498_> Atoms,selected atoms,random_seed,amplitude:     1613      457        1        4.0000\n",
      "randomi_496_> Amplitude is > 0; randomization is done.\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1927\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1933\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1938\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1956\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "iupac_m_487_> NH1/2 swapped:      170.3274        6  6\n",
      "iupac_m_486_> OE1/2 will be swapped:     -105.1969       13  13\n",
      "iupac_m_487_> NH1/2 swapped:      179.7925       26  26\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:      169.7476       28  28\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_485_> OD1/2 will be swapped:     -105.6172       34  34\n",
      "iupac_m_487_> NH1/2 swapped:      179.9604       68  68\n",
      "iupac_m_487_> NH1/2 swapped:      179.9441       98  98\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      103  103\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0163      104  104\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      107  107\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      109  109\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      110  110\n",
      "iupac_m_485_> OD1/2 will be swapped:     -179.9802      111  111\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      114  114\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0292      124  124\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9657      127  127\n",
      "iupac_m_487_> NH1/2 swapped:      179.9558      129  129\n",
      "iupac_m_486_> OE1/2 will be swapped:      154.4214      135  135\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      136  136\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0239      138  138\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9558      139  139\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9720      140  140\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9407      146  146\n",
      "iupac_m_486_> OE1/2 will be swapped:      179.9604      152  152\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      160  160\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      164  164\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      165  165\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      179  179\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:     -176.2274      180  180\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -180.0000      183  183\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      196  196\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      200  200\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0309      202  202\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      207  207\n",
      "\n",
      "\n",
      ">> ENERGY; Differences between the model's features and restraints:\n",
      "Number of all residues in MODEL                   :      207\n",
      "Number of all, selected real atoms                :     1613     457\n",
      "Number of all, selected pseudo atoms              :        0       0\n",
      "Number of all static, selected restraints         :    12693    1983\n",
      "COVALENT_CYS                                      :        F\n",
      "NONBONDED_SEL_ATOMS                               :        1\n",
      "Number of non-bonded pairs (excluding 1-2,1-3,1-4):      560\n",
      "Dynamic pairs routine                             : 5, NATM_SEL x NATM double loop\n",
      "Atomic shift for contacts update (UPDATE_DYNAMIC) :    0.390\n",
      "LENNARD_JONES_SWITCH                              :    6.500   7.500\n",
      "COULOMB_JONES_SWITCH                              :    6.500   7.500\n",
      "RESIDUE_SPAN_RANGE                                :        0   99999\n",
      "NLOGN_USE                                         :       15\n",
      "CONTACT_SHELL                                     :    4.000\n",
      "DYNAMIC_PAIRS,_SPHERE,_COULOMB,_LENNARD,_MODELLER :        T       T       F       F       F\n",
      "SPHERE_STDV                                       :    0.050\n",
      "RADII_FACTOR                                      :    0.820\n",
      "Current energy                                    :         264.5612\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Summary of the restraint violations: \n",
      "\n",
      "   NUM     ... number of restraints.\n",
      "   NUMVI   ... number of restraints with RVIOL > VIOL_REPORT_CUT[i].\n",
      "   RVIOL   ... relative difference from the best value.\n",
      "   NUMVP   ... number of restraints with -Ln(pdf) > VIOL_REPORT_CUT2[i].\n",
      "   RMS_1   ... RMS(feature, minimally_violated_basis_restraint, NUMB).\n",
      "   RMS_2   ... RMS(feature, best_value, NUMB).\n",
      "   MOL.PDF ... scaled contribution to -Ln(Molecular pdf).\n",
      "\n",
      " #                     RESTRAINT_GROUP      NUM   NUMVI  NUMVP   RMS_1   RMS_2         MOL.PDF     S_i\n",
      "------------------------------------------------------------------------------------------------------\n",
      " 1 Bond length potential              :     478       0      0   0.002   0.002     0.54066       1.000\n",
      " 2 Bond angle potential               :     667       0      0   1.994   1.994      51.969       1.000\n",
      " 3 Stereochemical cosine torsion poten:     343       0      3  51.382  51.382      118.69       1.000\n",
      " 4 Stereochemical improper torsion pot:     181       0      0   0.612   0.612      1.6402       1.000\n",
      " 5 Soft-sphere overlap restraints     :     560       0      0   0.002   0.002     0.16394       1.000\n",
      " 6 Lennard-Jones 6-12 potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 7 Coulomb point-point electrostatic p:       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 8 H-bonding potential                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 9 Distance restraints 1 (CA-CA)      :      29       0      0   0.012   0.012     0.48381E-03   1.000\n",
      "10 Distance restraints 2 (N-O)        :      30       0      0   0.643   0.643      1.5562       1.000\n",
      "11 Mainchain Phi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "12 Mainchain Psi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "13 Mainchain Omega dihedral restraints:      70       0      0   2.101   2.101      3.6454       1.000\n",
      "14 Sidechain Chi_1 dihedral restraints:      43       0      1  76.225  76.225      11.137       1.000\n",
      "15 Sidechain Chi_2 dihedral restraints:      31       0      0  78.276  78.276      7.0184       1.000\n",
      "16 Sidechain Chi_3 dihedral restraints:      21       0      0  60.379  60.379      4.2227       1.000\n",
      "17 Sidechain Chi_4 dihedral restraints:       3       0      0  76.916  76.916      2.0503       1.000\n",
      "18 Disulfide distance restraints      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "19 Disulfide angle restraints         :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "20 Disulfide dihedral angle restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "21 Lower bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "22 Upper bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "23 Distance restraints 3 (SDCH-MNCH)  :      16       0      0   0.449   0.449     0.51953E-01   1.000\n",
      "24 Sidechain Chi_5 dihedral restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "25 Phi/Psi pair of dihedral restraints:      71      50      6  24.430 149.269      61.878       1.000\n",
      "26 Distance restraints 4 (SDCH-SDCH)  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "27 Distance restraints 5 (X-Y)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "28 NMR distance restraints 6 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "29 NMR distance restraints 7 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "30 Minimal distance restraints        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "31 Non-bonded restraints              :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "32 Atomic accessibility restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "33 Atomic density restraints          :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "34 Absolute position restraints       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "35 Dihedral angle difference restraint:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "36 GBSA implicit solvent potential    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "37 EM density fitting potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "38 SAXS restraints                    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "39 Symmetry restraints                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "\n",
      "\n",
      "\n",
      "openf___224_> Open           Target.V99990005\n",
      "# Heavy relative violation of each residue is written to: Target.V99990005\n",
      "# The profile is NOT normalized by the number of restraints.\n",
      "# The profiles are smoothed over a window of residues:    1\n",
      "# The sum of all numbers in the file:   12411.2129\n",
      "\n",
      "\n",
      "\n",
      "List of the violated restraints:\n",
      "   A restraint is violated when the relative difference\n",
      "   from the best value (RVIOL) is larger than CUTOFF.\n",
      "\n",
      "   ICSR   ... index of a restraint in the current set.\n",
      "   RESNO  ... residue numbers of the first two atoms.\n",
      "   ATM    ... IUPAC atom names of the first two atoms.\n",
      "   FEAT   ... the value of the feature in the model.\n",
      "   restr  ... the mean of the basis restraint with the smallest\n",
      "              difference from the model (local minimum).\n",
      "   viol   ... difference from the local minimum.\n",
      "   rviol  ... relative difference from the local minimum.\n",
      "   RESTR  ... the best value (global minimum).\n",
      "   VIOL   ... difference from the best value.\n",
      "   RVIOL  ... relative difference from the best value.\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 25                           : Phi/Psi pair of dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       6.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   5792   2A   3H C   N      12   14 -117.03 -125.60   20.56    0.91  -63.20  169.01   24.66\n",
      "    1          3H   3H N   CA     14   15  157.49  138.80                  -42.30\n",
      "    2   5793   3H   4A C   N      22   24  -68.91  -68.20   20.97    1.72  -62.50  165.37   26.79\n",
      "    2          4A   4A N   CA     24   25  124.35  145.30                  -40.90\n",
      "    3   5794   4A   5G C   N      27   29 -102.91  -80.20   53.91    1.26   82.20 -132.49    7.93\n",
      "    3          5G   5G N   CA     29   30 -137.00  174.10                    8.50\n",
      "    4   5795   5G   6R C   N      31   33  -64.29  -72.10   15.64    1.04  -63.00  169.46   22.81\n",
      "    4          6R   6R N   CA     33   34  128.36  141.90                  -41.10\n",
      "    5   5796   6R   7T C   N      42   44 -111.25 -124.80   35.58    1.17  -63.20  160.09   17.80\n",
      "    5          7T   7T N   CA     44   45  110.60  143.50                  -42.10\n",
      "    6   5797   7T   8G C   N      49   51 -160.97 -167.20   14.20    0.44   82.20 -167.22   14.31\n",
      "    6          8G   8G N   CA     51   52  161.84  174.60                    8.50\n",
      "    7   5798   8G   9Y C   N      53   55 -131.88 -124.30   63.40    3.04  -63.50  136.60   24.95\n",
      "    7          9Y   9Y N   CA     55   56 -161.65  135.40                  -43.40\n",
      "    8   5799   9Y  10D C   N      65   67 -130.82  -63.30   71.53   12.12  -63.30   71.53   12.12\n",
      "    8         10D  10D N   CA     67   68  -63.60  -40.00                  -40.00\n",
      "    9   5821  31D  32A C   N     265  267  -84.81  -68.20   16.76    1.49  -62.50  177.44   30.22\n",
      "    9         32A  32A N   CA    267  268  143.07  145.30                  -40.90\n",
      "   10   5822  32A  33G C   N     270  272 -174.12 -167.20   45.70    1.85   82.20  159.28   12.01\n",
      "   10         33G  33G N   CA    272  273  129.42  174.60                    8.50\n",
      "   11   5823  33G  34D C   N     274  276  -97.02  -70.90   38.64    1.49  -63.30  145.20   20.09\n",
      "   11         34D  34D N   CA    276  277  178.77  150.30                  -40.00\n",
      "   12   5825  35V  36G C   N     289  291  152.11 -167.20   41.36    0.64   82.20 -172.95   12.35\n",
      "   12         36G  36G N   CA    291  292 -178.00  174.60                    8.50\n",
      "   13   5826  36G  37A C   N     293  295 -139.94 -134.00   13.37    0.57  -62.50  177.86   33.30\n",
      "   13         37A  37A N   CA    295  296  158.98  147.00                  -40.90\n",
      "   14   5827  37A  38A C   N     298  300  -70.33  -68.20    8.33    0.60  -62.50  165.93   27.58\n",
      "   14         38A  38A N   CA    300  301  153.36  145.30                  -40.90\n",
      "   15   5829  39P  40P C   N     310  312  -69.61  -58.70   15.45    1.09  -64.50  166.84   12.72\n",
      "   15         40P  40P N   CA    312  313  -19.56  -30.50                  147.20\n",
      "   16   5830  40P  41G C   N     317  319  -65.61  -80.20   33.33    0.78   82.20 -159.39    6.59\n",
      "   16         41G  41G N   CA    319  320  144.13  174.10                    8.50\n",
      "   17   5831  41G  42A C   N     321  323 -138.68 -134.00    9.82    0.40  -62.50 -179.65   33.64\n",
      "   17         42A  42A N   CA    323  324  155.63  147.00                  -40.90\n",
      "   18   5832  42A  43A C   N     326  328 -106.21 -134.00   29.99    0.69  -62.50 -178.05   28.01\n",
      "   18         43A  43A N   CA    328  329  135.73  147.00                  -40.90\n",
      "   19   5834  44P  45A C   N     338  340 -116.33 -134.00   19.26    0.44  -62.50 -172.34   33.61\n",
      "   19         45A  45A N   CA    340  341  139.33  147.00                  -40.90\n",
      "   20   5836  46P  47G C   N     350  352 -163.40 -167.20   22.48    0.83   82.20 -176.14   13.73\n",
      "   20         47G  47G N   CA    352  353  152.44  174.60                    8.50\n",
      "   21   5838  48I  49F C   N     362  364 -114.19 -124.20   10.26    0.33  -63.20 -178.04   28.67\n",
      "   21         49F  49F N   CA    364  365  141.03  143.30                  -44.30\n",
      "   22   5839  49F  50S C   N     373  375 -136.12 -136.60    6.20    0.32  -64.10 -177.56   18.34\n",
      "   22         50S  50S N   CA    375  376  157.37  151.20                  -35.00\n",
      "   23   5840  50S  51S C   N     379  381 -130.13 -136.60    6.79    0.32  -64.10 -176.01   18.00\n",
      "   23         51S  51S N   CA    381  382  153.26  151.20                  -35.00\n",
      "   24   5841  51S  52Q C   N     385  387 -112.50 -121.10    9.15    0.29  -63.80 -176.55   24.69\n",
      "   24         52Q  52Q N   CA    387  388  136.57  139.70                  -40.30\n",
      "   25   5843  53P  54G C   N     401  403 -169.76 -167.20   19.56    0.79   82.20 -177.81   13.45\n",
      "   25         54G  54G N   CA    403  404  155.20  174.60                    8.50\n",
      "   26   5844  54G  55H C   N     405  407 -133.37 -125.60   17.32    0.50  -63.20  177.85   26.82\n",
      "   26         55H  55H N   CA    407  408  154.27  138.80                  -42.30\n",
      "   27   5845  55H  56T C   N     415  417 -114.14 -124.80   13.72    0.47  -63.20 -175.84   20.67\n",
      "   27         56T  56T N   CA    417  418  134.87  143.50                  -42.10\n",
      "   28   5847  57P  58H C   N     429  431 -122.69 -125.60    8.50    0.37  -63.20 -179.03   26.52\n",
      "   28         58H  58H N   CA    431  432  146.79  138.80                  -42.30\n",
      "   29   5849  59P  60A C   N     446  448 -118.78 -134.00   16.30    0.38  -62.50 -173.38   33.58\n",
      "   29         60A  60A N   CA    448  449  141.17  147.00                  -40.90\n",
      "   30   5850  60A  61A C   N     451  453 -146.52 -134.00   16.53    0.46  -62.50 -178.12   34.30\n",
      "   30         61A  61A N   CA    453  454  157.80  147.00                  -40.90\n",
      "   31   5851  61A  62S C   N     456  458  -74.94  -72.40   15.04    0.80  -64.10  158.14   12.14\n",
      "   31         62S  62S N   CA    458  459  167.23  152.40                  -35.00\n",
      "   32   5852  62S  63R C   N     462  464  -83.61  -72.10   44.81    3.68  -63.00  141.21   17.77\n",
      "   32         63R  63R N   CA    464  465   98.60  141.90                  -41.10\n",
      "   33   5853  63R  64D C   N     473  475 -130.23  -96.50   39.52    1.64  -63.30 -172.82   20.10\n",
      "   33         64D  64D N   CA    475  476  134.80  114.20                  -40.00\n",
      "   34   5856  66V  67A C   N     495  497  -68.78  -68.20    3.11    0.23  -62.50  170.86   28.31\n",
      "   34         67A  67A N   CA    497  498  148.36  145.30                  -40.90\n",
      "   35   5857  67A  68R C   N     500  502  -73.88  -72.10    1.78    0.14  -63.00  177.32   24.69\n",
      "   35         68R  68R N   CA    502  503  141.92  141.90                  -41.10\n",
      "   36   5858  68R  69T C   N     511  513 -120.94 -124.80    5.55    0.33  -63.20  179.92   26.90\n",
      "   36         69T  69T N   CA    513  514  147.49  143.50                  -42.10\n",
      "   37   5859  69T  70S C   N     518  520  -75.33  -72.40    5.00    0.23  -64.10  168.92   12.94\n",
      "   37         70S  70S N   CA    520  521  156.45  152.40                  -35.00\n",
      "   38   5861  71P  72L C   N     531  533  -70.30  -70.70    1.07    0.07  -63.50  178.32   25.15\n",
      "   38         72L  72L N   CA    533  534  140.61  141.60                  -41.20\n",
      "   39   5862  72L  73Q C   N     539  541 -118.61 -121.10    2.81    0.09  -63.80 -173.09   24.97\n",
      "   39         73Q  73Q N   CA    541  542  138.39  139.70                  -40.30\n",
      "   40   5863  73Q  74T C   N     548  550  -96.38  -78.10   24.52    0.94  -63.20  155.35   22.14\n",
      "   40         74T  74T N   CA    550  551  166.14  149.80                  -42.10\n",
      "   41   5865  75P  76A C   N     562  564  -68.14  -68.20    0.87    0.07  -62.50  174.76   28.92\n",
      "   41         76A  76A N   CA    564  565  144.44  145.30                  -40.90\n",
      "   42   5866  76A  77A C   N     567  569  -76.52  -68.20   10.66    0.68  -62.50  167.72   28.20\n",
      "   42         77A  77A N   CA    569  570  151.97  145.30                  -40.90\n",
      "   43   5868  78P  79G C   N     579  581  -76.62  -80.20   11.09    0.28   82.20 -138.01    7.04\n",
      "   43         79G  79G N   CA    581  582  163.60  174.10                    8.50\n",
      "   44   5869  79G  80A C   N     583  585  -69.22  -68.20    2.35    0.16  -62.50  171.81   28.49\n",
      "   44         80A  80A N   CA    585  586  147.43  145.30                  -40.90\n",
      "   45   5870  80A  81A C   N     588  590 -109.99 -134.00   27.44    0.63  -62.50 -179.05   27.74\n",
      "   45         81A  81A N   CA    590  591  133.71  147.00                  -40.90\n",
      "   46   5871  81A  82A C   N     593  595 -145.13 -134.00   16.37    0.51  -62.50 -179.82   33.95\n",
      "   46         82A  82A N   CA    595  596  158.99  147.00                  -40.90\n",
      "   47   5872  82A  83G C   N     598  600 -104.19  -80.20   29.53    2.16   82.20 -131.63   17.72\n",
      "   47         83G  83G N   CA    600  601  156.87  174.10                    8.50\n",
      "   48   5874  84P  85A C   N     609  611  -75.28  -68.20   10.43    0.66  -62.50  166.63   27.95\n",
      "   48         85A  85A N   CA    611  612  152.96  145.30                  -40.90\n",
      "   49   5875  85A  86L C   N     614  616 -107.08 -108.50    7.80    0.43  -63.50 -176.13   28.20\n",
      "   49         86L  86L N   CA    616  617  140.17  132.50                  -41.20\n",
      "   50   5876  86L  87S C   N     622  624 -123.90 -136.60   25.20    0.95  -64.10  174.97   10.79\n",
      "   50         87S  87S N   CA    624  625  129.43  151.20                  -35.00\n",
      "\n",
      "\n",
      "report______> Distribution of short non-bonded contacts:\n",
      "\n",
      "  serious non-bonded atom clash:    90 1354   1.714\n",
      "  serious non-bonded atom clash:   122 1320   2.215\n",
      "  serious non-bonded atom clash:   144  727   2.264\n",
      "  serious non-bonded atom clash:   151 1196   1.945\n",
      "  serious non-bonded atom clash:   190 1197   1.758\n",
      "  serious non-bonded atom clash:   664  694   2.295\n",
      "  serious non-bonded atom clash:   671  702   1.565\n",
      "  serious non-bonded atom clash:   719 1120   2.189\n",
      "  serious non-bonded atom clash:   720 1119   1.741\n",
      "  serious non-bonded atom clash:   721 1119   1.648\n",
      "  serious non-bonded atom clash:   726 1152   1.768\n",
      "  serious non-bonded atom clash:   726 1154   2.269\n",
      "  serious non-bonded atom clash:   727 1152   2.294\n",
      "  serious non-bonded atom clash:   728 1152   2.162\n",
      "  serious non-bonded atom clash:   728 1154   1.683\n",
      "  serious non-bonded atom clash:   729 1152   2.041\n",
      "  serious non-bonded atom clash:   729 1153   2.155\n",
      "  serious non-bonded atom clash:   729 1154   1.124\n",
      "  serious non-bonded atom clash:   757  797   2.239\n",
      "  serious non-bonded atom clash:   757  798   2.053\n",
      "  serious non-bonded atom clash:   809  817   2.094\n",
      "  serious non-bonded atom clash:   823  827   2.273\n",
      "  serious non-bonded atom clash:   882 1210   2.153\n",
      "  serious non-bonded atom clash:  1000 1371   1.984\n",
      "  serious non-bonded atom clash:  1019 1041   1.587\n",
      "  serious non-bonded atom clash:  1019 1043   1.762\n",
      "  serious non-bonded atom clash:  1063 1067   2.231\n",
      "  serious non-bonded atom clash:  1111 1455   1.857\n",
      "  serious non-bonded atom clash:  1111 1457   1.212\n",
      "  serious non-bonded atom clash:  1111 1459   1.921\n",
      "  serious non-bonded atom clash:  1139 1383   2.285\n",
      "  serious non-bonded atom clash:  1141 1382   2.235\n",
      "  serious non-bonded atom clash:  1141 1383   2.023\n",
      "  serious non-bonded atom clash:  1168 1342   1.700\n",
      "  serious non-bonded atom clash:  1168 1343   2.160\n",
      "  serious non-bonded atom clash:  1169 1340   1.956\n",
      "  serious non-bonded atom clash:  1169 1342   2.045\n",
      "  serious non-bonded atom clash:  1169 1343   1.522\n",
      "  serious non-bonded atom clash:  1169 1345   2.168\n",
      "  serious non-bonded atom clash:  1170 1312   1.597\n",
      "  serious non-bonded atom clash:  1170 1339   2.166\n",
      "  serious non-bonded atom clash:  1170 1340   0.933\n",
      "  serious non-bonded atom clash:  1170 1342   2.001\n",
      "  serious non-bonded atom clash:  1170 1343   0.763\n",
      "  serious non-bonded atom clash:  1170 1344   1.750\n",
      "  serious non-bonded atom clash:  1170 1345   1.567\n",
      "  serious non-bonded atom clash:  1170 1346   2.186\n",
      "  serious non-bonded atom clash:  1170 1347   2.117\n",
      "  serious non-bonded atom clash:  1172 1310   2.198\n",
      "  serious non-bonded atom clash:  1172 1312   1.497\n",
      "  serious non-bonded atom clash:  1172 1313   2.013\n",
      "  serious non-bonded atom clash:  1172 1314   1.631\n",
      "  serious non-bonded atom clash:  1172 1340   1.411\n",
      "  serious non-bonded atom clash:  1172 1343   2.110\n",
      "  serious non-bonded atom clash:  1172 1344   0.608\n",
      "  serious non-bonded atom clash:  1172 1346   1.043\n",
      "  serious non-bonded atom clash:  1172 1347   1.878\n",
      "  serious non-bonded atom clash:  1174 1313   2.173\n",
      "  serious non-bonded atom clash:  1174 1315   2.177\n",
      "  serious non-bonded atom clash:  1174 1344   1.525\n",
      "  serious non-bonded atom clash:  1174 1346   1.595\n",
      "  serious non-bonded atom clash:  1195 1289   1.622\n",
      "  serious non-bonded atom clash:  1196 1289   1.533\n",
      "  serious non-bonded atom clash:  1197 1287   2.140\n",
      "  serious non-bonded atom clash:  1197 1288   1.553\n",
      "  serious non-bonded atom clash:  1197 1289   2.144\n",
      "  serious non-bonded atom clash:  1216 1262   2.299\n",
      "  serious non-bonded atom clash:  1218 1284   1.935\n",
      "  serious non-bonded atom clash:  1219 1262   2.021\n",
      "  serious non-bonded atom clash:  1219 1277   1.863\n",
      "  serious non-bonded atom clash:  1219 1278   2.125\n",
      "  serious non-bonded atom clash:  1219 1279   1.883\n",
      "  serious non-bonded atom clash:  1219 1280   2.163\n",
      "  serious non-bonded atom clash:  1262 1279   1.800\n",
      "  serious non-bonded atom clash:  1308 1346   1.922\n",
      "  serious non-bonded atom clash:  1309 1346   1.814\n",
      "  serious non-bonded atom clash:  1309 1347   1.856\n",
      "  serious non-bonded atom clash:  1310 1345   1.752\n",
      "  serious non-bonded atom clash:  1310 1346   1.549\n",
      "  serious non-bonded atom clash:  1310 1347   0.442\n",
      "  serious non-bonded atom clash:  1311 1345   1.843\n",
      "  serious non-bonded atom clash:  1311 1347   1.407\n",
      "  serious non-bonded atom clash:  1312 1340   2.114\n",
      "  serious non-bonded atom clash:  1312 1343   1.842\n",
      "  serious non-bonded atom clash:  1312 1344   1.964\n",
      "  serious non-bonded atom clash:  1312 1345   1.299\n",
      "  serious non-bonded atom clash:  1312 1346   1.407\n",
      "  serious non-bonded atom clash:  1312 1347   0.954\n",
      "  serious non-bonded atom clash:  1313 1344   1.825\n",
      "  serious non-bonded atom clash:  1313 1346   1.568\n",
      "  serious non-bonded atom clash:  1314 1344   1.208\n",
      "  serious non-bonded atom clash:  1314 1346   1.796\n",
      "  serious non-bonded atom clash:  1436 1474   1.768\n",
      "  serious non-bonded atom clash:  1436 1475   0.966\n",
      "  serious non-bonded atom clash:  1437 1475   2.218\n",
      "  serious non-bonded atom clash:  1438 1475   1.747\n",
      "  serious non-bonded atom clash:  1486 1489   2.181\n",
      "  serious non-bonded atom clash:  1486 1522   2.155\n",
      "  serious non-bonded atom clash:  1489 1522   2.077\n",
      "  serious non-bonded atom clash:  1490 1519   1.863\n",
      "  serious non-bonded atom clash:  1490 1520   2.124\n",
      "  serious non-bonded atom clash:  1491 1519   1.983\n",
      "  serious non-bonded atom clash:  1491 1520   1.662\n",
      "  serious non-bonded atom clash:  1491 1521   1.787\n",
      "  serious non-bonded atom clash:  1492 1494   2.068\n",
      "  serious non-bonded atom clash:  1564 1583   1.439\n",
      "  serious non-bonded atom clash:  1564 1585   2.034\n",
      "  serious non-bonded atom clash:  1575 1583   2.278\n",
      "  serious non-bonded atom clash:  1576 1583   2.243\n",
      "\n",
      "DISTANCE1:  0.00 2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40\n",
      "DISTANCE2:  2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40 3.50\n",
      "FREQUENCY:    74   21   14    7   24   31   38   71  106  143  150  214  211  215  273\n",
      "\n",
      "\n",
      "<< end of ENERGY.\n",
      "openf___224_> Open           Target.B99990005.pdb\n",
      "wrpdb___568_> Residues, atoms, selected atoms:      207     1613     1613\n",
      "openf___224_> Open           Target.ini\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      3936160    3843.906     3.754\n",
      "read_mo_297_> Segments, residues, atoms:        1      207     1613\n",
      "read_mo_298_> Segment:        1  1:A  207:A     1613\n",
      "randomi_498_> Atoms,selected atoms,random_seed,amplitude:     1613      457        1        4.0000\n",
      "randomi_496_> Amplitude is > 0; randomization is done.\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1927\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1933\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1938\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1956\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "iupac_m_486_> OE1/2 will be swapped:     -105.1969       13  13\n",
      "iupac_m_487_> NH1/2 swapped:      179.7925       26  26\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:      169.7476       28  28\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_485_> OD1/2 will be swapped:      138.1106       34  34\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9234       63  63\n",
      "iupac_m_487_> NH1/2 swapped:     -177.0247       68  68\n",
      "iupac_m_487_> NH1/2 swapped:      179.9441       98  98\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      103  103\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0163      104  104\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      107  107\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      109  109\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      110  110\n",
      "iupac_m_485_> OD1/2 will be swapped:     -179.9802      111  111\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      114  114\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0292      124  124\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9657      127  127\n",
      "iupac_m_487_> NH1/2 swapped:      179.9558      129  129\n",
      "iupac_m_486_> OE1/2 will be swapped:      154.4214      135  135\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      136  136\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0239      138  138\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9558      139  139\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9720      140  140\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9407      146  146\n",
      "iupac_m_486_> OE1/2 will be swapped:      179.9604      152  152\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      160  160\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      164  164\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      165  165\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      179  179\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:     -176.2274      180  180\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -180.0000      183  183\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      196  196\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      200  200\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0309      202  202\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      207  207\n",
      "\n",
      "\n",
      ">> ENERGY; Differences between the model's features and restraints:\n",
      "Number of all residues in MODEL                   :      207\n",
      "Number of all, selected real atoms                :     1613     457\n",
      "Number of all, selected pseudo atoms              :        0       0\n",
      "Number of all static, selected restraints         :    12693    1983\n",
      "COVALENT_CYS                                      :        F\n",
      "NONBONDED_SEL_ATOMS                               :        1\n",
      "Number of non-bonded pairs (excluding 1-2,1-3,1-4):      570\n",
      "Dynamic pairs routine                             : 5, NATM_SEL x NATM double loop\n",
      "Atomic shift for contacts update (UPDATE_DYNAMIC) :    0.390\n",
      "LENNARD_JONES_SWITCH                              :    6.500   7.500\n",
      "COULOMB_JONES_SWITCH                              :    6.500   7.500\n",
      "RESIDUE_SPAN_RANGE                                :        0   99999\n",
      "NLOGN_USE                                         :       15\n",
      "CONTACT_SHELL                                     :    4.000\n",
      "DYNAMIC_PAIRS,_SPHERE,_COULOMB,_LENNARD,_MODELLER :        T       T       F       F       F\n",
      "SPHERE_STDV                                       :    0.050\n",
      "RADII_FACTOR                                      :    0.820\n",
      "Current energy                                    :         312.9248\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Summary of the restraint violations: \n",
      "\n",
      "   NUM     ... number of restraints.\n",
      "   NUMVI   ... number of restraints with RVIOL > VIOL_REPORT_CUT[i].\n",
      "   RVIOL   ... relative difference from the best value.\n",
      "   NUMVP   ... number of restraints with -Ln(pdf) > VIOL_REPORT_CUT2[i].\n",
      "   RMS_1   ... RMS(feature, minimally_violated_basis_restraint, NUMB).\n",
      "   RMS_2   ... RMS(feature, best_value, NUMB).\n",
      "   MOL.PDF ... scaled contribution to -Ln(Molecular pdf).\n",
      "\n",
      " #                     RESTRAINT_GROUP      NUM   NUMVI  NUMVP   RMS_1   RMS_2         MOL.PDF     S_i\n",
      "------------------------------------------------------------------------------------------------------\n",
      " 1 Bond length potential              :     478       0      0   0.003   0.003     0.85277       1.000\n",
      " 2 Bond angle potential               :     667       0      0   2.098   2.098      55.160       1.000\n",
      " 3 Stereochemical cosine torsion poten:     343       0      8  52.973  52.973      127.41       1.000\n",
      " 4 Stereochemical improper torsion pot:     181       0      0   0.714   0.714      2.1716       1.000\n",
      " 5 Soft-sphere overlap restraints     :     570       0      0   0.003   0.003     0.47549       1.000\n",
      " 6 Lennard-Jones 6-12 potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 7 Coulomb point-point electrostatic p:       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 8 H-bonding potential                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 9 Distance restraints 1 (CA-CA)      :      29       0      0   0.607   0.607      1.5159       1.000\n",
      "10 Distance restraints 2 (N-O)        :      30       0      0   1.751   1.751      12.439       1.000\n",
      "11 Mainchain Phi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "12 Mainchain Psi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "13 Mainchain Omega dihedral restraints:      70       0      0   2.111   2.111      3.6795       1.000\n",
      "14 Sidechain Chi_1 dihedral restraints:      43       0      0  80.522  80.522      9.5916       1.000\n",
      "15 Sidechain Chi_2 dihedral restraints:      31       0      0  70.259  70.259      4.5974       1.000\n",
      "16 Sidechain Chi_3 dihedral restraints:      21       0      0  56.238  56.238      3.4708       1.000\n",
      "17 Sidechain Chi_4 dihedral restraints:       3       0      0 134.677 134.677      2.1682       1.000\n",
      "18 Disulfide distance restraints      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "19 Disulfide angle restraints         :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "20 Disulfide dihedral angle restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "21 Lower bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "22 Upper bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "23 Distance restraints 3 (SDCH-MNCH)  :      16       0      0   1.045   1.045     0.27979       1.000\n",
      "24 Sidechain Chi_5 dihedral restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "25 Phi/Psi pair of dihedral restraints:      71      54      6  29.963 152.193      89.109       1.000\n",
      "26 Distance restraints 4 (SDCH-SDCH)  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "27 Distance restraints 5 (X-Y)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "28 NMR distance restraints 6 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "29 NMR distance restraints 7 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "30 Minimal distance restraints        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "31 Non-bonded restraints              :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "32 Atomic accessibility restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "33 Atomic density restraints          :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "34 Absolute position restraints       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "35 Dihedral angle difference restraint:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "36 GBSA implicit solvent potential    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "37 EM density fitting potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "38 SAXS restraints                    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "39 Symmetry restraints                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "\n",
      "\n",
      "\n",
      "openf___224_> Open           Target.V99990006\n",
      "# Heavy relative violation of each residue is written to: Target.V99990006\n",
      "# The profile is NOT normalized by the number of restraints.\n",
      "# The profiles are smoothed over a window of residues:    1\n",
      "# The sum of all numbers in the file:   13010.4707\n",
      "\n",
      "\n",
      "\n",
      "List of the violated restraints:\n",
      "   A restraint is violated when the relative difference\n",
      "   from the best value (RVIOL) is larger than CUTOFF.\n",
      "\n",
      "   ICSR   ... index of a restraint in the current set.\n",
      "   RESNO  ... residue numbers of the first two atoms.\n",
      "   ATM    ... IUPAC atom names of the first two atoms.\n",
      "   FEAT   ... the value of the feature in the model.\n",
      "   restr  ... the mean of the basis restraint with the smallest\n",
      "              difference from the model (local minimum).\n",
      "   viol   ... difference from the local minimum.\n",
      "   rviol  ... relative difference from the local minimum.\n",
      "   RESTR  ... the best value (global minimum).\n",
      "   VIOL   ... difference from the best value.\n",
      "   RVIOL  ... relative difference from the best value.\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 25                           : Phi/Psi pair of dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       6.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   5791   1M   2A C   N       7    9 -109.42 -134.00   24.68    0.74  -62.50 -179.54   32.06\n",
      "    1          2A   2A N   CA      9   10  144.85  147.00                  -40.90\n",
      "    2   5792   2A   3H C   N      12   14 -109.44 -125.60   16.50    0.64  -63.20 -178.41   25.72\n",
      "    2          3H   3H N   CA     14   15  142.10  138.80                  -42.30\n",
      "    3   5793   3H   4A C   N      22   24 -116.59 -134.00   17.45    0.53  -62.50 -178.37   32.64\n",
      "    3          4A   4A N   CA     24   25  145.72  147.00                  -40.90\n",
      "    4   5794   4A   5G C   N      27   29  170.87 -167.20   22.53    0.34   82.20 -167.15   13.38\n",
      "    4          5G   5G N   CA     29   30  179.76  174.60                    8.50\n",
      "    5   5795   5G   6R C   N      31   33  -77.62  -72.10   13.25    0.90  -63.00  165.61   23.37\n",
      "    5          6R   6R N   CA     33   34  153.94  141.90                  -41.10\n",
      "    6   5796   6R   7T C   N      42   44 -123.27 -124.80   17.52    0.80  -63.20  168.05   25.52\n",
      "    6          7T   7T N   CA     44   45  160.95  143.50                  -42.10\n",
      "    7   5797   7T   8G C   N      49   51   82.02   78.70   13.69    0.26   82.20  172.12    8.35\n",
      "    7          8G   8G N   CA     51   52 -179.38 -166.10                    8.50\n",
      "    8   5798   8G   9Y C   N      53   55 -133.65 -124.30   46.89    2.11  -63.50  152.36   27.59\n",
      "    8          9Y   9Y N   CA     55   56 -178.65  135.40                  -43.40\n",
      "    9   5799   9Y  10D C   N      65   67 -110.75  -63.30   53.00    9.05  -63.30   53.00    9.05\n",
      "    9         10D  10D N   CA     67   68  -63.60  -40.00                  -40.00\n",
      "   10   5821  31D  32A C   N     265  267  -84.23  -68.20   17.16    1.24  -62.50  169.07   28.82\n",
      "   10         32A  32A N   CA    267  268  151.44  145.30                  -40.90\n",
      "   11   5822  32A  33G C   N     270  272  175.88 -167.20   49.46    2.03   82.20  151.93   11.32\n",
      "   11         33G  33G N   CA    272  273  128.12  174.60                    8.50\n",
      "   12   5823  33G  34D C   N     274  276 -143.00  -96.50   47.06    1.93  -63.30 -179.98   19.11\n",
      "   12         34D  34D N   CA    276  277  121.42  114.20                  -40.00\n",
      "   13   5825  35V  36G C   N     289  291  146.89 -167.20   53.37    1.74   82.20  153.20   10.41\n",
      "   13         36G  36G N   CA    291  292  147.38  174.60                    8.50\n",
      "   14   5826  36G  37A C   N     293  295 -168.35 -134.00   38.88    0.89  -62.50 -173.21   36.21\n",
      "   14         37A  37A N   CA    295  296  165.20  147.00                  -40.90\n",
      "   15   5827  37A  38A C   N     298  300  -68.69  -68.20    3.78    0.32  -62.50  177.65   29.42\n",
      "   15         38A  38A N   CA    300  301  141.56  145.30                  -40.90\n",
      "   16   5828  38A  39P C   N     303  305  -65.59  -58.70    7.15    0.99  -64.50  179.63   13.46\n",
      "   16         39P  39P N   CA    305  306  -32.43  -30.50                  147.20\n",
      "   17   5829  39P  40P C   N     310  312  -56.78  -58.70    2.25    0.20  -64.50  179.04   12.96\n",
      "   17         40P  40P N   CA    312  313  -31.67  -30.50                  147.20\n",
      "   18   5830  40P  41G C   N     317  319   65.83   78.70   25.99    0.39   82.20  152.90    8.24\n",
      "   18         41G  41G N   CA    319  320 -143.52 -166.10                    8.50\n",
      "   19   5831  41G  42A C   N     321  323 -106.76 -134.00   27.40    0.80  -62.50 -179.39   31.94\n",
      "   19         42A  42A N   CA    323  324  144.00  147.00                  -40.90\n",
      "   20   5832  42A  43A C   N     326  328  -71.23  -68.20    4.76    0.30  -62.50  170.35   28.35\n",
      "   20         43A  43A N   CA    328  329  148.97  145.30                  -40.90\n",
      "   21   5834  44P  45A C   N     338  340 -104.67 -134.00   31.17    0.72  -62.50 -177.73   28.11\n",
      "   21         45A  45A N   CA    340  341  136.42  147.00                  -40.90\n",
      "   22   5836  46P  47G C   N     350  352 -153.06 -167.20   19.21    0.41   82.20 -162.53   14.80\n",
      "   22         47G  47G N   CA    352  353  161.59  174.60                    8.50\n",
      "   23   5838  48I  49F C   N     362  364 -113.46 -124.20   10.75    0.43  -63.20  179.11   28.22\n",
      "   23         49F  49F N   CA    364  365  143.79  143.30                  -44.30\n",
      "   24   5839  49F  50S C   N     373  375 -130.43 -136.60    9.00    0.50  -64.10  179.91   17.73\n",
      "   24         50S  50S N   CA    375  376  157.76  151.20                  -35.00\n",
      "   25   5840  50S  51S C   N     379  381 -126.41 -136.60   12.43    0.67  -64.10  177.95   17.29\n",
      "   25         51S  51S N   CA    381  382  158.31  151.20                  -35.00\n",
      "   26   5841  51S  52Q C   N     385  387 -107.84 -121.10   13.28    0.48  -63.80 -175.39   25.05\n",
      "   26         52Q  52Q N   CA    387  388  138.98  139.70                  -40.30\n",
      "   27   5843  53P  54G C   N     401  403 -147.40 -167.20   29.96    0.72   82.20 -166.03   14.73\n",
      "   27         54G  54G N   CA    403  404  152.11  174.60                    8.50\n",
      "   28   5844  54G  55H C   N     405  407 -128.24 -125.60   18.36    0.65  -63.20  173.39   25.93\n",
      "   28         55H  55H N   CA    407  408  156.97  138.80                  -42.30\n",
      "   29   5845  55H  56T C   N     415  417 -114.85 -124.80   11.98    0.44  -63.20 -173.76   20.90\n",
      "   29         56T  56T N   CA    417  418  136.83  143.50                  -42.10\n",
      "   30   5847  57P  58H C   N     429  431  -99.78 -125.60   30.14    0.77  -63.20  169.54   19.19\n",
      "   30         58H  58H N   CA    431  432  123.25  138.80                  -42.30\n",
      "   31   5849  59P  60A C   N     446  448 -110.59 -134.00   24.70    0.58  -62.50 -173.70   33.08\n",
      "   31         60A  60A N   CA    448  449  139.11  147.00                  -40.90\n",
      "   32   5850  60A  61A C   N     451  453 -166.50 -134.00   41.38    1.09  -62.50  179.65   34.93\n",
      "   32         61A  61A N   CA    453  454  172.61  147.00                  -40.90\n",
      "   33   5851  61A  62S C   N     456  458  -61.95  -72.40   10.89    0.85  -64.10  169.53   12.16\n",
      "   33         62S  62S N   CA    458  459  155.49  152.40                  -35.00\n",
      "   34   5852  62S  63R C   N     462  464  -66.86  -72.10   34.01    2.50  -63.00  149.45   19.93\n",
      "   34         63R  63R N   CA    464  465  108.30  141.90                  -41.10\n",
      "   35   5853  63R  64D C   N     473  475 -122.39  -96.50   26.41    1.07  -63.30  160.31   17.18\n",
      "   35         64D  64D N   CA    475  476  109.02  114.20                  -40.00\n",
      "   36   5856  66V  67A C   N     495  497 -109.05 -134.00   27.24    0.62  -62.50 -177.02   28.09\n",
      "   36         67A  67A N   CA    497  498  136.06  147.00                  -40.90\n",
      "   37   5857  67A  68R C   N     500  502   68.79   57.30  110.94    8.15  -63.00 -144.45   24.64\n",
      "   37         68R  68R N   CA    502  503  148.34   38.00                  -41.10\n",
      "   38   5858  68R  69T C   N     511  513 -110.75 -124.80   21.00    1.25  -63.20  165.76   24.42\n",
      "   38         69T  69T N   CA    513  514  159.10  143.50                  -42.10\n",
      "   39   5859  69T  70S C   N     518  520  -76.38  -72.40   14.47    0.72  -64.10  159.16   12.30\n",
      "   39         70S  70S N   CA    520  521  166.31  152.40                  -35.00\n",
      "   40   5861  71P  72L C   N     531  533  -77.10  -70.70   14.79    0.93  -63.50  164.43   23.63\n",
      "   40         72L  72L N   CA    533  534  154.93  141.60                  -41.20\n",
      "   41   5862  72L  73Q C   N     539  541  -65.22  -73.00   19.21    1.12  -63.80  163.44   23.96\n",
      "   41         73Q  73Q N   CA    541  542  123.14  140.70                  -40.30\n",
      "   42   5863  73Q  74T C   N     548  550 -125.06 -124.80   23.65    1.02  -63.20  173.36   18.89\n",
      "   42         74T  74T N   CA    550  551  119.85  143.50                  -42.10\n",
      "   43   5865  75P  76A C   N     562  564 -104.20 -134.00   36.60    0.92  -62.50  171.78   26.44\n",
      "   43         76A  76A N   CA    564  565  125.75  147.00                  -40.90\n",
      "   44   5866  76A  77A C   N     567  569 -175.09 -134.00   43.30    1.02  -62.50 -165.61   37.79\n",
      "   44         77A  77A N   CA    569  570  160.64  147.00                  -40.90\n",
      "   45   5868  78P  79G C   N     579  581  -81.95  -80.20   23.78    1.02   82.20 -143.02    7.39\n",
      "   45         79G  79G N   CA    581  582  150.39  174.10                    8.50\n",
      "   46   5869  79G  80A C   N     583  585 -109.66 -134.00   32.70    0.92  -62.50  172.62   26.41\n",
      "   46         80A  80A N   CA    585  586  125.16  147.00                  -40.90\n",
      "   47   5870  80A  81A C   N     588  590 -159.26 -134.00   29.69    0.70  -62.50 -175.99   35.30\n",
      "   47         81A  81A N   CA    590  591  162.59  147.00                  -40.90\n",
      "   48   5871  81A  82A C   N     593  595  -81.73  -68.20   13.54    1.13  -62.50  174.41   29.57\n",
      "   48         82A  82A N   CA    595  596  145.75  145.30                  -40.90\n",
      "   49   5872  82A  83G C   N     598  600 -155.49 -167.20   15.65    0.33   82.20 -162.00   14.76\n",
      "   49         83G  83G N   CA    600  601  164.21  174.60                    8.50\n",
      "   50   5874  84P  85A C   N     609  611 -122.84 -134.00   17.40    1.04  -62.50  169.83   31.06\n",
      "   50         85A  85A N   CA    611  612  160.36  147.00                  -40.90\n",
      "   51   5875  85A  86L C   N     614  616  -97.43 -108.50   23.34    1.33  -63.50  169.19   25.56\n",
      "   51         86L  86L N   CA    616  617  153.05  132.50                  -41.20\n",
      "   52   5876  86L  87S C   N     622  624 -153.30 -136.60   32.44    1.21  -64.10  171.08   18.83\n",
      "   52         87S  87S N   CA    624  625  179.01  151.20                  -35.00\n",
      "   53   5880  90P  91P C   N     649  651  -76.78  -64.50   12.54    1.36  -58.70  176.09   13.14\n",
      "   53         91P  91P N   CA    651  652  144.66  147.20                  -30.50\n",
      "   54   5881  91P  92V C   N     656  658 -154.13  -62.40   92.91   13.98  -62.40   92.91   13.98\n",
      "   54         92V  92V N   CA    658  659  -27.62  -42.40                  -42.40\n",
      "\n",
      "\n",
      "report______> Distribution of short non-bonded contacts:\n",
      "\n",
      "  serious non-bonded atom clash:    90 1354   1.714\n",
      "  serious non-bonded atom clash:   122 1320   2.215\n",
      "  serious non-bonded atom clash:   144  727   2.264\n",
      "  serious non-bonded atom clash:   151 1196   1.945\n",
      "  serious non-bonded atom clash:   190 1197   1.758\n",
      "  serious non-bonded atom clash:   664  694   2.295\n",
      "  serious non-bonded atom clash:   671  702   1.565\n",
      "  serious non-bonded atom clash:   719 1120   2.189\n",
      "  serious non-bonded atom clash:   720 1119   1.741\n",
      "  serious non-bonded atom clash:   721 1119   1.648\n",
      "  serious non-bonded atom clash:   726 1152   1.768\n",
      "  serious non-bonded atom clash:   726 1154   2.269\n",
      "  serious non-bonded atom clash:   727 1152   2.294\n",
      "  serious non-bonded atom clash:   728 1152   2.162\n",
      "  serious non-bonded atom clash:   728 1154   1.683\n",
      "  serious non-bonded atom clash:   729 1152   2.041\n",
      "  serious non-bonded atom clash:   729 1153   2.155\n",
      "  serious non-bonded atom clash:   729 1154   1.124\n",
      "  serious non-bonded atom clash:   757  797   2.239\n",
      "  serious non-bonded atom clash:   757  798   2.053\n",
      "  serious non-bonded atom clash:   809  817   2.094\n",
      "  serious non-bonded atom clash:   823  827   2.273\n",
      "  serious non-bonded atom clash:   882 1210   2.153\n",
      "  serious non-bonded atom clash:  1000 1371   1.984\n",
      "  serious non-bonded atom clash:  1019 1041   1.587\n",
      "  serious non-bonded atom clash:  1019 1043   1.762\n",
      "  serious non-bonded atom clash:  1063 1067   2.231\n",
      "  serious non-bonded atom clash:  1111 1455   1.857\n",
      "  serious non-bonded atom clash:  1111 1457   1.212\n",
      "  serious non-bonded atom clash:  1111 1459   1.921\n",
      "  serious non-bonded atom clash:  1139 1383   2.285\n",
      "  serious non-bonded atom clash:  1141 1382   2.235\n",
      "  serious non-bonded atom clash:  1141 1383   2.023\n",
      "  serious non-bonded atom clash:  1168 1342   1.700\n",
      "  serious non-bonded atom clash:  1168 1343   2.160\n",
      "  serious non-bonded atom clash:  1169 1340   1.956\n",
      "  serious non-bonded atom clash:  1169 1342   2.045\n",
      "  serious non-bonded atom clash:  1169 1343   1.522\n",
      "  serious non-bonded atom clash:  1169 1345   2.168\n",
      "  serious non-bonded atom clash:  1170 1312   1.597\n",
      "  serious non-bonded atom clash:  1170 1339   2.166\n",
      "  serious non-bonded atom clash:  1170 1340   0.933\n",
      "  serious non-bonded atom clash:  1170 1342   2.001\n",
      "  serious non-bonded atom clash:  1170 1343   0.763\n",
      "  serious non-bonded atom clash:  1170 1344   1.750\n",
      "  serious non-bonded atom clash:  1170 1345   1.567\n",
      "  serious non-bonded atom clash:  1170 1346   2.186\n",
      "  serious non-bonded atom clash:  1170 1347   2.117\n",
      "  serious non-bonded atom clash:  1172 1310   2.198\n",
      "  serious non-bonded atom clash:  1172 1312   1.497\n",
      "  serious non-bonded atom clash:  1172 1313   2.013\n",
      "  serious non-bonded atom clash:  1172 1314   1.631\n",
      "  serious non-bonded atom clash:  1172 1340   1.411\n",
      "  serious non-bonded atom clash:  1172 1343   2.110\n",
      "  serious non-bonded atom clash:  1172 1344   0.608\n",
      "  serious non-bonded atom clash:  1172 1346   1.043\n",
      "  serious non-bonded atom clash:  1172 1347   1.878\n",
      "  serious non-bonded atom clash:  1174 1313   2.173\n",
      "  serious non-bonded atom clash:  1174 1315   2.177\n",
      "  serious non-bonded atom clash:  1174 1344   1.525\n",
      "  serious non-bonded atom clash:  1174 1346   1.595\n",
      "  serious non-bonded atom clash:  1195 1289   1.622\n",
      "  serious non-bonded atom clash:  1196 1289   1.533\n",
      "  serious non-bonded atom clash:  1197 1287   2.140\n",
      "  serious non-bonded atom clash:  1197 1288   1.553\n",
      "  serious non-bonded atom clash:  1197 1289   2.144\n",
      "  serious non-bonded atom clash:  1216 1262   2.299\n",
      "  serious non-bonded atom clash:  1218 1284   1.935\n",
      "  serious non-bonded atom clash:  1219 1262   2.021\n",
      "  serious non-bonded atom clash:  1219 1277   1.863\n",
      "  serious non-bonded atom clash:  1219 1278   2.125\n",
      "  serious non-bonded atom clash:  1219 1279   1.883\n",
      "  serious non-bonded atom clash:  1219 1280   2.163\n",
      "  serious non-bonded atom clash:  1262 1279   1.800\n",
      "  serious non-bonded atom clash:  1308 1346   1.922\n",
      "  serious non-bonded atom clash:  1309 1346   1.814\n",
      "  serious non-bonded atom clash:  1309 1347   1.856\n",
      "  serious non-bonded atom clash:  1310 1345   1.752\n",
      "  serious non-bonded atom clash:  1310 1346   1.549\n",
      "  serious non-bonded atom clash:  1310 1347   0.442\n",
      "  serious non-bonded atom clash:  1311 1345   1.843\n",
      "  serious non-bonded atom clash:  1311 1347   1.407\n",
      "  serious non-bonded atom clash:  1312 1340   2.114\n",
      "  serious non-bonded atom clash:  1312 1343   1.842\n",
      "  serious non-bonded atom clash:  1312 1344   1.964\n",
      "  serious non-bonded atom clash:  1312 1345   1.299\n",
      "  serious non-bonded atom clash:  1312 1346   1.407\n",
      "  serious non-bonded atom clash:  1312 1347   0.954\n",
      "  serious non-bonded atom clash:  1313 1344   1.825\n",
      "  serious non-bonded atom clash:  1313 1346   1.568\n",
      "  serious non-bonded atom clash:  1314 1344   1.208\n",
      "  serious non-bonded atom clash:  1314 1346   1.796\n",
      "  serious non-bonded atom clash:  1436 1474   1.768\n",
      "  serious non-bonded atom clash:  1436 1475   0.966\n",
      "  serious non-bonded atom clash:  1437 1475   2.218\n",
      "  serious non-bonded atom clash:  1438 1475   1.747\n",
      "  serious non-bonded atom clash:  1486 1489   2.181\n",
      "  serious non-bonded atom clash:  1486 1522   2.155\n",
      "  serious non-bonded atom clash:  1489 1522   2.077\n",
      "  serious non-bonded atom clash:  1490 1519   1.863\n",
      "  serious non-bonded atom clash:  1490 1520   2.124\n",
      "  serious non-bonded atom clash:  1491 1519   1.983\n",
      "  serious non-bonded atom clash:  1491 1520   1.662\n",
      "  serious non-bonded atom clash:  1491 1521   1.787\n",
      "  serious non-bonded atom clash:  1492 1494   2.068\n",
      "  serious non-bonded atom clash:  1564 1583   1.439\n",
      "  serious non-bonded atom clash:  1564 1585   2.034\n",
      "  serious non-bonded atom clash:  1575 1583   2.278\n",
      "  serious non-bonded atom clash:  1576 1583   2.243\n",
      "\n",
      "DISTANCE1:  0.00 2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40\n",
      "DISTANCE2:  2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40 3.50\n",
      "FREQUENCY:    74   21   14    7   24   32   38   76  111  140  149  211  221  211  283\n",
      "\n",
      "\n",
      "<< end of ENERGY.\n",
      "openf___224_> Open           Target.B99990006.pdb\n",
      "wrpdb___568_> Residues, atoms, selected atoms:      207     1613     1613\n",
      "openf___224_> Open           Target.ini\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      3936160    3843.906     3.754\n",
      "read_mo_297_> Segments, residues, atoms:        1      207     1613\n",
      "read_mo_298_> Segment:        1  1:A  207:A     1613\n",
      "randomi_498_> Atoms,selected atoms,random_seed,amplitude:     1613      457        1        4.0000\n",
      "randomi_496_> Amplitude is > 0; randomization is done.\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1927\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1933\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1938\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1956\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "iupac_m_487_> NH1/2 swapped:      179.8275        6  6\n",
      "iupac_m_486_> OE1/2 will be swapped:     -105.1969       13  13\n",
      "iupac_m_487_> NH1/2 swapped:      179.7925       26  26\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:      169.7476       28  28\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_485_> OD1/2 will be swapped:      164.4717       34  34\n",
      "iupac_m_487_> NH1/2 swapped:     -179.7259       63  63\n",
      "iupac_m_487_> NH1/2 swapped:      179.9441       98  98\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      103  103\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0163      104  104\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      107  107\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      109  109\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      110  110\n",
      "iupac_m_485_> OD1/2 will be swapped:     -179.9802      111  111\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      114  114\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0292      124  124\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9657      127  127\n",
      "iupac_m_487_> NH1/2 swapped:      179.9558      129  129\n",
      "iupac_m_486_> OE1/2 will be swapped:      154.4214      135  135\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      136  136\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0239      138  138\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9558      139  139\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9720      140  140\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9407      146  146\n",
      "iupac_m_486_> OE1/2 will be swapped:      179.9604      152  152\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      160  160\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      164  164\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      165  165\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      179  179\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:     -176.2274      180  180\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -180.0000      183  183\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      196  196\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      200  200\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0309      202  202\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      207  207\n",
      "\n",
      "\n",
      ">> ENERGY; Differences between the model's features and restraints:\n",
      "Number of all residues in MODEL                   :      207\n",
      "Number of all, selected real atoms                :     1613     457\n",
      "Number of all, selected pseudo atoms              :        0       0\n",
      "Number of all static, selected restraints         :    12693    1983\n",
      "COVALENT_CYS                                      :        F\n",
      "NONBONDED_SEL_ATOMS                               :        1\n",
      "Number of non-bonded pairs (excluding 1-2,1-3,1-4):      603\n",
      "Dynamic pairs routine                             : 5, NATM_SEL x NATM double loop\n",
      "Atomic shift for contacts update (UPDATE_DYNAMIC) :    0.390\n",
      "LENNARD_JONES_SWITCH                              :    6.500   7.500\n",
      "COULOMB_JONES_SWITCH                              :    6.500   7.500\n",
      "RESIDUE_SPAN_RANGE                                :        0   99999\n",
      "NLOGN_USE                                         :       15\n",
      "CONTACT_SHELL                                     :    4.000\n",
      "DYNAMIC_PAIRS,_SPHERE,_COULOMB,_LENNARD,_MODELLER :        T       T       F       F       F\n",
      "SPHERE_STDV                                       :    0.050\n",
      "RADII_FACTOR                                      :    0.820\n",
      "Current energy                                    :         277.1028\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Summary of the restraint violations: \n",
      "\n",
      "   NUM     ... number of restraints.\n",
      "   NUMVI   ... number of restraints with RVIOL > VIOL_REPORT_CUT[i].\n",
      "   RVIOL   ... relative difference from the best value.\n",
      "   NUMVP   ... number of restraints with -Ln(pdf) > VIOL_REPORT_CUT2[i].\n",
      "   RMS_1   ... RMS(feature, minimally_violated_basis_restraint, NUMB).\n",
      "   RMS_2   ... RMS(feature, best_value, NUMB).\n",
      "   MOL.PDF ... scaled contribution to -Ln(Molecular pdf).\n",
      "\n",
      " #                     RESTRAINT_GROUP      NUM   NUMVI  NUMVP   RMS_1   RMS_2         MOL.PDF     S_i\n",
      "------------------------------------------------------------------------------------------------------\n",
      " 1 Bond length potential              :     478       0      0   0.002   0.002     0.48695       1.000\n",
      " 2 Bond angle potential               :     667       0      0   1.987   1.987      51.334       1.000\n",
      " 3 Stereochemical cosine torsion poten:     343       0      4  51.953  51.953      122.29       1.000\n",
      " 4 Stereochemical improper torsion pot:     181       0      0   0.755   0.755      2.2836       1.000\n",
      " 5 Soft-sphere overlap restraints     :     603       0      0   0.002   0.002     0.18000       1.000\n",
      " 6 Lennard-Jones 6-12 potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 7 Coulomb point-point electrostatic p:       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 8 H-bonding potential                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 9 Distance restraints 1 (CA-CA)      :      29       0      0   0.011   0.011     0.63311E-03   1.000\n",
      "10 Distance restraints 2 (N-O)        :      30       0      0   0.584   0.584      1.2492       1.000\n",
      "11 Mainchain Phi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "12 Mainchain Psi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "13 Mainchain Omega dihedral restraints:      70       0      0   2.399   2.399      4.7501       1.000\n",
      "14 Sidechain Chi_1 dihedral restraints:      43       0      0  79.819  79.819      8.8677       1.000\n",
      "15 Sidechain Chi_2 dihedral restraints:      31       0      0  72.665  72.665      5.5533       1.000\n",
      "16 Sidechain Chi_3 dihedral restraints:      21       0      0  47.383  47.383      2.3389       1.000\n",
      "17 Sidechain Chi_4 dihedral restraints:       3       0      0 143.957 143.957      1.5977       1.000\n",
      "18 Disulfide distance restraints      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "19 Disulfide angle restraints         :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "20 Disulfide dihedral angle restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "21 Lower bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "22 Upper bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "23 Distance restraints 3 (SDCH-MNCH)  :      16       0      0   0.410   0.410     0.43199E-01   1.000\n",
      "24 Sidechain Chi_5 dihedral restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "25 Phi/Psi pair of dihedral restraints:      71      50     10  29.500 149.332      76.130       1.000\n",
      "26 Distance restraints 4 (SDCH-SDCH)  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "27 Distance restraints 5 (X-Y)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "28 NMR distance restraints 6 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "29 NMR distance restraints 7 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "30 Minimal distance restraints        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "31 Non-bonded restraints              :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "32 Atomic accessibility restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "33 Atomic density restraints          :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "34 Absolute position restraints       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "35 Dihedral angle difference restraint:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "36 GBSA implicit solvent potential    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "37 EM density fitting potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "38 SAXS restraints                    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "39 Symmetry restraints                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "\n",
      "\n",
      "\n",
      "openf___224_> Open           Target.V99990007\n",
      "# Heavy relative violation of each residue is written to: Target.V99990007\n",
      "# The profile is NOT normalized by the number of restraints.\n",
      "# The profiles are smoothed over a window of residues:    1\n",
      "# The sum of all numbers in the file:   12689.6143\n",
      "\n",
      "\n",
      "\n",
      "List of the violated restraints:\n",
      "   A restraint is violated when the relative difference\n",
      "   from the best value (RVIOL) is larger than CUTOFF.\n",
      "\n",
      "   ICSR   ... index of a restraint in the current set.\n",
      "   RESNO  ... residue numbers of the first two atoms.\n",
      "   ATM    ... IUPAC atom names of the first two atoms.\n",
      "   FEAT   ... the value of the feature in the model.\n",
      "   restr  ... the mean of the basis restraint with the smallest\n",
      "              difference from the model (local minimum).\n",
      "   viol   ... difference from the local minimum.\n",
      "   rviol  ... relative difference from the local minimum.\n",
      "   RESTR  ... the best value (global minimum).\n",
      "   VIOL   ... difference from the best value.\n",
      "   RVIOL  ... relative difference from the best value.\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 25                           : Phi/Psi pair of dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       6.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   5791   1M   2A C   N       7    9 -109.69 -134.00   24.46    0.71  -62.50 -178.90   32.18\n",
      "    1          2A   2A N   CA      9   10  144.26  147.00                  -40.90\n",
      "    2   5792   2A   3H C   N      12   14 -105.98 -125.60   19.62    0.69  -63.20 -176.25   25.76\n",
      "    2          3H   3H N   CA     14   15  139.00  138.80                  -42.30\n",
      "    3   5793   3H   4A C   N      22   24 -126.31 -134.00    8.73    0.45  -62.50  179.68   32.86\n",
      "    3          4A   4A N   CA     24   25  151.14  147.00                  -40.90\n",
      "    4   5794   4A   5G C   N      27   29 -174.82 -167.20    7.69    0.13   82.20 -163.72   14.07\n",
      "    4          5G   5G N   CA     29   30  175.60  174.60                    8.50\n",
      "    5   5795   5G   6R C   N      31   33 -117.27 -125.20   20.56    1.12  -63.00  168.32   26.48\n",
      "    5          6R   6R N   CA     33   34  159.57  140.60                  -41.10\n",
      "    6   5798   8G   9Y C   N      53   55 -122.84 -124.30   36.67    2.00  -63.50  156.26   27.75\n",
      "    6          9Y   9Y N   CA     55   56  172.05  135.40                  -43.40\n",
      "    7   5799   9Y  10D C   N      65   67 -102.68  -63.30   45.92    7.83  -63.30   45.92    7.83\n",
      "    7         10D  10D N   CA     67   68  -63.60  -40.00                  -40.00\n",
      "    8   5821  31D  32A C   N     265  267  -73.43  -68.20    5.84    0.40  -62.50  171.55   28.66\n",
      "    8         32A  32A N   CA    267  268  147.90  145.30                  -40.90\n",
      "    9   5822  32A  33G C   N     270  272 -144.58 -167.20   54.60    1.72   82.20  176.90   13.71\n",
      "    9         33G  33G N   CA    272  273  124.90  174.60                    8.50\n",
      "   10   5823  33G  34D C   N     274  276 -116.50  -96.50   20.00    0.82  -63.30  163.18   17.65\n",
      "   10         34D  34D N   CA    276  277  114.26  114.20                  -40.00\n",
      "   11   5825  35V  36G C   N     289  291  167.30 -167.20   26.98    0.75   82.20  178.83   12.51\n",
      "   11         36G  36G N   CA    291  292  165.78  174.60                    8.50\n",
      "   12   5826  36G  37A C   N     293  295 -128.49 -134.00    5.66    0.14  -62.50 -174.46   33.94\n",
      "   12         37A  37A N   CA    295  296  145.69  147.00                  -40.90\n",
      "   13   5827  37A  38A C   N     298  300 -132.87 -134.00    3.08    0.19  -62.50 -176.72   33.81\n",
      "   13         38A  38A N   CA    300  301  149.87  147.00                  -40.90\n",
      "   14   5829  39P  40P C   N     310  312  -55.61  -58.70    4.11    0.31  -64.50  179.81   13.89\n",
      "   14         40P  40P N   CA    312  313  -33.21  -30.50                  147.20\n",
      "   15   5830  40P  41G C   N     317  319   88.20   78.70   32.76    0.57   82.20  154.16    7.79\n",
      "   15         41G  41G N   CA    319  320  162.54 -166.10                    8.50\n",
      "   16   5831  41G  42A C   N     321  323 -129.36 -134.00    6.10    0.17  -62.50 -171.67   34.44\n",
      "   16         42A  42A N   CA    323  324  143.04  147.00                  -40.90\n",
      "   17   5832  42A  43A C   N     326  328 -141.65 -134.00    7.96    0.19  -62.50 -172.55   34.96\n",
      "   17         43A  43A N   CA    328  329  149.18  147.00                  -40.90\n",
      "   18   5834  44P  45A C   N     338  340 -122.55 -134.00   15.06    0.41  -62.50 -172.03   28.53\n",
      "   18         45A  45A N   CA    340  341  137.22  147.00                  -40.90\n",
      "   19   5836  46P  47G C   N     350  352 -155.89 -167.20   35.16    1.19   82.20 -179.73   13.71\n",
      "   19         47G  47G N   CA    352  353  141.30  174.60                    8.50\n",
      "   20   5838  48I  49F C   N     362  364 -113.36 -124.20   14.78    0.39  -63.20 -175.49   22.52\n",
      "   20         49F  49F N   CA    364  365  133.26  143.30                  -44.30\n",
      "   21   5839  49F  50S C   N     373  375 -141.10 -136.60    6.49    0.21  -64.10 -174.17   18.96\n",
      "   21         50S  50S N   CA    375  376  155.87  151.20                  -35.00\n",
      "   22   5840  50S  51S C   N     379  381 -133.26 -136.60    4.07    0.22  -64.10 -175.11   18.30\n",
      "   22         51S  51S N   CA    381  382  153.53  151.20                  -35.00\n",
      "   23   5841  51S  52Q C   N     385  387 -108.74 -121.10   19.06    0.64  -63.80  171.48   23.11\n",
      "   23         52Q  52Q N   CA    387  388  125.19  139.70                  -40.30\n",
      "   24   5843  53P  54G C   N     401  403  154.37 -167.20   42.57    1.30   82.20  164.48   11.28\n",
      "   24         54G  54G N   CA    403  404  156.30  174.60                    8.50\n",
      "   25   5844  54G  55H C   N     405  407 -133.77 -125.60    8.67    0.24  -63.20 -170.39   28.32\n",
      "   25         55H  55H N   CA    407  408  141.71  138.80                  -42.30\n",
      "   26   5845  55H  56T C   N     415  417 -125.04 -124.80    1.54    0.07  -63.20 -173.52   28.01\n",
      "   26         56T  56T N   CA    417  418  141.97  143.50                  -42.10\n",
      "   27   5847  57P  58H C   N     429  431 -137.98  -63.20   85.27   14.65  -63.20   85.27   14.65\n",
      "   27         58H  58H N   CA    431  432  -83.26  -42.30                  -42.30\n",
      "   28   5849  59P  60A C   N     446  448 -137.00 -134.00    5.74    0.35  -62.50 -167.97   35.46\n",
      "   28         60A  60A N   CA    448  449  142.11  147.00                  -40.90\n",
      "   29   5850  60A  61A C   N     451  453 -164.09 -134.00   33.95    0.78  -62.50 -173.50   35.95\n",
      "   29         61A  61A N   CA    453  454  162.71  147.00                  -40.90\n",
      "   30   5851  61A  62S C   N     456  458 -135.18 -136.60   18.30    0.96  -64.10  171.02   17.46\n",
      "   30         62S  62S N   CA    458  459  169.44  151.20                  -35.00\n",
      "   31   5852  62S  63R C   N     462  464  -81.61  -72.10   53.22    4.29  -63.00  131.95   16.65\n",
      "   31         63R  63R N   CA    464  465   89.54  141.90                  -41.10\n",
      "   32   5853  63R  64D C   N     473  475 -129.06  -96.50   52.61    2.21  -63.30  177.14   26.12\n",
      "   32         64D  64D N   CA    475  476  155.52  114.20                  -40.00\n",
      "   33   5856  66V  67A C   N     495  497 -116.35 -134.00   17.91    0.49  -62.50 -176.78   32.89\n",
      "   33         67A  67A N   CA    497  498  143.98  147.00                  -40.90\n",
      "   34   5857  67A  68R C   N     500  502 -122.96 -125.20    6.75    0.36  -63.00 -177.91   28.73\n",
      "   34         68R  68R N   CA    502  503  146.97  140.60                  -41.10\n",
      "   35   5858  68R  69T C   N     511  513  -81.45  -78.10    3.71    0.18  -63.20  167.51   22.69\n",
      "   35         69T  69T N   CA    513  514  151.38  149.80                  -42.10\n",
      "   36   5859  69T  70S C   N     518  520  -85.00  -72.40   18.88    0.88  -64.10  159.91   12.94\n",
      "   36         70S  70S N   CA    520  521  166.46  152.40                  -35.00\n",
      "   37   5861  71P  72L C   N     531  533 -106.50 -108.50    3.76    0.21  -63.50 -177.96   23.18\n",
      "   37         72L  72L N   CA    533  534  135.69  132.50                  -41.20\n",
      "   38   5862  72L  73Q C   N     539  541  -85.16  -73.00   19.87    1.14  -63.80  164.68   25.35\n",
      "   38         73Q  73Q N   CA    541  542  156.41  140.70                  -40.30\n",
      "   39   5863  73Q  74T C   N     548  550  -76.36  -78.10    6.45    0.33  -63.20  162.43   21.70\n",
      "   39         74T  74T N   CA    550  551  156.01  149.80                  -42.10\n",
      "   40   5865  75P  76A C   N     562  564  -89.81  -68.20   25.21    2.46  -62.50  175.36   27.53\n",
      "   40         76A  76A N   CA    564  565  132.32  145.30                  -40.90\n",
      "   41   5866  76A  77A C   N     567  569  -73.47  -68.20    9.24    0.59  -62.50  166.58   27.85\n",
      "   41         77A  77A N   CA    569  570  152.89  145.30                  -40.90\n",
      "   42   5868  78P  79G C   N     579  581  -79.83  -80.20   13.23    0.50   82.20 -137.58    7.20\n",
      "   42         79G  79G N   CA    581  582  160.87  174.10                    8.50\n",
      "   43   5869  79G  80A C   N     583  585  -73.69  -68.20    9.16    0.58  -62.50  166.85   27.91\n",
      "   43         80A  80A N   CA    585  586  152.63  145.30                  -40.90\n",
      "   44   5870  80A  81A C   N     588  590 -117.99 -134.00   16.42    0.42  -62.50 -175.70   33.16\n",
      "   44         81A  81A N   CA    590  591  143.35  147.00                  -40.90\n",
      "   45   5871  81A  82A C   N     593  595  -83.38  -68.20   23.36    1.47  -62.50  157.44   26.87\n",
      "   45         82A  82A N   CA    595  596  163.06  145.30                  -40.90\n",
      "   46   5872  82A  83G C   N     598  600 -140.28 -167.20   45.02    1.18   82.20 -170.75   14.58\n",
      "   46         83G  83G N   CA    600  601  138.51  174.60                    8.50\n",
      "   47   5874  84P  85A C   N     609  611 -128.69 -134.00    5.76    0.28  -62.50 -177.69   33.42\n",
      "   47         85A  85A N   CA    611  612  149.23  147.00                  -40.90\n",
      "   48   5875  85A  86L C   N     614  616 -110.16 -108.50   27.74    1.48  -63.50  165.33   25.83\n",
      "   48         86L  86L N   CA    616  617  160.19  132.50                  -41.20\n",
      "   49   5876  86L  87S C   N     622  624  -83.52  -72.40   19.54    1.50  -64.10  172.43   11.48\n",
      "   49         87S  87S N   CA    624  625  136.33  152.40                  -35.00\n",
      "   50   5877  87S  88P C   N     628  630  -72.92  -58.70   17.20    1.46  -64.50  168.23   13.00\n",
      "   50         88P  88P N   CA    630  631  -20.82  -30.50                  147.20\n",
      "\n",
      "\n",
      "report______> Distribution of short non-bonded contacts:\n",
      "\n",
      "  serious non-bonded atom clash:    90 1354   1.714\n",
      "  serious non-bonded atom clash:   122 1320   2.215\n",
      "  serious non-bonded atom clash:   144  727   2.264\n",
      "  serious non-bonded atom clash:   151 1196   1.945\n",
      "  serious non-bonded atom clash:   190 1197   1.758\n",
      "  serious non-bonded atom clash:   664  694   2.295\n",
      "  serious non-bonded atom clash:   671  702   1.565\n",
      "  serious non-bonded atom clash:   719 1120   2.189\n",
      "  serious non-bonded atom clash:   720 1119   1.741\n",
      "  serious non-bonded atom clash:   721 1119   1.648\n",
      "  serious non-bonded atom clash:   726 1152   1.768\n",
      "  serious non-bonded atom clash:   726 1154   2.269\n",
      "  serious non-bonded atom clash:   727 1152   2.294\n",
      "  serious non-bonded atom clash:   728 1152   2.162\n",
      "  serious non-bonded atom clash:   728 1154   1.683\n",
      "  serious non-bonded atom clash:   729 1152   2.041\n",
      "  serious non-bonded atom clash:   729 1153   2.155\n",
      "  serious non-bonded atom clash:   729 1154   1.124\n",
      "  serious non-bonded atom clash:   757  797   2.239\n",
      "  serious non-bonded atom clash:   757  798   2.053\n",
      "  serious non-bonded atom clash:   809  817   2.094\n",
      "  serious non-bonded atom clash:   823  827   2.273\n",
      "  serious non-bonded atom clash:   882 1210   2.153\n",
      "  serious non-bonded atom clash:  1000 1371   1.984\n",
      "  serious non-bonded atom clash:  1019 1041   1.587\n",
      "  serious non-bonded atom clash:  1019 1043   1.762\n",
      "  serious non-bonded atom clash:  1063 1067   2.231\n",
      "  serious non-bonded atom clash:  1111 1455   1.857\n",
      "  serious non-bonded atom clash:  1111 1457   1.212\n",
      "  serious non-bonded atom clash:  1111 1459   1.921\n",
      "  serious non-bonded atom clash:  1139 1383   2.285\n",
      "  serious non-bonded atom clash:  1141 1382   2.235\n",
      "  serious non-bonded atom clash:  1141 1383   2.023\n",
      "  serious non-bonded atom clash:  1168 1342   1.700\n",
      "  serious non-bonded atom clash:  1168 1343   2.160\n",
      "  serious non-bonded atom clash:  1169 1340   1.956\n",
      "  serious non-bonded atom clash:  1169 1342   2.045\n",
      "  serious non-bonded atom clash:  1169 1343   1.522\n",
      "  serious non-bonded atom clash:  1169 1345   2.168\n",
      "  serious non-bonded atom clash:  1170 1312   1.597\n",
      "  serious non-bonded atom clash:  1170 1339   2.166\n",
      "  serious non-bonded atom clash:  1170 1340   0.933\n",
      "  serious non-bonded atom clash:  1170 1342   2.001\n",
      "  serious non-bonded atom clash:  1170 1343   0.763\n",
      "  serious non-bonded atom clash:  1170 1344   1.750\n",
      "  serious non-bonded atom clash:  1170 1345   1.567\n",
      "  serious non-bonded atom clash:  1170 1346   2.186\n",
      "  serious non-bonded atom clash:  1170 1347   2.117\n",
      "  serious non-bonded atom clash:  1172 1310   2.198\n",
      "  serious non-bonded atom clash:  1172 1312   1.497\n",
      "  serious non-bonded atom clash:  1172 1313   2.013\n",
      "  serious non-bonded atom clash:  1172 1314   1.631\n",
      "  serious non-bonded atom clash:  1172 1340   1.411\n",
      "  serious non-bonded atom clash:  1172 1343   2.110\n",
      "  serious non-bonded atom clash:  1172 1344   0.608\n",
      "  serious non-bonded atom clash:  1172 1346   1.043\n",
      "  serious non-bonded atom clash:  1172 1347   1.878\n",
      "  serious non-bonded atom clash:  1174 1313   2.173\n",
      "  serious non-bonded atom clash:  1174 1315   2.177\n",
      "  serious non-bonded atom clash:  1174 1344   1.525\n",
      "  serious non-bonded atom clash:  1174 1346   1.595\n",
      "  serious non-bonded atom clash:  1195 1289   1.622\n",
      "  serious non-bonded atom clash:  1196 1289   1.533\n",
      "  serious non-bonded atom clash:  1197 1287   2.140\n",
      "  serious non-bonded atom clash:  1197 1288   1.553\n",
      "  serious non-bonded atom clash:  1197 1289   2.144\n",
      "  serious non-bonded atom clash:  1216 1262   2.299\n",
      "  serious non-bonded atom clash:  1218 1284   1.935\n",
      "  serious non-bonded atom clash:  1219 1262   2.021\n",
      "  serious non-bonded atom clash:  1219 1277   1.863\n",
      "  serious non-bonded atom clash:  1219 1278   2.125\n",
      "  serious non-bonded atom clash:  1219 1279   1.883\n",
      "  serious non-bonded atom clash:  1219 1280   2.163\n",
      "  serious non-bonded atom clash:  1262 1279   1.800\n",
      "  serious non-bonded atom clash:  1308 1346   1.922\n",
      "  serious non-bonded atom clash:  1309 1346   1.814\n",
      "  serious non-bonded atom clash:  1309 1347   1.856\n",
      "  serious non-bonded atom clash:  1310 1345   1.752\n",
      "  serious non-bonded atom clash:  1310 1346   1.549\n",
      "  serious non-bonded atom clash:  1310 1347   0.442\n",
      "  serious non-bonded atom clash:  1311 1345   1.843\n",
      "  serious non-bonded atom clash:  1311 1347   1.407\n",
      "  serious non-bonded atom clash:  1312 1340   2.114\n",
      "  serious non-bonded atom clash:  1312 1343   1.842\n",
      "  serious non-bonded atom clash:  1312 1344   1.964\n",
      "  serious non-bonded atom clash:  1312 1345   1.299\n",
      "  serious non-bonded atom clash:  1312 1346   1.407\n",
      "  serious non-bonded atom clash:  1312 1347   0.954\n",
      "  serious non-bonded atom clash:  1313 1344   1.825\n",
      "  serious non-bonded atom clash:  1313 1346   1.568\n",
      "  serious non-bonded atom clash:  1314 1344   1.208\n",
      "  serious non-bonded atom clash:  1314 1346   1.796\n",
      "  serious non-bonded atom clash:  1436 1474   1.768\n",
      "  serious non-bonded atom clash:  1436 1475   0.966\n",
      "  serious non-bonded atom clash:  1437 1475   2.218\n",
      "  serious non-bonded atom clash:  1438 1475   1.747\n",
      "  serious non-bonded atom clash:  1486 1489   2.181\n",
      "  serious non-bonded atom clash:  1486 1522   2.155\n",
      "  serious non-bonded atom clash:  1489 1522   2.077\n",
      "  serious non-bonded atom clash:  1490 1519   1.863\n",
      "  serious non-bonded atom clash:  1490 1520   2.124\n",
      "  serious non-bonded atom clash:  1491 1519   1.983\n",
      "  serious non-bonded atom clash:  1491 1520   1.662\n",
      "  serious non-bonded atom clash:  1491 1521   1.787\n",
      "  serious non-bonded atom clash:  1492 1494   2.068\n",
      "  serious non-bonded atom clash:  1564 1583   1.439\n",
      "  serious non-bonded atom clash:  1564 1585   2.034\n",
      "  serious non-bonded atom clash:  1575 1583   2.278\n",
      "  serious non-bonded atom clash:  1576 1583   2.243\n",
      "\n",
      "DISTANCE1:  0.00 2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40\n",
      "DISTANCE2:  2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40 3.50\n",
      "FREQUENCY:    74   21   14    7   24   33   40   74  110  139  144  227  214  214  265\n",
      "\n",
      "\n",
      "<< end of ENERGY.\n",
      "openf___224_> Open           Target.B99990007.pdb\n",
      "wrpdb___568_> Residues, atoms, selected atoms:      207     1613     1613\n",
      "openf___224_> Open           Target.ini\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      3936160    3843.906     3.754\n",
      "read_mo_297_> Segments, residues, atoms:        1      207     1613\n",
      "read_mo_298_> Segment:        1  1:A  207:A     1613\n",
      "randomi_498_> Atoms,selected atoms,random_seed,amplitude:     1613      457        1        4.0000\n",
      "randomi_496_> Amplitude is > 0; randomization is done.\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1927\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1933\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1938\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1956\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "iupac_m_487_> NH1/2 swapped:     -179.1834        6  6\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:      102.0474        9  9\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_486_> OE1/2 will be swapped:     -105.1969       13  13\n",
      "iupac_m_487_> NH1/2 swapped:      179.7925       26  26\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:      169.7476       28  28\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_485_> OD1/2 will be swapped:      174.3501       34  34\n",
      "iupac_m_487_> NH1/2 swapped:      179.9441       98  98\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      103  103\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0163      104  104\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      107  107\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      109  109\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      110  110\n",
      "iupac_m_485_> OD1/2 will be swapped:     -179.9802      111  111\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      114  114\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0292      124  124\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9657      127  127\n",
      "iupac_m_487_> NH1/2 swapped:      179.9558      129  129\n",
      "iupac_m_486_> OE1/2 will be swapped:      154.4214      135  135\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      136  136\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0239      138  138\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9558      139  139\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9720      140  140\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9407      146  146\n",
      "iupac_m_486_> OE1/2 will be swapped:      179.9604      152  152\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      160  160\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      164  164\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      165  165\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      179  179\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:     -176.2274      180  180\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -180.0000      183  183\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      196  196\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      200  200\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0309      202  202\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      207  207\n",
      "\n",
      "\n",
      ">> ENERGY; Differences between the model's features and restraints:\n",
      "Number of all residues in MODEL                   :      207\n",
      "Number of all, selected real atoms                :     1613     457\n",
      "Number of all, selected pseudo atoms              :        0       0\n",
      "Number of all static, selected restraints         :    12693    1983\n",
      "COVALENT_CYS                                      :        F\n",
      "NONBONDED_SEL_ATOMS                               :        1\n",
      "Number of non-bonded pairs (excluding 1-2,1-3,1-4):      592\n",
      "Dynamic pairs routine                             : 5, NATM_SEL x NATM double loop\n",
      "Atomic shift for contacts update (UPDATE_DYNAMIC) :    0.390\n",
      "LENNARD_JONES_SWITCH                              :    6.500   7.500\n",
      "COULOMB_JONES_SWITCH                              :    6.500   7.500\n",
      "RESIDUE_SPAN_RANGE                                :        0   99999\n",
      "NLOGN_USE                                         :       15\n",
      "CONTACT_SHELL                                     :    4.000\n",
      "DYNAMIC_PAIRS,_SPHERE,_COULOMB,_LENNARD,_MODELLER :        T       T       F       F       F\n",
      "SPHERE_STDV                                       :    0.050\n",
      "RADII_FACTOR                                      :    0.820\n",
      "Current energy                                    :         296.2316\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Summary of the restraint violations: \n",
      "\n",
      "   NUM     ... number of restraints.\n",
      "   NUMVI   ... number of restraints with RVIOL > VIOL_REPORT_CUT[i].\n",
      "   RVIOL   ... relative difference from the best value.\n",
      "   NUMVP   ... number of restraints with -Ln(pdf) > VIOL_REPORT_CUT2[i].\n",
      "   RMS_1   ... RMS(feature, minimally_violated_basis_restraint, NUMB).\n",
      "   RMS_2   ... RMS(feature, best_value, NUMB).\n",
      "   MOL.PDF ... scaled contribution to -Ln(Molecular pdf).\n",
      "\n",
      " #                     RESTRAINT_GROUP      NUM   NUMVI  NUMVP   RMS_1   RMS_2         MOL.PDF     S_i\n",
      "------------------------------------------------------------------------------------------------------\n",
      " 1 Bond length potential              :     478       0      0   0.003   0.003      1.0310       1.000\n",
      " 2 Bond angle potential               :     667       0      1   2.256   2.256      63.856       1.000\n",
      " 3 Stereochemical cosine torsion poten:     343       0      7  53.060  53.060      126.10       1.000\n",
      " 4 Stereochemical improper torsion pot:     181       0      0   0.888   0.888      3.1497       1.000\n",
      " 5 Soft-sphere overlap restraints     :     592       0      0   0.002   0.002     0.33351       1.000\n",
      " 6 Lennard-Jones 6-12 potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 7 Coulomb point-point electrostatic p:       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 8 H-bonding potential                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 9 Distance restraints 1 (CA-CA)      :      29       0      0   0.412   0.412     0.43132       1.000\n",
      "10 Distance restraints 2 (N-O)        :      30       0      0   0.767   0.767      1.6836       1.000\n",
      "11 Mainchain Phi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "12 Mainchain Psi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "13 Mainchain Omega dihedral restraints:      70       0      0   2.940   2.940      7.1370       1.000\n",
      "14 Sidechain Chi_1 dihedral restraints:      43       0      0  75.630  75.630      9.8911       1.000\n",
      "15 Sidechain Chi_2 dihedral restraints:      31       0      0  73.134  73.134      5.6513       1.000\n",
      "16 Sidechain Chi_3 dihedral restraints:      21       0      0  45.682  45.682      3.8909       1.000\n",
      "17 Sidechain Chi_4 dihedral restraints:       3       0      0  75.264  75.264      2.1573       1.000\n",
      "18 Disulfide distance restraints      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "19 Disulfide angle restraints         :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "20 Disulfide dihedral angle restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "21 Lower bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "22 Upper bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "23 Distance restraints 3 (SDCH-MNCH)  :      16       0      0   0.692   0.692     0.11699       1.000\n",
      "24 Sidechain Chi_5 dihedral restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "25 Phi/Psi pair of dihedral restraints:      71      49      5  29.194 147.231      70.803       1.000\n",
      "26 Distance restraints 4 (SDCH-SDCH)  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "27 Distance restraints 5 (X-Y)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "28 NMR distance restraints 6 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "29 NMR distance restraints 7 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "30 Minimal distance restraints        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "31 Non-bonded restraints              :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "32 Atomic accessibility restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "33 Atomic density restraints          :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "34 Absolute position restraints       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "35 Dihedral angle difference restraint:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "36 GBSA implicit solvent potential    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "37 EM density fitting potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "38 SAXS restraints                    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "39 Symmetry restraints                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "\n",
      "\n",
      "\n",
      "openf___224_> Open           Target.V99990008\n",
      "# Heavy relative violation of each residue is written to: Target.V99990008\n",
      "# The profile is NOT normalized by the number of restraints.\n",
      "# The profiles are smoothed over a window of residues:    1\n",
      "# The sum of all numbers in the file:   12646.1709\n",
      "\n",
      "\n",
      "\n",
      "List of the violated restraints:\n",
      "   A restraint is violated when the relative difference\n",
      "   from the best value (RVIOL) is larger than CUTOFF.\n",
      "\n",
      "   ICSR   ... index of a restraint in the current set.\n",
      "   RESNO  ... residue numbers of the first two atoms.\n",
      "   ATM    ... IUPAC atom names of the first two atoms.\n",
      "   FEAT   ... the value of the feature in the model.\n",
      "   restr  ... the mean of the basis restraint with the smallest\n",
      "              difference from the model (local minimum).\n",
      "   viol   ... difference from the local minimum.\n",
      "   rviol  ... relative difference from the local minimum.\n",
      "   RESTR  ... the best value (global minimum).\n",
      "   VIOL   ... difference from the best value.\n",
      "   RVIOL  ... relative difference from the best value.\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 25                           : Phi/Psi pair of dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       6.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   5791   1M   2A C   N       7    9  -70.60  -68.20    5.68    0.38  -62.50  168.84   28.07\n",
      "    1          2A   2A N   CA      9   10  150.45  145.30                  -40.90\n",
      "    2   5792   2A   3H C   N      12   14 -107.74 -125.60   17.95    0.59  -63.20 -175.21   20.71\n",
      "    2          3H   3H N   CA     14   15  137.04  138.80                  -42.30\n",
      "    3   5793   3H   4A C   N      22   24 -119.97 -134.00   14.15    0.40  -62.50 -176.84   33.08\n",
      "    3          4A   4A N   CA     24   25  145.19  147.00                  -40.90\n",
      "    4   5794   4A   5G C   N      27   29 -146.53 -167.20   28.03    0.59   82.20 -162.80   14.94\n",
      "    4          5G   5G N   CA     29   30  155.67  174.60                    8.50\n",
      "    5   5795   5G   6R C   N      31   33  -77.74  -72.10    7.64    0.50  -63.00  172.48   24.30\n",
      "    5          6R   6R N   CA     33   34  147.06  141.90                  -41.10\n",
      "    6   5796   6R   7T C   N      42   44 -116.63 -124.80   13.24    0.78  -63.20  172.47   25.67\n",
      "    6          7T   7T N   CA     44   45  153.91  143.50                  -42.10\n",
      "    7   5798   8G   9Y C   N      53   55  101.70 -124.30  134.26   10.33  -63.50 -122.57   44.74\n",
      "    7          9Y   9Y N   CA     55   56  127.13  135.40                  -43.40\n",
      "    8   5799   9Y  10D C   N      65   67 -128.59  -63.30   69.43   11.78  -63.30   69.43   11.78\n",
      "    8         10D  10D N   CA     67   68  -63.60  -40.00                  -40.00\n",
      "    9   5821  31D  32A C   N     265  267 -104.87 -134.00   29.13    0.97  -62.50  177.01   31.24\n",
      "    9         32A  32A N   CA    267  268  147.24  147.00                  -40.90\n",
      "   10   5822  32A  33G C   N     270  272 -109.35  -80.20   58.56    3.69   82.20 -156.15   15.93\n",
      "   10         33G  33G N   CA    272  273  123.31  174.10                    8.50\n",
      "   11   5823  33G  34D C   N     274  276 -127.36  -96.50   38.20    1.56  -63.30  146.45   15.55\n",
      "   11         34D  34D N   CA    276  277   91.70  114.20                  -40.00\n",
      "   12   5825  35V  36G C   N     289  291 -170.08 -167.20    6.71    0.27   82.20 -167.09   14.04\n",
      "   12         36G  36G N   CA    291  292  168.54  174.60                    8.50\n",
      "   13   5826  36G  37A C   N     293  295 -138.80 -134.00   16.75    0.81  -62.50  173.71   32.55\n",
      "   13         37A  37A N   CA    295  296  163.04  147.00                  -40.90\n",
      "   14   5827  37A  38A C   N     298  300  -62.87  -68.20    5.39    0.48  -62.50  172.97   28.36\n",
      "   14         38A  38A N   CA    300  301  146.14  145.30                  -40.90\n",
      "   15   5830  40P  41G C   N     317  319  160.89 -167.20   34.22    0.98   82.20  172.70   11.95\n",
      "   15         41G  41G N   CA    319  320  162.23  174.60                    8.50\n",
      "   16   5831  41G  42A C   N     321  323 -130.90 -134.00    4.23    0.12  -62.50 -172.12   34.45\n",
      "   16         42A  42A N   CA    323  324  144.12  147.00                  -40.90\n",
      "   17   5832  42A  43A C   N     326  328 -136.33 -134.00    2.34    0.07  -62.50 -172.92   34.61\n",
      "   17         43A  43A N   CA    328  329  147.20  147.00                  -40.90\n",
      "   18   5834  44P  45A C   N     338  340 -123.16 -134.00   14.59    0.41  -62.50 -171.82   28.55\n",
      "   18         45A  45A N   CA    340  341  137.24  147.00                  -40.90\n",
      "   19   5836  46P  47G C   N     350  352  160.20 -167.20   32.80    0.55   82.20 -173.25   12.65\n",
      "   19         47G  47G N   CA    352  353  178.19  174.60                    8.50\n",
      "   20   5838  48I  49F C   N     362  364 -104.64 -124.20   19.62    0.81  -63.20  175.89   27.21\n",
      "   20         49F  49F N   CA    364  365  144.76  143.30                  -44.30\n",
      "   21   5839  49F  50S C   N     373  375 -133.47 -136.60   12.50    0.69  -64.10  175.95   17.68\n",
      "   21         50S  50S N   CA    375  376  163.30  151.20                  -35.00\n",
      "   22   5840  50S  51S C   N     379  381 -129.87 -136.60   15.14    0.86  -64.10  173.21   17.21\n",
      "   22         51S  51S N   CA    381  382  164.75  151.20                  -35.00\n",
      "   23   5841  51S  52Q C   N     385  387  -77.78  -73.00    5.16    0.40  -63.80  179.60   25.69\n",
      "   23         52Q  52Q N   CA    387  388  138.76  140.70                  -40.30\n",
      "   24   5843  53P  54G C   N     401  403 -138.81 -167.20   40.22    0.90   82.20 -164.42   15.01\n",
      "   24         54G  54G N   CA    403  404  146.10  174.60                    8.50\n",
      "   25   5844  54G  55H C   N     405  407 -124.88 -125.60   33.30    1.28  -63.20  158.14   23.80\n",
      "   25         55H  55H N   CA    407  408  172.09  138.80                  -42.30\n",
      "   26   5845  55H  56T C   N     415  417 -104.75 -124.80   26.01    0.89  -63.20  174.07   19.86\n",
      "   26         56T  56T N   CA    417  418  126.94  143.50                  -42.10\n",
      "   27   5847  57P  58H C   N     429  431 -109.76 -125.60   50.35    1.59  -63.20  141.20   15.37\n",
      "   27         58H  58H N   CA    431  432   91.00  138.80                  -42.30\n",
      "   28   5849  59P  60A C   N     446  448  -84.82  -68.20   19.93    1.31  -62.50  164.32   28.08\n",
      "   28         60A  60A N   CA    448  449  156.30  145.30                  -40.90\n",
      "   29   5850  60A  61A C   N     451  453  -64.37  -68.20    5.45    0.34  -62.50  177.69   29.21\n",
      "   29         61A  61A N   CA    453  454  141.42  145.30                  -40.90\n",
      "   30   5851  61A  62S C   N     456  458 -149.10 -136.60   12.66    0.56  -64.10 -164.73   20.23\n",
      "   30         62S  62S N   CA    458  459  149.19  151.20                  -35.00\n",
      "   31   5853  63R  64D C   N     473  475  -72.19  -70.90   13.24    0.72  -63.30  177.35   21.37\n",
      "   31         64D  64D N   CA    475  476  137.12  150.30                  -40.00\n",
      "   32   5856  66V  67A C   N     495  497 -166.42 -134.00   39.07    0.96  -62.50 -177.26   35.44\n",
      "   32         67A  67A N   CA    497  498  168.80  147.00                  -40.90\n",
      "   33   5857  67A  68R C   N     500  502   52.43   57.30   11.80    0.57  -63.00  146.27   26.55\n",
      "   33         68R  68R N   CA    502  503   48.74   38.00                  -41.10\n",
      "   34   5858  68R  69T C   N     511  513  -93.62  -78.10   46.03    2.49  -63.20  151.64   17.60\n",
      "   34         69T  69T N   CA    513  514  106.46  149.80                  -42.10\n",
      "   35   5859  69T  70S C   N     518  520  -71.40  -72.40   11.50    0.64  -64.10  176.10   12.34\n",
      "   35         70S  70S N   CA    520  521  140.95  152.40                  -35.00\n",
      "   36   5861  71P  72L C   N     531  533 -113.62 -108.50    5.16    0.25  -63.50 -179.88   22.67\n",
      "   36         72L  72L N   CA    533  534  131.81  132.50                  -41.20\n",
      "   37   5862  72L  73Q C   N     539  541 -138.89 -121.10   34.48    1.27  -63.80  168.16   28.63\n",
      "   37         73Q  73Q N   CA    541  542  169.23  139.70                  -40.30\n",
      "   38   5863  73Q  74T C   N     548  550  -70.74  -78.10    7.61    0.41  -63.20  170.23   22.33\n",
      "   38         74T  74T N   CA    550  551  147.84  149.80                  -42.10\n",
      "   39   5865  75P  76A C   N     562  564 -129.95 -134.00   19.39    1.17  -62.50  167.34   31.04\n",
      "   39         76A  76A N   CA    564  565  165.96  147.00                  -40.90\n",
      "   40   5866  76A  77A C   N     567  569  -59.51  -68.20   12.42    1.22  -62.50  164.95   26.89\n",
      "   40         77A  77A N   CA    569  570  154.18  145.30                  -40.90\n",
      "   41   5868  78P  79G C   N     579  581  -81.34  -80.20   14.67    0.64   82.20 -137.42    7.29\n",
      "   41         79G  79G N   CA    581  582  159.48  174.10                    8.50\n",
      "   42   5869  79G  80A C   N     583  585 -126.04 -134.00   11.85    0.70  -62.50  175.26   32.12\n",
      "   42         80A  80A N   CA    585  586  155.77  147.00                  -40.90\n",
      "   43   5870  80A  81A C   N     588  590  -63.78  -68.20   12.99    1.19  -62.50  161.59   26.54\n",
      "   43         81A  81A N   CA    590  591  157.51  145.30                  -40.90\n",
      "   44   5871  81A  82A C   N     593  595  -65.99  -68.20    5.17    0.34  -62.50  178.51   29.43\n",
      "   44         82A  82A N   CA    595  596  140.63  145.30                  -40.90\n",
      "   45   5872  82A  83G C   N     598  600  175.85 -167.20   47.52    1.95   82.20  153.56   11.41\n",
      "   45         83G  83G N   CA    600  601  130.20  174.60                    8.50\n",
      "   46   5874  84P  85A C   N     609  611  -96.05  -68.20   33.62    2.21  -62.50  158.56   27.75\n",
      "   46         85A  85A N   CA    611  612  164.13  145.30                  -40.90\n",
      "   47   5875  85A  86L C   N     614  616 -121.29 -108.50   23.99    1.13  -63.50  175.77   27.98\n",
      "   47         86L  86L N   CA    616  617  152.80  132.50                  -41.20\n",
      "   48   5876  86L  87S C   N     622  624  -86.11  -72.40   16.92    0.86  -64.10  164.16   13.33\n",
      "   48         87S  87S N   CA    624  625  162.32  152.40                  -35.00\n",
      "   49   5880  90P  91P C   N     649  651  -79.28  -64.50   36.01    2.11  -58.70  150.88   14.17\n",
      "   49         91P  91P N   CA    651  652 -179.97  147.20                  -30.50\n",
      "\n",
      "\n",
      "report______> Distribution of short non-bonded contacts:\n",
      "\n",
      "  serious non-bonded atom clash:    90 1354   1.714\n",
      "  serious non-bonded atom clash:   122 1320   2.215\n",
      "  serious non-bonded atom clash:   144  727   2.264\n",
      "  serious non-bonded atom clash:   151 1196   1.945\n",
      "  serious non-bonded atom clash:   190 1197   1.758\n",
      "  serious non-bonded atom clash:   664  694   2.295\n",
      "  serious non-bonded atom clash:   671  702   1.565\n",
      "  serious non-bonded atom clash:   719 1120   2.189\n",
      "  serious non-bonded atom clash:   720 1119   1.741\n",
      "  serious non-bonded atom clash:   721 1119   1.648\n",
      "  serious non-bonded atom clash:   726 1152   1.768\n",
      "  serious non-bonded atom clash:   726 1154   2.269\n",
      "  serious non-bonded atom clash:   727 1152   2.294\n",
      "  serious non-bonded atom clash:   728 1152   2.162\n",
      "  serious non-bonded atom clash:   728 1154   1.683\n",
      "  serious non-bonded atom clash:   729 1152   2.041\n",
      "  serious non-bonded atom clash:   729 1153   2.155\n",
      "  serious non-bonded atom clash:   729 1154   1.124\n",
      "  serious non-bonded atom clash:   757  797   2.239\n",
      "  serious non-bonded atom clash:   757  798   2.053\n",
      "  serious non-bonded atom clash:   809  817   2.094\n",
      "  serious non-bonded atom clash:   823  827   2.273\n",
      "  serious non-bonded atom clash:   882 1210   2.153\n",
      "  serious non-bonded atom clash:  1000 1371   1.984\n",
      "  serious non-bonded atom clash:  1019 1041   1.587\n",
      "  serious non-bonded atom clash:  1019 1043   1.762\n",
      "  serious non-bonded atom clash:  1063 1067   2.231\n",
      "  serious non-bonded atom clash:  1111 1455   1.857\n",
      "  serious non-bonded atom clash:  1111 1457   1.212\n",
      "  serious non-bonded atom clash:  1111 1459   1.921\n",
      "  serious non-bonded atom clash:  1139 1383   2.285\n",
      "  serious non-bonded atom clash:  1141 1382   2.235\n",
      "  serious non-bonded atom clash:  1141 1383   2.023\n",
      "  serious non-bonded atom clash:  1168 1342   1.700\n",
      "  serious non-bonded atom clash:  1168 1343   2.160\n",
      "  serious non-bonded atom clash:  1169 1340   1.956\n",
      "  serious non-bonded atom clash:  1169 1342   2.045\n",
      "  serious non-bonded atom clash:  1169 1343   1.522\n",
      "  serious non-bonded atom clash:  1169 1345   2.168\n",
      "  serious non-bonded atom clash:  1170 1312   1.597\n",
      "  serious non-bonded atom clash:  1170 1339   2.166\n",
      "  serious non-bonded atom clash:  1170 1340   0.933\n",
      "  serious non-bonded atom clash:  1170 1342   2.001\n",
      "  serious non-bonded atom clash:  1170 1343   0.763\n",
      "  serious non-bonded atom clash:  1170 1344   1.750\n",
      "  serious non-bonded atom clash:  1170 1345   1.567\n",
      "  serious non-bonded atom clash:  1170 1346   2.186\n",
      "  serious non-bonded atom clash:  1170 1347   2.117\n",
      "  serious non-bonded atom clash:  1172 1310   2.198\n",
      "  serious non-bonded atom clash:  1172 1312   1.497\n",
      "  serious non-bonded atom clash:  1172 1313   2.013\n",
      "  serious non-bonded atom clash:  1172 1314   1.631\n",
      "  serious non-bonded atom clash:  1172 1340   1.411\n",
      "  serious non-bonded atom clash:  1172 1343   2.110\n",
      "  serious non-bonded atom clash:  1172 1344   0.608\n",
      "  serious non-bonded atom clash:  1172 1346   1.043\n",
      "  serious non-bonded atom clash:  1172 1347   1.878\n",
      "  serious non-bonded atom clash:  1174 1313   2.173\n",
      "  serious non-bonded atom clash:  1174 1315   2.177\n",
      "  serious non-bonded atom clash:  1174 1344   1.525\n",
      "  serious non-bonded atom clash:  1174 1346   1.595\n",
      "  serious non-bonded atom clash:  1195 1289   1.622\n",
      "  serious non-bonded atom clash:  1196 1289   1.533\n",
      "  serious non-bonded atom clash:  1197 1287   2.140\n",
      "  serious non-bonded atom clash:  1197 1288   1.553\n",
      "  serious non-bonded atom clash:  1197 1289   2.144\n",
      "  serious non-bonded atom clash:  1216 1262   2.299\n",
      "  serious non-bonded atom clash:  1218 1284   1.935\n",
      "  serious non-bonded atom clash:  1219 1262   2.021\n",
      "  serious non-bonded atom clash:  1219 1277   1.863\n",
      "  serious non-bonded atom clash:  1219 1278   2.125\n",
      "  serious non-bonded atom clash:  1219 1279   1.883\n",
      "  serious non-bonded atom clash:  1219 1280   2.163\n",
      "  serious non-bonded atom clash:  1262 1279   1.800\n",
      "  serious non-bonded atom clash:  1308 1346   1.922\n",
      "  serious non-bonded atom clash:  1309 1346   1.814\n",
      "  serious non-bonded atom clash:  1309 1347   1.856\n",
      "  serious non-bonded atom clash:  1310 1345   1.752\n",
      "  serious non-bonded atom clash:  1310 1346   1.549\n",
      "  serious non-bonded atom clash:  1310 1347   0.442\n",
      "  serious non-bonded atom clash:  1311 1345   1.843\n",
      "  serious non-bonded atom clash:  1311 1347   1.407\n",
      "  serious non-bonded atom clash:  1312 1340   2.114\n",
      "  serious non-bonded atom clash:  1312 1343   1.842\n",
      "  serious non-bonded atom clash:  1312 1344   1.964\n",
      "  serious non-bonded atom clash:  1312 1345   1.299\n",
      "  serious non-bonded atom clash:  1312 1346   1.407\n",
      "  serious non-bonded atom clash:  1312 1347   0.954\n",
      "  serious non-bonded atom clash:  1313 1344   1.825\n",
      "  serious non-bonded atom clash:  1313 1346   1.568\n",
      "  serious non-bonded atom clash:  1314 1344   1.208\n",
      "  serious non-bonded atom clash:  1314 1346   1.796\n",
      "  serious non-bonded atom clash:  1436 1474   1.768\n",
      "  serious non-bonded atom clash:  1436 1475   0.966\n",
      "  serious non-bonded atom clash:  1437 1475   2.218\n",
      "  serious non-bonded atom clash:  1438 1475   1.747\n",
      "  serious non-bonded atom clash:  1486 1489   2.181\n",
      "  serious non-bonded atom clash:  1486 1522   2.155\n",
      "  serious non-bonded atom clash:  1489 1522   2.077\n",
      "  serious non-bonded atom clash:  1490 1519   1.863\n",
      "  serious non-bonded atom clash:  1490 1520   2.124\n",
      "  serious non-bonded atom clash:  1491 1519   1.983\n",
      "  serious non-bonded atom clash:  1491 1520   1.662\n",
      "  serious non-bonded atom clash:  1491 1521   1.787\n",
      "  serious non-bonded atom clash:  1492 1494   2.068\n",
      "  serious non-bonded atom clash:  1564 1583   1.439\n",
      "  serious non-bonded atom clash:  1564 1585   2.034\n",
      "  serious non-bonded atom clash:  1575 1583   2.278\n",
      "  serious non-bonded atom clash:  1576 1583   2.243\n",
      "\n",
      "DISTANCE1:  0.00 2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40\n",
      "DISTANCE2:  2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40 3.50\n",
      "FREQUENCY:    74   21   14    7   24   33   38   77  110  142  153  209  225  219  262\n",
      "\n",
      "\n",
      "<< end of ENERGY.\n",
      "openf___224_> Open           Target.B99990008.pdb\n",
      "wrpdb___568_> Residues, atoms, selected atoms:      207     1613     1613\n",
      "openf___224_> Open           Target.ini\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      3936160    3843.906     3.754\n",
      "read_mo_297_> Segments, residues, atoms:        1      207     1613\n",
      "read_mo_298_> Segment:        1  1:A  207:A     1613\n",
      "randomi_498_> Atoms,selected atoms,random_seed,amplitude:     1613      457        1        4.0000\n",
      "randomi_496_> Amplitude is > 0; randomization is done.\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1927\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1933\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1938\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1956\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "iupac_m_486_> OE1/2 will be swapped:     -105.1969       13  13\n",
      "iupac_m_487_> NH1/2 swapped:      179.7925       26  26\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:      169.7476       28  28\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      179.7762       63  63\n",
      "iupac_m_487_> NH1/2 swapped:      179.9441       98  98\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      103  103\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0163      104  104\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      107  107\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      109  109\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      110  110\n",
      "iupac_m_485_> OD1/2 will be swapped:     -179.9802      111  111\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      114  114\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0292      124  124\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9657      127  127\n",
      "iupac_m_487_> NH1/2 swapped:      179.9558      129  129\n",
      "iupac_m_486_> OE1/2 will be swapped:      154.4214      135  135\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      136  136\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0239      138  138\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9558      139  139\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9720      140  140\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9407      146  146\n",
      "iupac_m_486_> OE1/2 will be swapped:      179.9604      152  152\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      160  160\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      164  164\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      165  165\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      179  179\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:     -176.2274      180  180\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -180.0000      183  183\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      196  196\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      200  200\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0309      202  202\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      207  207\n",
      "\n",
      "\n",
      ">> ENERGY; Differences between the model's features and restraints:\n",
      "Number of all residues in MODEL                   :      207\n",
      "Number of all, selected real atoms                :     1613     457\n",
      "Number of all, selected pseudo atoms              :        0       0\n",
      "Number of all static, selected restraints         :    12693    1983\n",
      "COVALENT_CYS                                      :        F\n",
      "NONBONDED_SEL_ATOMS                               :        1\n",
      "Number of non-bonded pairs (excluding 1-2,1-3,1-4):      581\n",
      "Dynamic pairs routine                             : 5, NATM_SEL x NATM double loop\n",
      "Atomic shift for contacts update (UPDATE_DYNAMIC) :    0.390\n",
      "LENNARD_JONES_SWITCH                              :    6.500   7.500\n",
      "COULOMB_JONES_SWITCH                              :    6.500   7.500\n",
      "RESIDUE_SPAN_RANGE                                :        0   99999\n",
      "NLOGN_USE                                         :       15\n",
      "CONTACT_SHELL                                     :    4.000\n",
      "DYNAMIC_PAIRS,_SPHERE,_COULOMB,_LENNARD,_MODELLER :        T       T       F       F       F\n",
      "SPHERE_STDV                                       :    0.050\n",
      "RADII_FACTOR                                      :    0.820\n",
      "Current energy                                    :         278.8297\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Summary of the restraint violations: \n",
      "\n",
      "   NUM     ... number of restraints.\n",
      "   NUMVI   ... number of restraints with RVIOL > VIOL_REPORT_CUT[i].\n",
      "   RVIOL   ... relative difference from the best value.\n",
      "   NUMVP   ... number of restraints with -Ln(pdf) > VIOL_REPORT_CUT2[i].\n",
      "   RMS_1   ... RMS(feature, minimally_violated_basis_restraint, NUMB).\n",
      "   RMS_2   ... RMS(feature, best_value, NUMB).\n",
      "   MOL.PDF ... scaled contribution to -Ln(Molecular pdf).\n",
      "\n",
      " #                     RESTRAINT_GROUP      NUM   NUMVI  NUMVP   RMS_1   RMS_2         MOL.PDF     S_i\n",
      "------------------------------------------------------------------------------------------------------\n",
      " 1 Bond length potential              :     478       0      0   0.002   0.002     0.33493       1.000\n",
      " 2 Bond angle potential               :     667       0      2   2.190   2.190      60.920       1.000\n",
      " 3 Stereochemical cosine torsion poten:     343       0      4  53.014  53.014      129.29       1.000\n",
      " 4 Stereochemical improper torsion pot:     181       0      0   0.756   0.756      2.3542       1.000\n",
      " 5 Soft-sphere overlap restraints     :     581       0      0   0.002   0.002     0.25438       1.000\n",
      " 6 Lennard-Jones 6-12 potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 7 Coulomb point-point electrostatic p:       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 8 H-bonding potential                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 9 Distance restraints 1 (CA-CA)      :      29       0      0   0.270   0.270     0.19193       1.000\n",
      "10 Distance restraints 2 (N-O)        :      30       0      0   0.823   0.823      1.6609       1.000\n",
      "11 Mainchain Phi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "12 Mainchain Psi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "13 Mainchain Omega dihedral restraints:      70       0      0   1.673   1.673      2.3117       1.000\n",
      "14 Sidechain Chi_1 dihedral restraints:      43       0      0  86.574  86.574      15.217       1.000\n",
      "15 Sidechain Chi_2 dihedral restraints:      31       0      0  72.307  72.307      6.7678       1.000\n",
      "16 Sidechain Chi_3 dihedral restraints:      21       0      0  41.500  41.500     0.80857       1.000\n",
      "17 Sidechain Chi_4 dihedral restraints:       3       0      0 126.619 126.619      1.9622       1.000\n",
      "18 Disulfide distance restraints      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "19 Disulfide angle restraints         :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "20 Disulfide dihedral angle restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "21 Lower bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "22 Upper bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "23 Distance restraints 3 (SDCH-MNCH)  :      16       0      0   0.738   0.738     0.13766       1.000\n",
      "24 Sidechain Chi_5 dihedral restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "25 Phi/Psi pair of dihedral restraints:      71      50      6  24.015 150.365      56.622       1.000\n",
      "26 Distance restraints 4 (SDCH-SDCH)  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "27 Distance restraints 5 (X-Y)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "28 NMR distance restraints 6 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "29 NMR distance restraints 7 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "30 Minimal distance restraints        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "31 Non-bonded restraints              :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "32 Atomic accessibility restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "33 Atomic density restraints          :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "34 Absolute position restraints       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "35 Dihedral angle difference restraint:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "36 GBSA implicit solvent potential    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "37 EM density fitting potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "38 SAXS restraints                    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "39 Symmetry restraints                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "\n",
      "\n",
      "\n",
      "openf___224_> Open           Target.V99990009\n",
      "# Heavy relative violation of each residue is written to: Target.V99990009\n",
      "# The profile is NOT normalized by the number of restraints.\n",
      "# The profiles are smoothed over a window of residues:    1\n",
      "# The sum of all numbers in the file:   12935.0615\n",
      "\n",
      "\n",
      "\n",
      "List of the violated restraints:\n",
      "   A restraint is violated when the relative difference\n",
      "   from the best value (RVIOL) is larger than CUTOFF.\n",
      "\n",
      "   ICSR   ... index of a restraint in the current set.\n",
      "   RESNO  ... residue numbers of the first two atoms.\n",
      "   ATM    ... IUPAC atom names of the first two atoms.\n",
      "   FEAT   ... the value of the feature in the model.\n",
      "   restr  ... the mean of the basis restraint with the smallest\n",
      "              difference from the model (local minimum).\n",
      "   viol   ... difference from the local minimum.\n",
      "   rviol  ... relative difference from the local minimum.\n",
      "   RESTR  ... the best value (global minimum).\n",
      "   VIOL   ... difference from the best value.\n",
      "   RVIOL  ... relative difference from the best value.\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 25                           : Phi/Psi pair of dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       6.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   5791   1M   2A C   N       7    9  -75.17  -68.20    9.57    0.61  -62.50  167.72   28.13\n",
      "    1          2A   2A N   CA      9   10  151.86  145.30                  -40.90\n",
      "    2   5792   2A   3H C   N      12   14 -100.52 -125.60   25.94    1.03  -63.20  176.29   24.47\n",
      "    2          3H   3H N   CA     14   15  145.41  138.80                  -42.30\n",
      "    3   5793   3H   4A C   N      22   24 -126.52 -134.00   16.93    1.04  -62.50  169.48   31.20\n",
      "    3          4A   4A N   CA     24   25  162.19  147.00                  -40.90\n",
      "    4   5794   4A   5G C   N      27   29 -124.71 -167.20   51.49    0.92   82.20 -154.55   15.89\n",
      "    4          5G   5G N   CA     29   30  145.52  174.60                    8.50\n",
      "    5   5795   5G   6R C   N      31   33 -139.83 -125.20   30.85    1.11  -63.00  169.54   28.12\n",
      "    5          6R   6R N   CA     33   34  167.76  140.60                  -41.10\n",
      "    6   5796   6R   7T C   N      42   44  -83.76  -78.10   32.96    1.24  -63.20  137.18   18.96\n",
      "    6          7T   7T N   CA     44   45 -177.73  149.80                  -42.10\n",
      "    7   5797   7T   8G C   N      49   51  -82.91  -80.20    3.80    0.12   82.20 -124.26    7.30\n",
      "    7          8G   8G N   CA     51   52  176.76  174.10                    8.50\n",
      "    8   5798   8G   9Y C   N      53   55  -63.89  -98.40   48.47    3.70  -63.50  154.17   24.48\n",
      "    8          9Y   9Y N   CA     55   56  162.43  128.40                  -43.40\n",
      "    9   5821  31D  32A C   N     265  267  -76.43  -68.20    8.98    0.63  -62.50  170.78   28.69\n",
      "    9         32A  32A N   CA    267  268  148.89  145.30                  -40.90\n",
      "   10   5822  32A  33G C   N     270  272  174.64 -167.20   41.36    1.69   82.20  158.64   11.67\n",
      "   10         33G  33G N   CA    272  273  137.43  174.60                    8.50\n",
      "   11   5823  33G  34D C   N     274  276 -130.07  -96.50   47.11    1.97  -63.30 -174.79   27.18\n",
      "   11         34D  34D N   CA    276  277  147.25  114.20                  -40.00\n",
      "   12   5825  35V  36G C   N     289  291  174.87 -167.20   32.99    1.32   82.20  166.56   12.11\n",
      "   12         36G  36G N   CA    291  292  146.91  174.60                    8.50\n",
      "   13   5826  36G  37A C   N     293  295  -79.92  -68.20   13.06    0.90  -62.50  168.95   28.57\n",
      "   13         37A  37A N   CA    295  296  151.06  145.30                  -40.90\n",
      "   14   5827  37A  38A C   N     298  300 -130.85 -134.00    3.61    0.08  -62.50 -173.17   34.28\n",
      "   14         38A  38A N   CA    300  301  145.22  147.00                  -40.90\n",
      "   15   5829  39P  40P C   N     310  312  -55.73  -58.70    6.90    0.40  -64.50  176.29   13.62\n",
      "   15         40P  40P N   CA    312  313  -36.73  -30.50                  147.20\n",
      "   16   5830  40P  41G C   N     317  319   87.45   78.70   37.08    0.71   82.20  149.46    7.52\n",
      "   16         41G  41G N   CA    319  320  157.87 -166.10                    8.50\n",
      "   17   5831  41G  42A C   N     321  323 -146.54 -134.00   13.07    0.32  -62.50 -171.78   35.35\n",
      "   17         42A  42A N   CA    323  324  150.69  147.00                  -40.90\n",
      "   18   5832  42A  43A C   N     326  328 -146.44 -134.00   12.84    0.32  -62.50 -171.35   35.41\n",
      "   18         43A  43A N   CA    328  329  150.15  147.00                  -40.90\n",
      "   19   5834  44P  45A C   N     338  340 -125.67 -134.00   12.53    0.40  -62.50 -170.63   28.69\n",
      "   19         45A  45A N   CA    340  341  137.63  147.00                  -40.90\n",
      "   20   5836  46P  47G C   N     350  352 -158.65 -167.20   34.04    1.20   82.20  178.67   13.55\n",
      "   20         47G  47G N   CA    352  353  141.65  174.60                    8.50\n",
      "   21   5838  48I  49F C   N     362  364 -114.50 -124.20   11.65    0.30  -63.20 -173.93   29.26\n",
      "   21         49F  49F N   CA    364  365  136.85  143.30                  -44.30\n",
      "   22   5839  49F  50S C   N     373  375 -140.05 -136.60    5.79    0.20  -64.10 -174.58   18.85\n",
      "   22         50S  50S N   CA    375  376  155.85  151.20                  -35.00\n",
      "   23   5840  50S  51S C   N     379  381 -132.21 -136.60    4.49    0.20  -64.10 -174.21   18.28\n",
      "   23         51S  51S N   CA    381  382  152.15  151.20                  -35.00\n",
      "   24   5841  51S  52Q C   N     385  387 -114.42 -121.10    7.02    0.23  -63.80 -175.11   24.83\n",
      "   24         52Q  52Q N   CA    387  388  137.53  139.70                  -40.30\n",
      "   25   5843  53P  54G C   N     401  403  171.72 -167.20   26.84    0.95   82.20  174.24   12.42\n",
      "   25         54G  54G N   CA    403  404  157.99  174.60                    8.50\n",
      "   26   5844  54G  55H C   N     405  407 -140.35 -125.60   23.17    0.59  -63.20  178.56   27.35\n",
      "   26         55H  55H N   CA    407  408  156.67  138.80                  -42.30\n",
      "   27   5845  55H  56T C   N     415  417 -111.08 -124.80   22.86    0.71  -63.20  174.03   19.55\n",
      "   27         56T  56T N   CA    417  418  125.21  143.50                  -42.10\n",
      "   28   5847  57P  58H C   N     429  431 -123.19 -125.60    2.47    0.07  -63.20 -170.79   27.58\n",
      "   28         58H  58H N   CA    431  432  138.25  138.80                  -42.30\n",
      "   29   5849  59P  60A C   N     446  448 -147.74 -134.00   13.95    0.38  -62.50 -170.06   35.69\n",
      "   29         60A  60A N   CA    448  449  149.37  147.00                  -40.90\n",
      "   30   5850  60A  61A C   N     451  453 -155.78 -134.00   23.92    0.54  -62.50 -172.88   35.64\n",
      "   30         61A  61A N   CA    453  454  156.89  147.00                  -40.90\n",
      "   31   5851  61A  62S C   N     456  458 -110.45 -136.60   43.09    1.50  -64.10  158.86    9.80\n",
      "   31         62S  62S N   CA    458  459  116.95  151.20                  -35.00\n",
      "   32   5852  62S  63R C   N     462  464 -163.51 -125.20   56.78    1.77  -63.00  169.42   29.46\n",
      "   32         63R  63R N   CA    464  465 -177.49  140.60                  -41.10\n",
      "   33   5853  63R  64D C   N     473  475  -58.68  -70.90   28.77    0.97  -63.30  164.31   20.56\n",
      "   33         64D  64D N   CA    475  476  124.25  150.30                  -40.00\n",
      "   34   5857  67A  68R C   N     500  502 -145.82 -125.20   20.89    0.93  -63.00 -163.38   22.45\n",
      "   34         68R  68R N   CA    502  503  137.23  140.60                  -41.10\n",
      "   35   5858  68R  69T C   N     511  513 -134.25 -124.80   13.51    0.81  -63.20 -170.24   20.56\n",
      "   35         69T  69T N   CA    513  514  133.85  143.50                  -42.10\n",
      "   36   5859  69T  70S C   N     518  520 -160.88 -136.60   24.72    1.10  -64.10 -157.01   21.67\n",
      "   36         70S  70S N   CA    520  521  146.56  151.20                  -35.00\n",
      "   37   5861  71P  72L C   N     531  533 -107.71 -108.50   15.99    0.86  -63.50  163.81   20.67\n",
      "   37         72L  72L N   CA    533  534  116.53  132.50                  -41.20\n",
      "   38   5862  72L  73Q C   N     539  541 -154.62 -121.10   39.59    1.22  -63.80 -176.96   31.58\n",
      "   38         73Q  73Q N   CA    541  542  160.78  139.70                  -40.30\n",
      "   39   5863  73Q  74T C   N     548  550 -120.77 -124.80    5.31    0.32  -63.20 -179.62   26.95\n",
      "   39         74T  74T N   CA    550  551  146.96  143.50                  -42.10\n",
      "   40   5865  75P  76A C   N     562  564 -132.48 -134.00    1.53    0.06  -62.50 -174.41   34.16\n",
      "   40         76A  76A N   CA    564  565  147.21  147.00                  -40.90\n",
      "   41   5866  76A  77A C   N     567  569 -132.66 -134.00    2.65    0.16  -62.50 -176.26   33.87\n",
      "   41         77A  77A N   CA    569  570  149.28  147.00                  -40.90\n",
      "   42   5868  78P  79G C   N     579  581 -112.58  -80.20   52.62    3.56   82.20 -153.35   16.12\n",
      "   42         79G  79G N   CA    581  582  132.62  174.10                    8.50\n",
      "   43   5869  79G  80A C   N     583  585  -78.47  -68.20   15.68    0.99  -62.50  162.74   27.48\n",
      "   43         80A  80A N   CA    585  586  157.15  145.30                  -40.90\n",
      "   44   5870  80A  81A C   N     588  590  -68.09  -68.20    4.97    0.40  -62.50  168.93   27.96\n",
      "   44         81A  81A N   CA    590  591  150.27  145.30                  -40.90\n",
      "   45   5871  81A  82A C   N     593  595 -111.25 -134.00   23.64    0.58  -62.50 -174.94   32.91\n",
      "   45         82A  82A N   CA    595  596  140.58  147.00                  -40.90\n",
      "   46   5872  82A  83G C   N     598  600   94.33   78.70   52.64    0.91   82.20  135.67    7.19\n",
      "   46         83G  83G N   CA    600  601  143.63 -166.10                    8.50\n",
      "   47   5874  84P  85A C   N     609  611  -75.30  -68.20    7.15    0.58  -62.50  173.46   29.07\n",
      "   47         85A  85A N   CA    611  612  146.11  145.30                  -40.90\n",
      "   48   5875  85A  86L C   N     614  616 -118.49 -108.50   10.93    0.48  -63.50 -173.55   23.37\n",
      "   48         86L  86L N   CA    616  617  136.95  132.50                  -41.20\n",
      "   49   5876  86L  87S C   N     622  624 -121.49 -136.60   24.95    0.87  -64.10  175.96   10.84\n",
      "   49         87S  87S N   CA    624  625  131.34  151.20                  -35.00\n",
      "   50   5880  90P  91P C   N     649  651  -77.31  -64.50   16.80    1.13  -58.70  172.45   15.78\n",
      "   50         91P  91P N   CA    651  652  158.06  147.20                  -30.50\n",
      "\n",
      "\n",
      "report______> Distribution of short non-bonded contacts:\n",
      "\n",
      "  serious non-bonded atom clash:    90 1354   1.714\n",
      "  serious non-bonded atom clash:   122 1320   2.215\n",
      "  serious non-bonded atom clash:   144  727   2.264\n",
      "  serious non-bonded atom clash:   151 1196   1.945\n",
      "  serious non-bonded atom clash:   190 1197   1.758\n",
      "  serious non-bonded atom clash:   664  694   2.295\n",
      "  serious non-bonded atom clash:   671  702   1.565\n",
      "  serious non-bonded atom clash:   719 1120   2.189\n",
      "  serious non-bonded atom clash:   720 1119   1.741\n",
      "  serious non-bonded atom clash:   721 1119   1.648\n",
      "  serious non-bonded atom clash:   726 1152   1.768\n",
      "  serious non-bonded atom clash:   726 1154   2.269\n",
      "  serious non-bonded atom clash:   727 1152   2.294\n",
      "  serious non-bonded atom clash:   728 1152   2.162\n",
      "  serious non-bonded atom clash:   728 1154   1.683\n",
      "  serious non-bonded atom clash:   729 1152   2.041\n",
      "  serious non-bonded atom clash:   729 1153   2.155\n",
      "  serious non-bonded atom clash:   729 1154   1.124\n",
      "  serious non-bonded atom clash:   757  797   2.239\n",
      "  serious non-bonded atom clash:   757  798   2.053\n",
      "  serious non-bonded atom clash:   809  817   2.094\n",
      "  serious non-bonded atom clash:   823  827   2.273\n",
      "  serious non-bonded atom clash:   882 1210   2.153\n",
      "  serious non-bonded atom clash:  1000 1371   1.984\n",
      "  serious non-bonded atom clash:  1019 1041   1.587\n",
      "  serious non-bonded atom clash:  1019 1043   1.762\n",
      "  serious non-bonded atom clash:  1063 1067   2.231\n",
      "  serious non-bonded atom clash:  1111 1455   1.857\n",
      "  serious non-bonded atom clash:  1111 1457   1.212\n",
      "  serious non-bonded atom clash:  1111 1459   1.921\n",
      "  serious non-bonded atom clash:  1139 1383   2.285\n",
      "  serious non-bonded atom clash:  1141 1382   2.235\n",
      "  serious non-bonded atom clash:  1141 1383   2.023\n",
      "  serious non-bonded atom clash:  1168 1342   1.700\n",
      "  serious non-bonded atom clash:  1168 1343   2.160\n",
      "  serious non-bonded atom clash:  1169 1340   1.956\n",
      "  serious non-bonded atom clash:  1169 1342   2.045\n",
      "  serious non-bonded atom clash:  1169 1343   1.522\n",
      "  serious non-bonded atom clash:  1169 1345   2.168\n",
      "  serious non-bonded atom clash:  1170 1312   1.597\n",
      "  serious non-bonded atom clash:  1170 1339   2.166\n",
      "  serious non-bonded atom clash:  1170 1340   0.933\n",
      "  serious non-bonded atom clash:  1170 1342   2.001\n",
      "  serious non-bonded atom clash:  1170 1343   0.763\n",
      "  serious non-bonded atom clash:  1170 1344   1.750\n",
      "  serious non-bonded atom clash:  1170 1345   1.567\n",
      "  serious non-bonded atom clash:  1170 1346   2.186\n",
      "  serious non-bonded atom clash:  1170 1347   2.117\n",
      "  serious non-bonded atom clash:  1172 1310   2.198\n",
      "  serious non-bonded atom clash:  1172 1312   1.497\n",
      "  serious non-bonded atom clash:  1172 1313   2.013\n",
      "  serious non-bonded atom clash:  1172 1314   1.631\n",
      "  serious non-bonded atom clash:  1172 1340   1.411\n",
      "  serious non-bonded atom clash:  1172 1343   2.110\n",
      "  serious non-bonded atom clash:  1172 1344   0.608\n",
      "  serious non-bonded atom clash:  1172 1346   1.043\n",
      "  serious non-bonded atom clash:  1172 1347   1.878\n",
      "  serious non-bonded atom clash:  1174 1313   2.173\n",
      "  serious non-bonded atom clash:  1174 1315   2.177\n",
      "  serious non-bonded atom clash:  1174 1344   1.525\n",
      "  serious non-bonded atom clash:  1174 1346   1.595\n",
      "  serious non-bonded atom clash:  1195 1289   1.622\n",
      "  serious non-bonded atom clash:  1196 1289   1.533\n",
      "  serious non-bonded atom clash:  1197 1287   2.140\n",
      "  serious non-bonded atom clash:  1197 1288   1.553\n",
      "  serious non-bonded atom clash:  1197 1289   2.144\n",
      "  serious non-bonded atom clash:  1216 1262   2.299\n",
      "  serious non-bonded atom clash:  1218 1284   1.935\n",
      "  serious non-bonded atom clash:  1219 1262   2.021\n",
      "  serious non-bonded atom clash:  1219 1277   1.863\n",
      "  serious non-bonded atom clash:  1219 1278   2.125\n",
      "  serious non-bonded atom clash:  1219 1279   1.883\n",
      "  serious non-bonded atom clash:  1219 1280   2.163\n",
      "  serious non-bonded atom clash:  1262 1279   1.800\n",
      "  serious non-bonded atom clash:  1308 1346   1.922\n",
      "  serious non-bonded atom clash:  1309 1346   1.814\n",
      "  serious non-bonded atom clash:  1309 1347   1.856\n",
      "  serious non-bonded atom clash:  1310 1345   1.752\n",
      "  serious non-bonded atom clash:  1310 1346   1.549\n",
      "  serious non-bonded atom clash:  1310 1347   0.442\n",
      "  serious non-bonded atom clash:  1311 1345   1.843\n",
      "  serious non-bonded atom clash:  1311 1347   1.407\n",
      "  serious non-bonded atom clash:  1312 1340   2.114\n",
      "  serious non-bonded atom clash:  1312 1343   1.842\n",
      "  serious non-bonded atom clash:  1312 1344   1.964\n",
      "  serious non-bonded atom clash:  1312 1345   1.299\n",
      "  serious non-bonded atom clash:  1312 1346   1.407\n",
      "  serious non-bonded atom clash:  1312 1347   0.954\n",
      "  serious non-bonded atom clash:  1313 1344   1.825\n",
      "  serious non-bonded atom clash:  1313 1346   1.568\n",
      "  serious non-bonded atom clash:  1314 1344   1.208\n",
      "  serious non-bonded atom clash:  1314 1346   1.796\n",
      "  serious non-bonded atom clash:  1436 1474   1.768\n",
      "  serious non-bonded atom clash:  1436 1475   0.966\n",
      "  serious non-bonded atom clash:  1437 1475   2.218\n",
      "  serious non-bonded atom clash:  1438 1475   1.747\n",
      "  serious non-bonded atom clash:  1486 1489   2.181\n",
      "  serious non-bonded atom clash:  1486 1522   2.155\n",
      "  serious non-bonded atom clash:  1489 1522   2.077\n",
      "  serious non-bonded atom clash:  1490 1519   1.863\n",
      "  serious non-bonded atom clash:  1490 1520   2.124\n",
      "  serious non-bonded atom clash:  1491 1519   1.983\n",
      "  serious non-bonded atom clash:  1491 1520   1.662\n",
      "  serious non-bonded atom clash:  1491 1521   1.787\n",
      "  serious non-bonded atom clash:  1492 1494   2.068\n",
      "  serious non-bonded atom clash:  1564 1583   1.439\n",
      "  serious non-bonded atom clash:  1564 1585   2.034\n",
      "  serious non-bonded atom clash:  1575 1583   2.278\n",
      "  serious non-bonded atom clash:  1576 1583   2.243\n",
      "\n",
      "DISTANCE1:  0.00 2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40\n",
      "DISTANCE2:  2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40 3.50\n",
      "FREQUENCY:    74   21   14    7   24   31   38   75  104  143  155  202  219  217  279\n",
      "\n",
      "\n",
      "<< end of ENERGY.\n",
      "openf___224_> Open           Target.B99990009.pdb\n",
      "wrpdb___568_> Residues, atoms, selected atoms:      207     1613     1613\n",
      "openf___224_> Open           Target.ini\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      3936160    3843.906     3.754\n",
      "read_mo_297_> Segments, residues, atoms:        1      207     1613\n",
      "read_mo_298_> Segment:        1  1:A  207:A     1613\n",
      "randomi_498_> Atoms,selected atoms,random_seed,amplitude:     1613      457        1        4.0000\n",
      "randomi_496_> Amplitude is > 0; randomization is done.\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1927\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1933\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1938\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1944\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1956\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      457     1983\n",
      "iupac_m_486_> OE1/2 will be swapped:     -105.1969       13  13\n",
      "iupac_m_487_> NH1/2 swapped:      179.7925       26  26\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:      169.7476       28  28\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_485_> OD1/2 will be swapped:      128.7383       34  34\n",
      "iupac_m_487_> NH1/2 swapped:      179.9441       98  98\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      103  103\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0163      104  104\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      107  107\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      109  109\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      110  110\n",
      "iupac_m_485_> OD1/2 will be swapped:     -179.9802      111  111\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      114  114\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0292      124  124\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9657      127  127\n",
      "iupac_m_487_> NH1/2 swapped:      179.9558      129  129\n",
      "iupac_m_486_> OE1/2 will be swapped:      154.4214      135  135\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      136  136\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0239      138  138\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9558      139  139\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9720      140  140\n",
      "iupac_m_487_> NH1/2 swapped:     -179.9407      146  146\n",
      "iupac_m_486_> OE1/2 will be swapped:      179.9604      152  152\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      160  160\n",
      "iupac_m_487_> NH1/2 swapped:      179.9720      164  164\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      165  165\n",
      "iupac_m_486_> OE1/2 will be swapped:     -179.9802      179  179\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:     -176.2274      180  180\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:     -180.0000      183  183\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      196  196\n",
      "iupac_m_486_> OE1/2 will be swapped:      180.0000      200  200\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       90.0309      202  202\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_487_> NH1/2 swapped:      180.0000      207  207\n",
      "\n",
      "\n",
      ">> ENERGY; Differences between the model's features and restraints:\n",
      "Number of all residues in MODEL                   :      207\n",
      "Number of all, selected real atoms                :     1613     457\n",
      "Number of all, selected pseudo atoms              :        0       0\n",
      "Number of all static, selected restraints         :    12693    1983\n",
      "COVALENT_CYS                                      :        F\n",
      "NONBONDED_SEL_ATOMS                               :        1\n",
      "Number of non-bonded pairs (excluding 1-2,1-3,1-4):      570\n",
      "Dynamic pairs routine                             : 5, NATM_SEL x NATM double loop\n",
      "Atomic shift for contacts update (UPDATE_DYNAMIC) :    0.390\n",
      "LENNARD_JONES_SWITCH                              :    6.500   7.500\n",
      "COULOMB_JONES_SWITCH                              :    6.500   7.500\n",
      "RESIDUE_SPAN_RANGE                                :        0   99999\n",
      "NLOGN_USE                                         :       15\n",
      "CONTACT_SHELL                                     :    4.000\n",
      "DYNAMIC_PAIRS,_SPHERE,_COULOMB,_LENNARD,_MODELLER :        T       T       F       F       F\n",
      "SPHERE_STDV                                       :    0.050\n",
      "RADII_FACTOR                                      :    0.820\n",
      "Current energy                                    :         328.5452\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Summary of the restraint violations: \n",
      "\n",
      "   NUM     ... number of restraints.\n",
      "   NUMVI   ... number of restraints with RVIOL > VIOL_REPORT_CUT[i].\n",
      "   RVIOL   ... relative difference from the best value.\n",
      "   NUMVP   ... number of restraints with -Ln(pdf) > VIOL_REPORT_CUT2[i].\n",
      "   RMS_1   ... RMS(feature, minimally_violated_basis_restraint, NUMB).\n",
      "   RMS_2   ... RMS(feature, best_value, NUMB).\n",
      "   MOL.PDF ... scaled contribution to -Ln(Molecular pdf).\n",
      "\n",
      " #                     RESTRAINT_GROUP      NUM   NUMVI  NUMVP   RMS_1   RMS_2         MOL.PDF     S_i\n",
      "------------------------------------------------------------------------------------------------------\n",
      " 1 Bond length potential              :     478       0      0   0.002   0.002     0.70326       1.000\n",
      " 2 Bond angle potential               :     667       0      1   2.195   2.195      60.857       1.000\n",
      " 3 Stereochemical cosine torsion poten:     343       0     10  54.502  54.502      133.90       1.000\n",
      " 4 Stereochemical improper torsion pot:     181       0      0   0.798   0.798      2.6987       1.000\n",
      " 5 Soft-sphere overlap restraints     :     570       0      0   0.003   0.003     0.69708       1.000\n",
      " 6 Lennard-Jones 6-12 potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 7 Coulomb point-point electrostatic p:       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 8 H-bonding potential                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 9 Distance restraints 1 (CA-CA)      :      29       0      0   0.460   0.460     0.86741       1.000\n",
      "10 Distance restraints 2 (N-O)        :      30       0      0   1.709   1.709      11.722       1.000\n",
      "11 Mainchain Phi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "12 Mainchain Psi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "13 Mainchain Omega dihedral restraints:      70       0      0   2.334   2.334      4.4960       1.000\n",
      "14 Sidechain Chi_1 dihedral restraints:      43       0      0  75.183  75.183      11.939       1.000\n",
      "15 Sidechain Chi_2 dihedral restraints:      31       0      0  64.280  64.280      4.4540       1.000\n",
      "16 Sidechain Chi_3 dihedral restraints:      21       0      0  56.261  56.261      4.4384       1.000\n",
      "17 Sidechain Chi_4 dihedral restraints:       3       0      0 146.063 146.063      2.0596       1.000\n",
      "18 Disulfide distance restraints      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "19 Disulfide angle restraints         :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "20 Disulfide dihedral angle restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "21 Lower bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "22 Upper bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "23 Distance restraints 3 (SDCH-MNCH)  :      16       0      0   0.981   0.981     0.24933       1.000\n",
      "24 Sidechain Chi_5 dihedral restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "25 Phi/Psi pair of dihedral restraints:      71      51      9  30.928 149.644      89.468       1.000\n",
      "26 Distance restraints 4 (SDCH-SDCH)  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "27 Distance restraints 5 (X-Y)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "28 NMR distance restraints 6 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "29 NMR distance restraints 7 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "30 Minimal distance restraints        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "31 Non-bonded restraints              :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "32 Atomic accessibility restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "33 Atomic density restraints          :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "34 Absolute position restraints       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "35 Dihedral angle difference restraint:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "36 GBSA implicit solvent potential    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "37 EM density fitting potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "38 SAXS restraints                    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "39 Symmetry restraints                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "\n",
      "\n",
      "\n",
      "openf___224_> Open           Target.V99990010\n",
      "# Heavy relative violation of each residue is written to: Target.V99990010\n",
      "# The profile is NOT normalized by the number of restraints.\n",
      "# The profiles are smoothed over a window of residues:    1\n",
      "# The sum of all numbers in the file:   12661.1787\n",
      "\n",
      "\n",
      "\n",
      "List of the violated restraints:\n",
      "   A restraint is violated when the relative difference\n",
      "   from the best value (RVIOL) is larger than CUTOFF.\n",
      "\n",
      "   ICSR   ... index of a restraint in the current set.\n",
      "   RESNO  ... residue numbers of the first two atoms.\n",
      "   ATM    ... IUPAC atom names of the first two atoms.\n",
      "   FEAT   ... the value of the feature in the model.\n",
      "   restr  ... the mean of the basis restraint with the smallest\n",
      "              difference from the model (local minimum).\n",
      "   viol   ... difference from the local minimum.\n",
      "   rviol  ... relative difference from the local minimum.\n",
      "   RESTR  ... the best value (global minimum).\n",
      "   VIOL   ... difference from the best value.\n",
      "   RVIOL  ... relative difference from the best value.\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 25                           : Phi/Psi pair of dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       6.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   5791   1M   2A C   N       7    9 -125.55 -134.00   10.79    0.29  -62.50 -170.39   34.44\n",
      "    1          2A   2A N   CA      9   10  140.28  147.00                  -40.90\n",
      "    2   5792   2A   3H C   N      12   14  -67.91  -67.60    2.41    0.19  -63.20  179.98   22.21\n",
      "    2          3H   3H N   CA     14   15  137.62  140.00                  -42.30\n",
      "    3   5793   3H   4A C   N      22   24 -100.50  -68.20   37.97    3.71  -62.50  170.53   26.35\n",
      "    3          4A   4A N   CA     24   25  125.35  145.30                  -40.90\n",
      "    4   5794   4A   5G C   N      27   29 -110.71  -80.20   52.84    3.50   82.20 -152.84   16.17\n",
      "    4          5G   5G N   CA     29   30  130.96  174.10                    8.50\n",
      "    5   5795   5G   6R C   N      31   33 -153.65 -125.20   38.75    1.17  -63.00  176.97   29.98\n",
      "    5          6R   6R N   CA     33   34  166.91  140.60                  -41.10\n",
      "    6   5796   6R   7T C   N      42   44 -137.97 -124.80   50.55    1.80  -63.20  146.17   23.52\n",
      "    6          7T   7T N   CA     44   45 -167.69  143.50                  -42.10\n",
      "    7   5798   8G   9Y C   N      53   55  -88.01  -98.40   47.94    5.04  -63.50  127.38   19.01\n",
      "    7          9Y   9Y N   CA     55   56   81.60  128.40                  -43.40\n",
      "    8   5799   9Y  10D C   N      65   67   31.25  -63.30   97.45   14.15  -63.30   97.45   14.15\n",
      "    8         10D  10D N   CA     67   68  -63.60  -40.00                  -40.00\n",
      "    9   5821  31D  32A C   N     265  267 -109.24 -134.00   24.78    0.78  -62.50  179.46   31.88\n",
      "    9         32A  32A N   CA    267  268  145.84  147.00                  -40.90\n",
      "   10   5822  32A  33G C   N     270  272  -71.44  -80.20   36.00    1.01   82.20 -158.30    6.94\n",
      "   10         33G  33G N   CA    272  273  139.18  174.10                    8.50\n",
      "   11   5823  33G  34D C   N     274  276 -115.21  -96.50   37.89    1.59  -63.30 -179.52   25.62\n",
      "   11         34D  34D N   CA    276  277  147.15  114.20                  -40.00\n",
      "   12   5825  35V  36G C   N     289  291  -81.56  -80.20   12.97    0.44   82.20 -117.77    7.25\n",
      "   12         36G  36G N   CA    291  292 -173.00  174.10                    8.50\n",
      "   13   5826  36G  37A C   N     293  295  -63.65  -68.20   10.35    0.68  -62.50  176.91   28.94\n",
      "   13         37A  37A N   CA    295  296  136.01  145.30                  -40.90\n",
      "   14   5829  39P  40P C   N     310  312  -71.48  -58.70   15.35    1.31  -64.50  169.34   13.01\n",
      "   14         40P  40P N   CA    312  313  -22.00  -30.50                  147.20\n",
      "   15   5830  40P  41G C   N     317  319  160.52 -167.20   32.39    0.56   82.20 -173.87   12.64\n",
      "   15         41G  41G N   CA    319  320  177.35  174.60                    8.50\n",
      "   16   5831  41G  42A C   N     321  323  -73.73  -68.20   10.50    0.67  -62.50  165.25   27.65\n",
      "   16         42A  42A N   CA    323  324  154.23  145.30                  -40.90\n",
      "   17   5832  42A  43A C   N     326  328  -70.33  -68.20   11.26    0.83  -62.50  162.93   27.09\n",
      "   17         43A  43A N   CA    328  329  156.36  145.30                  -40.90\n",
      "   18   5834  44P  45A C   N     338  340 -125.46 -134.00   11.14    0.63  -62.50  176.55   32.30\n",
      "   18         45A  45A N   CA    340  341  154.15  147.00                  -40.90\n",
      "   19   5836  46P  47G C   N     350  352 -163.02 -167.20    4.26    0.07   82.20 -158.75   14.72\n",
      "   19         47G  47G N   CA    352  353  173.81  174.60                    8.50\n",
      "   20   5838  48I  49F C   N     362  364 -115.11 -124.20    9.13    0.33  -63.20 -179.16   28.57\n",
      "   20         49F  49F N   CA    364  365  142.47  143.30                  -44.30\n",
      "   21   5839  49F  50S C   N     373  375 -124.85 -136.60   12.19    0.41  -64.10 -172.84   17.82\n",
      "   21         50S  50S N   CA    375  376  147.96  151.20                  -35.00\n",
      "   22   5840  50S  51S C   N     379  381 -135.37 -136.60    6.13    0.33  -64.10 -177.69   18.27\n",
      "   22         51S  51S N   CA    381  382  157.20  151.20                  -35.00\n",
      "   23   5841  51S  52Q C   N     385  387 -109.49 -121.10   11.83    0.49  -63.80 -176.52   29.42\n",
      "   23         52Q  52Q N   CA    387  388  142.00  139.70                  -40.30\n",
      "   24   5842  52Q  53P C   N     394  396  -66.62  -58.70    9.38    0.82  -64.50  172.67   13.00\n",
      "   24         53P  53P N   CA    396  397  -25.46  -30.50                  147.20\n",
      "   25   5843  53P  54G C   N     401  403   78.46   78.70   14.70    0.43   82.20  170.74    8.10\n",
      "   25         54G  54G N   CA    403  404  179.20 -166.10                    8.50\n",
      "   26   5844  54G  55H C   N     405  407 -130.17 -125.60   15.00    0.48  -63.20  177.72   26.60\n",
      "   26         55H  55H N   CA    407  408  153.08  138.80                  -42.30\n",
      "   27   5845  55H  56T C   N     415  417 -112.73 -124.80   16.78    0.55  -63.20 -179.14   20.33\n",
      "   27         56T  56T N   CA    417  418  131.84  143.50                  -42.10\n",
      "   28   5847  57P  58H C   N     429  431 -127.48 -125.60    4.09    0.12  -63.20 -173.32   27.55\n",
      "   28         58H  58H N   CA    431  432  142.43  138.80                  -42.30\n",
      "   29   5849  59P  60A C   N     446  448 -115.60 -134.00   21.38    0.50  -62.50 -175.20   28.20\n",
      "   29         60A  60A N   CA    448  449  136.11  147.00                  -40.90\n",
      "   30   5850  60A  61A C   N     451  453 -162.50 -134.00   33.07    0.77  -62.50 -175.26   35.58\n",
      "   30         61A  61A N   CA    453  454  163.77  147.00                  -40.90\n",
      "   31   5851  61A  62S C   N     456  458  -75.16  -72.40    9.95    0.50  -64.10  163.42   12.53\n",
      "   31         62S  62S N   CA    458  459  161.96  152.40                  -35.00\n",
      "   32   5852  62S  63R C   N     462  464  -89.46  -72.10   51.17    4.27  -63.00  137.43   16.94\n",
      "   32         63R  63R N   CA    464  465   93.76  141.90                  -41.10\n",
      "   33   5853  63R  64D C   N     473  475 -128.86  -96.50   34.21    1.39  -63.30  157.41   16.75\n",
      "   33         64D  64D N   CA    475  476  103.11  114.20                  -40.00\n",
      "   34   5856  66V  67A C   N     495  497 -104.93 -134.00   39.21    1.11  -62.50  167.05   25.65\n",
      "   34         67A  67A N   CA    497  498  120.67  147.00                  -40.90\n",
      "   35   5857  67A  68R C   N     500  502 -165.02 -125.20   44.86    1.36  -63.00 -172.23   32.15\n",
      "   35         68R  68R N   CA    502  503  161.26  140.60                  -41.10\n",
      "   36   5858  68R  69T C   N     511  513  -77.49  -78.10    2.00    0.11  -63.20  166.81   22.34\n",
      "   36         69T  69T N   CA    513  514  151.70  149.80                  -42.10\n",
      "   37   5859  69T  70S C   N     518  520  -75.15  -72.40    3.08    0.17  -64.10  171.55   13.12\n",
      "   37         70S  70S N   CA    520  521  153.80  152.40                  -35.00\n",
      "   38   5861  71P  72L C   N     531  533 -111.38 -108.50    3.52    0.19  -63.50  178.23   22.49\n",
      "   38         72L  72L N   CA    533  534  130.48  132.50                  -41.20\n",
      "   39   5862  72L  73Q C   N     539  541  -87.17  -73.00   16.32    1.00  -63.80  172.49   26.61\n",
      "   39         73Q  73Q N   CA    541  542  148.80  140.70                  -40.30\n",
      "   40   5863  73Q  74T C   N     548  550  -96.03 -124.80   53.87    1.67  -63.20  143.85   16.49\n",
      "   40         74T  74T N   CA    550  551   97.95  143.50                  -42.10\n",
      "   41   5865  75P  76A C   N     562  564 -136.87 -134.00    7.82    0.48  -62.50 -165.82   35.81\n",
      "   41         76A  76A N   CA    564  565  139.73  147.00                  -40.90\n",
      "   42   5866  76A  77A C   N     567  569 -134.92 -134.00    8.48    0.50  -62.50 -166.47   29.17\n",
      "   42         77A  77A N   CA    569  570  138.57  147.00                  -40.90\n",
      "   43   5868  78P  79G C   N     579  581  170.15 -167.20   37.27    1.46   82.20  162.38   11.73\n",
      "   43         79G  79G N   CA    581  582  145.00  174.60                    8.50\n",
      "   44   5869  79G  80A C   N     583  585  175.05 -134.00   54.46    1.26  -62.50 -164.12   38.49\n",
      "   44         80A  80A N   CA    585  586  166.22  147.00                  -40.90\n",
      "   45   5870  80A  81A C   N     588  590  -68.97  -68.20    5.25    0.45  -62.50  179.11   29.67\n",
      "   45         81A  81A N   CA    590  591  140.11  145.30                  -40.90\n",
      "   46   5871  81A  82A C   N     593  595 -142.66 -134.00    8.66    0.28  -62.50 -170.25   35.39\n",
      "   46         82A  82A N   CA    595  596  147.11  147.00                  -40.90\n",
      "   47   5874  84P  85A C   N     609  611  -83.54  -68.20   15.35    1.28  -62.50  174.35   29.65\n",
      "   47         85A  85A N   CA    611  612  146.03  145.30                  -40.90\n",
      "   48   5875  85A  86L C   N     614  616 -126.79 -108.50   20.40    1.09  -63.50  176.41   21.83\n",
      "   48         86L  86L N   CA    616  617  123.46  132.50                  -41.20\n",
      "   49   5876  86L  87S C   N     622  624 -128.81 -136.60   43.57    2.04  -64.10  157.26    9.85\n",
      "   49         87S  87S N   CA    624  625  108.33  151.20                  -35.00\n",
      "   50   5880  90P  91P C   N     649  651  -79.14  -64.50   16.46    1.30  -58.70  175.98   16.23\n",
      "   50         91P  91P N   CA    651  652  154.72  147.20                  -30.50\n",
      "   51   5881  91P  92V C   N     656  658 -148.26  -62.40   87.12   13.06  -62.40   87.12   13.06\n",
      "   51         92V  92V N   CA    658  659  -27.62  -42.40                  -42.40\n",
      "\n",
      "\n",
      "report______> Distribution of short non-bonded contacts:\n",
      "\n",
      "  serious non-bonded atom clash:    90 1354   1.714\n",
      "  serious non-bonded atom clash:   122 1320   2.215\n",
      "  serious non-bonded atom clash:   144  727   2.264\n",
      "  serious non-bonded atom clash:   151 1196   1.945\n",
      "  serious non-bonded atom clash:   190 1197   1.758\n",
      "  serious non-bonded atom clash:   664  694   2.295\n",
      "  serious non-bonded atom clash:   671  702   1.565\n",
      "  serious non-bonded atom clash:   719 1120   2.189\n",
      "  serious non-bonded atom clash:   720 1119   1.741\n",
      "  serious non-bonded atom clash:   721 1119   1.648\n",
      "  serious non-bonded atom clash:   726 1152   1.768\n",
      "  serious non-bonded atom clash:   726 1154   2.269\n",
      "  serious non-bonded atom clash:   727 1152   2.294\n",
      "  serious non-bonded atom clash:   728 1152   2.162\n",
      "  serious non-bonded atom clash:   728 1154   1.683\n",
      "  serious non-bonded atom clash:   729 1152   2.041\n",
      "  serious non-bonded atom clash:   729 1153   2.155\n",
      "  serious non-bonded atom clash:   729 1154   1.124\n",
      "  serious non-bonded atom clash:   757  797   2.239\n",
      "  serious non-bonded atom clash:   757  798   2.053\n",
      "  serious non-bonded atom clash:   809  817   2.094\n",
      "  serious non-bonded atom clash:   823  827   2.273\n",
      "  serious non-bonded atom clash:   882 1210   2.153\n",
      "  serious non-bonded atom clash:  1000 1371   1.984\n",
      "  serious non-bonded atom clash:  1019 1041   1.587\n",
      "  serious non-bonded atom clash:  1019 1043   1.762\n",
      "  serious non-bonded atom clash:  1063 1067   2.231\n",
      "  serious non-bonded atom clash:  1111 1455   1.857\n",
      "  serious non-bonded atom clash:  1111 1457   1.212\n",
      "  serious non-bonded atom clash:  1111 1459   1.921\n",
      "  serious non-bonded atom clash:  1139 1383   2.285\n",
      "  serious non-bonded atom clash:  1141 1382   2.235\n",
      "  serious non-bonded atom clash:  1141 1383   2.023\n",
      "  serious non-bonded atom clash:  1168 1342   1.700\n",
      "  serious non-bonded atom clash:  1168 1343   2.160\n",
      "  serious non-bonded atom clash:  1169 1340   1.956\n",
      "  serious non-bonded atom clash:  1169 1342   2.045\n",
      "  serious non-bonded atom clash:  1169 1343   1.522\n",
      "  serious non-bonded atom clash:  1169 1345   2.168\n",
      "  serious non-bonded atom clash:  1170 1312   1.597\n",
      "  serious non-bonded atom clash:  1170 1339   2.166\n",
      "  serious non-bonded atom clash:  1170 1340   0.933\n",
      "  serious non-bonded atom clash:  1170 1342   2.001\n",
      "  serious non-bonded atom clash:  1170 1343   0.763\n",
      "  serious non-bonded atom clash:  1170 1344   1.750\n",
      "  serious non-bonded atom clash:  1170 1345   1.567\n",
      "  serious non-bonded atom clash:  1170 1346   2.186\n",
      "  serious non-bonded atom clash:  1170 1347   2.117\n",
      "  serious non-bonded atom clash:  1172 1310   2.198\n",
      "  serious non-bonded atom clash:  1172 1312   1.497\n",
      "  serious non-bonded atom clash:  1172 1313   2.013\n",
      "  serious non-bonded atom clash:  1172 1314   1.631\n",
      "  serious non-bonded atom clash:  1172 1340   1.411\n",
      "  serious non-bonded atom clash:  1172 1343   2.110\n",
      "  serious non-bonded atom clash:  1172 1344   0.608\n",
      "  serious non-bonded atom clash:  1172 1346   1.043\n",
      "  serious non-bonded atom clash:  1172 1347   1.878\n",
      "  serious non-bonded atom clash:  1174 1313   2.173\n",
      "  serious non-bonded atom clash:  1174 1315   2.177\n",
      "  serious non-bonded atom clash:  1174 1344   1.525\n",
      "  serious non-bonded atom clash:  1174 1346   1.595\n",
      "  serious non-bonded atom clash:  1195 1289   1.622\n",
      "  serious non-bonded atom clash:  1196 1289   1.533\n",
      "  serious non-bonded atom clash:  1197 1287   2.140\n",
      "  serious non-bonded atom clash:  1197 1288   1.553\n",
      "  serious non-bonded atom clash:  1197 1289   2.144\n",
      "  serious non-bonded atom clash:  1216 1262   2.299\n",
      "  serious non-bonded atom clash:  1218 1284   1.935\n",
      "  serious non-bonded atom clash:  1219 1262   2.021\n",
      "  serious non-bonded atom clash:  1219 1277   1.863\n",
      "  serious non-bonded atom clash:  1219 1278   2.125\n",
      "  serious non-bonded atom clash:  1219 1279   1.883\n",
      "  serious non-bonded atom clash:  1219 1280   2.163\n",
      "  serious non-bonded atom clash:  1262 1279   1.800\n",
      "  serious non-bonded atom clash:  1308 1346   1.922\n",
      "  serious non-bonded atom clash:  1309 1346   1.814\n",
      "  serious non-bonded atom clash:  1309 1347   1.856\n",
      "  serious non-bonded atom clash:  1310 1345   1.752\n",
      "  serious non-bonded atom clash:  1310 1346   1.549\n",
      "  serious non-bonded atom clash:  1310 1347   0.442\n",
      "  serious non-bonded atom clash:  1311 1345   1.843\n",
      "  serious non-bonded atom clash:  1311 1347   1.407\n",
      "  serious non-bonded atom clash:  1312 1340   2.114\n",
      "  serious non-bonded atom clash:  1312 1343   1.842\n",
      "  serious non-bonded atom clash:  1312 1344   1.964\n",
      "  serious non-bonded atom clash:  1312 1345   1.299\n",
      "  serious non-bonded atom clash:  1312 1346   1.407\n",
      "  serious non-bonded atom clash:  1312 1347   0.954\n",
      "  serious non-bonded atom clash:  1313 1344   1.825\n",
      "  serious non-bonded atom clash:  1313 1346   1.568\n",
      "  serious non-bonded atom clash:  1314 1344   1.208\n",
      "  serious non-bonded atom clash:  1314 1346   1.796\n",
      "  serious non-bonded atom clash:  1436 1474   1.768\n",
      "  serious non-bonded atom clash:  1436 1475   0.966\n",
      "  serious non-bonded atom clash:  1437 1475   2.218\n",
      "  serious non-bonded atom clash:  1438 1475   1.747\n",
      "  serious non-bonded atom clash:  1486 1489   2.181\n",
      "  serious non-bonded atom clash:  1486 1522   2.155\n",
      "  serious non-bonded atom clash:  1489 1522   2.077\n",
      "  serious non-bonded atom clash:  1490 1519   1.863\n",
      "  serious non-bonded atom clash:  1490 1520   2.124\n",
      "  serious non-bonded atom clash:  1491 1519   1.983\n",
      "  serious non-bonded atom clash:  1491 1520   1.662\n",
      "  serious non-bonded atom clash:  1491 1521   1.787\n",
      "  serious non-bonded atom clash:  1492 1494   2.068\n",
      "  serious non-bonded atom clash:  1564 1583   1.439\n",
      "  serious non-bonded atom clash:  1564 1585   2.034\n",
      "  serious non-bonded atom clash:  1575 1583   2.278\n",
      "  serious non-bonded atom clash:  1576 1583   2.243\n",
      "\n",
      "DISTANCE1:  0.00 2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40\n",
      "DISTANCE2:  2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40 3.50\n",
      "FREQUENCY:    74   21   14    7   24   32   37   76  102  139  156  213  220  224  266\n",
      "\n",
      "\n",
      "<< end of ENERGY.\n",
      "openf___224_> Open           Target.B99990010.pdb\n",
      "wrpdb___568_> Residues, atoms, selected atoms:      207     1613     1613\n",
      "\n",
      ">> Summary of successfully produced models:\n",
      "Filename                          molpdf\n",
      "----------------------------------------\n",
      "Target.B99990001.pdb           303.10901\n",
      "Target.B99990002.pdb           259.41501\n",
      "Target.B99990003.pdb           317.37903\n",
      "Target.B99990004.pdb           309.30780\n",
      "Target.B99990005.pdb           264.56122\n",
      "Target.B99990006.pdb           312.92484\n",
      "Target.B99990007.pdb           277.10275\n",
      "Target.B99990008.pdb           296.23157\n",
      "Target.B99990009.pdb           278.82965\n",
      "Target.B99990010.pdb           328.54523\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from modeller import *\n",
    "from modeller.automodel import *    # Load the AutoModel class\n",
    "\n",
    "log.verbose()\n",
    "env = Environ()\n",
    "\n",
    "# directories for input atom files\n",
    "env.io.atom_files_directory = ['.', '../atom_files']\n",
    "\n",
    "class MyModel(AutoModel):\n",
    "    def select_atoms(self):\n",
    "        return Selection(self.residue_range('1:A', '9:A'),\n",
    "                         self.residue_range('32:A', '91:A'))\n",
    "\n",
    "a = MyModel(env, alnfile = 'modeller.ali',\n",
    "            knowns = 'Template', sequence = 'Target')\n",
    "a.starting_model= 1\n",
    "a.ending_model  = 10\n",
    "\n",
    "a.make()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cb61f79",
   "metadata": {},
   "source": [
    "## Build structure using Multiple Templates\n",
    "- Target:    The complete protein sequence from 2xa0_A (206 AA)\n",
    "- Template1: The protein structure from 2xa0_A (Missing residues: 0-9, 32-91)\n",
    "- Template2: The protein structure of FLD from "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6ebdef53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "openf___224_> Open           $(LIB)/restyp.lib\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/resgrp.lib\n",
      "rdresgr_266_> Number of residue groups:        2\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/sstruc.lib\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:      7246856    7077.008     6.911\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:      7247384    7077.523     6.912\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/resdih.lib\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:      7295984    7124.984     6.958\n",
      "rdrdih__263_> Number of dihedral angle types         :        9\n",
      "              Maximal number of dihedral angle optima:        3\n",
      "              Dihedral angle names                   :  Alph Phi Psi Omeg chi1 chi2 chi3 chi4 chi5\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/radii.lib\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:      7309284    7137.973     6.971\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/af_mnchdef.lib\n",
      "rdwilmo_274_> Mainchain residue conformation classes:  APBLE\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/mnch.lib\n",
      "rdclass_257_> Number of classes:        5\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/mnch1.lib\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/mnch2.lib\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/mnch3.lib\n",
      "openf___224_> Open           ${MODINSTALL10v2}/modlib/xs4.mat\n",
      "rdrrwgh_268_> Number of residue types:       21\n",
      "openf___224_> Open           modeller_template_1gjh.ali\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      7338100    7166.113     6.998\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      7353633    7181.282     7.013\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      7369483    7196.761     7.028\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      7401183    7227.718     7.058\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      7464583    7289.632     7.119\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7465127    7290.163     7.119\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7465303    7290.335     7.119\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7466127    7291.140     7.120\n",
      "\n",
      "Read the alignment from file       : modeller_template_1gjh.ali\n",
      "\n",
      "Total number of alignment positions:   207\n",
      "\n",
      "  #  Code        #_Res #_Segm PDB_code    Name\n",
      "-------------------------------------------------------------------------------\n",
      "  1  Template1     137      1      2xa0_A\n",
      "  2  Template2      45      1      1gjh_A\n",
      "  3     Target     207      1\n",
      "check_a_343_> >> BEGINNING OF COMMAND\n",
      "openf___224_> Open           ./2xa0_A.pdb\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7508952    7332.961     7.161\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7508952    7332.961     7.161\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7509123    7333.128     7.161\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7509973    7333.958     7.162\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7511231    7335.187     7.163\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7513135    7337.046     7.165\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7515991    7339.835     7.168\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7515991    7339.835     7.168\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7516081    7339.923     7.168\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7520365    7344.106     7.172\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7520365    7344.106     7.172\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7520500    7344.238     7.172\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7526926    7350.514     7.178\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7526926    7350.514     7.178\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7527124    7350.707     7.178\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7536746    7360.104     7.188\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7536746    7360.104     7.188\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7537043    7360.394     7.188\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7537043    7360.394     7.188\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7537493    7360.833     7.188\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7551943    7374.944     7.202\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      7551826    7374.830     7.202\n",
      "openf___224_> Open           ./1gjh_A.pdb\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7567051    7389.698     7.217\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7567051    7389.698     7.217\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7567222    7389.865     7.217\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7568072    7390.695     7.217\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7569330    7391.924     7.219\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7571234    7393.783     7.220\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7574090    7396.572     7.223\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7574090    7396.572     7.223\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7574180    7396.660     7.223\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7578464    7400.844     7.227\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7578464    7400.844     7.227\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7578599    7400.976     7.228\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7578599    7400.976     7.228\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7578797    7401.169     7.228\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7585223    7407.444     7.234\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7585223    7407.444     7.234\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7585520    7407.734     7.234\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7595142    7417.131     7.243\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7595142    7417.131     7.243\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7595592    7417.570     7.244\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7610042    7431.682     7.258\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      7610042    7431.682     7.258\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7610717    7432.341     7.258\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      7632375    7453.491     7.279\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      7668002    7488.283     7.313\n",
      "read_te_290E> Number of residues in the alignment and  pdb files are different:       45      164\n",
      "              For alignment entry:        2  Template2\n",
      "                                  x  (mismatch at alignment position     46)\n",
      " Alignment   YEWDAGDDVEENRTEAPEGT/                                        \n",
      "       PDB   YEWDAGDDVEENRTEAPEGT-ESEVVHLTLRQAGDDFSRRYRRDFAEMSSQLHLTPFTAR \n",
      "     Match   ********************\n",
      "\n",
      "  Please check your alignment file header to be sure you correctly specified\n",
      "  the starting and ending residue numbers and chains. The alignment sequence\n",
      "  must match that from the atom file exactly.\n",
      "\n",
      "  Another possibility is that some residues in the atom file are missing,\n",
      "  perhaps because they could not be resolved experimentally. (Note that Modeller\n",
      "  reads only the ATOM and HETATM records in PDB, NOT the SEQRES records.)\n",
      "  In this case, simply replace the section of your alignment corresponding\n",
      "  to these missing residues with gaps.\n",
      "read_te_288W> Protein not accepted:        2  Template2\n"
     ]
    },
    {
     "ename": "ModellerError",
     "evalue": "read_te_290E> Number of residues in the alignment and  pdb files are different:       45      164 For alignment entry:        2  Template2",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModellerError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_105311/777323683.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mending_model\u001b[0m  \u001b[0;34m=\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 20\u001b[0;31m \u001b[0ma\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/opt/miniconda3/envs/biobb_env/lib/python3.7/site-packages/modeller/automodel/automodel.py\u001b[0m in \u001b[0;36mmake\u001b[0;34m(self, exit_stage)\u001b[0m\n\u001b[1;32m    139\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    140\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moutputs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 141\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhomcsr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexit_stage\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    142\u001b[0m         \u001b[0;31m# Exit early?\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    143\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mexit_stage\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/miniconda3/envs/biobb_env/lib/python3.7/site-packages/modeller/automodel/automodel.py\u001b[0m in \u001b[0;36mhomcsr\u001b[0;34m(self, exit_stage)\u001b[0m\n\u001b[1;32m    623\u001b[0m             \u001b[0;31m# modfile.delete(file='.tmp.ali')\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    624\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 625\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcheck_alignment\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maln\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    626\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    627\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_alignment_info\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maln\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/miniconda3/envs/biobb_env/lib/python3.7/site-packages/modeller/automodel/automodel.py\u001b[0m in \u001b[0;36mcheck_alignment\u001b[0;34m(self, aln)\u001b[0m\n\u001b[1;32m    576\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mcheck_alignment\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maln\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    577\u001b[0m         \u001b[0;34m\"\"\"Check the alignment for sanity\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 578\u001b[0;31m         \u001b[0maln\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcheck\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    579\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    580\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_get_sequence_info\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maln\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtgt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mknown\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/miniconda3/envs/biobb_env/lib/python3.7/site-packages/modeller/alignment.py\u001b[0m in \u001b[0;36mcheck\u001b[0;34m(self, io)\u001b[0m\n\u001b[1;32m    211\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mcheck\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mio\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    212\u001b[0m         \u001b[0;34m\"\"\"Check alignment for modeling\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 213\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcheck_structure_structure\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mio\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mio\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    214\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcheck_sequence_structure\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mio\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mio\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    215\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/miniconda3/envs/biobb_env/lib/python3.7/site-packages/modeller/alignment.py\u001b[0m in \u001b[0;36mcheck_structure_structure\u001b[0;34m(self, eqvdst, io)\u001b[0m\n\u001b[1;32m    220\u001b[0m             \u001b[0mio\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0menv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mio\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    221\u001b[0m         \u001b[0mf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_modeller\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmod_alignment_check_structures\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 222\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodpt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodpt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0menv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlibs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodpt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0meqvdst\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    223\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    224\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mcheck_sequence_structure\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgapdist\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m8.0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mio\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModellerError\u001b[0m: read_te_290E> Number of residues in the alignment and  pdb files are different:       45      164 For alignment entry:        2  Template2"
     ]
    }
   ],
   "source": [
    "from modeller import *\n",
    "from modeller.automodel import *    # Load the AutoModel class\n",
    "\n",
    "log.verbose()\n",
    "env = Environ()\n",
    "\n",
    "# directories for input atom files\n",
    "env.io.atom_files_directory = ['.', '../atom_files']\n",
    "\n",
    "class MyModel(AutoModel):\n",
    "    def select_atoms(self):\n",
    "        return Selection(self.residue_range('1:A', '9:A'),\n",
    "                         self.residue_range('32:A', '91:A'))\n",
    "\n",
    "a = MyModel(env, alnfile = 'modeller_template_1gjh.ali',\n",
    "            knowns = ('Template1', 'Template2'), sequence = 'Target')\n",
    "a.starting_model= 1\n",
    "a.ending_model  = 2\n",
    "\n",
    "a.make()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "e6bf92d361c19049dbef2997c619a01070cbb788e395c2d00386c10a34a2b941"
  },
  "kernelspec": {
   "display_name": "Python 3.7.12 ('biobb_env')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
